%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english,openany,oneside]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax
\ifdefined\pdfimageresolution
    \pdfimageresolution= \numexpr \dimexpr1in\relax/\sphinxpxdimen\relax
\fi
%% let collapsible pdf bookmarks panel have high depth per default
\PassOptionsToPackage{bookmarksdepth=5}{hyperref}

\PassOptionsToPackage{booktabs}{sphinx}
\PassOptionsToPackage{colorrows}{sphinx}

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{tgtermes}
\usepackage{tgheros}
\renewcommand{\ttdefault}{txtt}



\usepackage[Bjarne]{fncychap}
\usepackage[,numfigreset=1,mathnumfig]{sphinx}
\sphinxsetup{hmargin=0.8in, vmargin={1in,0.9in}}
\fvset{fontsize=auto}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}


\usepackage{sphinxmessages}
\setcounter{tocdepth}{2}

\DeclareRobustCommand{\and}{%
\end{tabular}\kern-\tabcolsep\\\begin{tabular}[t]{c}%
}%

\title{PyPop User Guide}
\date{Jul 26, 2023}
\release{0.0}
\author{Alexander K. Lancaster\and Mark P. Nelson\and Diogo Meyer\and Richard M. Single\and Owen D. Solberg}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\ifdefined\shorthandoff
  \ifnum\catcode`\=\string=\active\shorthandoff{=}\fi
  \ifnum\catcode`\"=\active\shorthandoff{"}\fi
\fi

\pagestyle{empty}
\newcommand\sphinxbackoftitlepage{\sphinxstrong{PyPop User Guide}\\ \\Copyright © 2003-2009 Regents of the University of California.\\Copyright © 2023 PyPop contributors \\ \\Permission is granted to copy, distribute and/or modify this document under the terms of the GNU Free Documentation License, Version 1.2 or any later version published by the Free Software Foundation; with no Invariant Sections no Front-Cover Texts and no Back-Cover Texts. A copy of the license is included in the License chapter.\\ \\Document release: 0.0.post2310+ga91d9e6.d20230726}\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{docs/index::doc}}


\sphinxAtStartPar
\sphinxstylestrong{PyPop (Python for Population Genomics)} is an environment for doing
large\sphinxhyphen{}scale population genetic analyses including:
\begin{itemize}
\item {} 
\sphinxAtStartPar
conformity to Hardy\sphinxhyphen{}Weinberg expectations

\item {} 
\sphinxAtStartPar
tests for balancing or directional selection

\item {} 
\sphinxAtStartPar
estimates of haplotype frequencies and measures and tests of
significance for linkage disequilibrium (LD).

\end{itemize}

\sphinxAtStartPar
It is an object\sphinxhyphen{}oriented framework implemented in \sphinxhref{http://www.pypop.org/}{Python} (http://www.pypop.org/), a language with powerful features for
interfacing with other languages, such as C (in which we have already
implemented many routines and which is particularly suited to
computationally intensive tasks).

\sphinxAtStartPar
The output of the analyses are stored in XML. These output files can
then be transformed using standard tools into many other data formats
suitable for machine input (such as PHYLIP or input for spreadsheet
programs such as Excel or statistical packages, such as R), plain
text, or HTML for human\sphinxhyphen{}readable format. Storing the output in XML
allows the final viewable output format to be redesigned at will,
without requiring the (often time\sphinxhyphen{}consuming) re\sphinxhyphen{}running of the
analyses themselves.

\sphinxAtStartPar
An outline of PyPop can be found in our 2007 \sphinxstyleemphasis{Tissue Antigens} and
2003 \sphinxstyleemphasis{PSB} \DUrole{xref,std,std-ref}{papers}.

\sphinxAtStartPar
\sphinxstylestrong{How to cite PyPop}

\sphinxAtStartPar
When citing PyPop, please cite the (2007) paper from \sphinxstyleemphasis{Tissue Antigens}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
A. K. Lancaster, R. M. Single, O. D. Solberg, M. P. Nelson and
G. Thomson (2007) “PyPop update \sphinxhyphen{} a software pipeline for
large\sphinxhyphen{}scale multilocus population genomics” \sphinxstyleemphasis{Tissue Antigens} 69 (s1), 192\sphinxhyphen{}197.
{[}\sphinxhref{http://dx.doi.org/10.1111/j.1399-0039.2006.00769.x}{journal page} (http://dx.doi.org/10.1111/j.1399\sphinxhyphen{}0039.2006.00769.x),
\sphinxhref{http://pypop.org/tissue-antigens-lancaster-2007.pdf}{preprint PDF (112 kB)} (http://pypop.org/tissue\sphinxhyphen{}antigens\sphinxhyphen{}lancaster\sphinxhyphen{}2007.pdf){]}.

\end{itemize}

\sphinxAtStartPar
In addition, you can also cite our 2003 Pacific Symposium on Biocomputing paper:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Alex Lancaster, Mark P. Nelson, Richard M. Single, Diogo Meyer, and
Glenys Thomson (2003) “PyPop: a software framework for population
genomics: analyzing large\sphinxhyphen{}scale multi\sphinxhyphen{}locus genotype data”, in
\sphinxstyleemphasis{Pacific Symposium on Biocomputing} vol. 8:514\sphinxhyphen{}525 (edited by R B
Altman. et al., World Scientific, Singapore, 2003) {[}\sphinxhref{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3891851/}{PubMed
Central} (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3891851/),
\sphinxhref{http://pypop.org/psb-pypop.pdf}{PDF (344 kB)} (http://pypop.org/psb\sphinxhyphen{}pypop.pdf){]}.

\end{itemize}

\sphinxAtStartPar
PyPop was originally developed for the analysis of data for the 13th
\sphinxhref{http://www.ihwg.org/}{International Histocompatiblity Workshop and
Conference} (http://www.ihwg.org/) held in Seattle, Washington in 2002
(\sphinxcite{docs/biblio:meyer-etal-2007}, \sphinxcite{docs/biblio:single-etal-2007a}, \sphinxcite{docs/biblio:single-etal-2007a}). For more
details on the design and technical details of PyPop, please consult
\sphinxcite{docs/biblio:lancaster-etal-2003}, \sphinxcite{docs/biblio:lancaster-etal-2007a} and \sphinxcite{docs/biblio:lancaster-etal-2007b}.

\phantomsection\label{\detokenize{docs/index:acknowlegements}}
\sphinxAtStartPar
\sphinxstylestrong{Acknowlegements}

\sphinxAtStartPar
This work has benefited from the support of NIH grant AI49213 (13th
IHW) and NIH/NIAID Contract number HHSN266200400076C
N01\sphinxhyphen{}AI\sphinxhyphen{}40076. Thanks to Steven J. Mack, Kristie A. Mather, Steve G.E.
Marsh, Mark Grote and Leslie Louie for helpful comments and testing.

\sphinxAtStartPar
\sphinxstylestrong{How to use this guide}

\sphinxAtStartPar
This guide to PyPop contains four main parts:
\begin{itemize}
\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{docs/guide-chapter-install::doc}]{\sphinxcrossref{\DUrole{doc}{Installing PyPop}}}} describes how to install PyPop,
including pre\sphinxhyphen{}release binaries.

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{docs/guide-chapter-usage::doc}]{\sphinxcrossref{\DUrole{doc}{Getting started with PyPop}}}} describes how to run PyPop.

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{docs/guide-chapter-instructions::doc}]{\sphinxcrossref{\DUrole{doc}{Interpreting PyPop output}}}} details the population genetic
methods and statistics that PyPop computes.

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{docs/guide-chapter-contributing::doc}]{\sphinxcrossref{\DUrole{doc}{Contributing to PyPop}}}} details how to contribute to
ongoing PyPop code and documentation.

\end{itemize}
\phantomsection\label{\detokenize{docs/index:user-guide-toc}}
\sphinxstepscope


\chapter{Installing PyPop}
\label{\detokenize{docs/guide-chapter-install:installing-pypop}}\label{\detokenize{docs/guide-chapter-install::doc}}
\begin{sphinxadmonition}{attention}{Attention:}
\sphinxAtStartPar
The package name for installation purposes is \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}} \sphinxhyphen{} to avoid
conflicting with an unrelated package with the name \sphinxcode{\sphinxupquote{pypop}}
already on \sphinxhref{https://pypi.org}{PyPI} (https://pypi.org). This is a working name and
may change and is not yet the final package name until the package
is released to PyPI.
\end{sphinxadmonition}


\section{Quickstart Guide}
\label{\detokenize{docs/guide-chapter-install:quickstart-guide}}
\sphinxAtStartPar
\sphinxstylestrong{Installing} \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}

\sphinxAtStartPar
If you already have Python and \sphinxcode{\sphinxupquote{pip}} {\hyperref[\detokenize{docs/guide-chapter-install:install-python-3-and-pip}]{\sphinxcrossref{installed}}}, install a test pre\sphinxhyphen{}releases using the following:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}\PYGZhy{}extra\PYGZhy{}index\PYGZhy{}url\PYG{+w}{ }https://test.pypi.org/simple/
\end{sphinxVerbatim}

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxAtStartPar
\sphinxstylestrong{These pre\sphinxhyphen{}release versions are being made available for initial
testing, they are not intended to be used for production
applications or analysis, and are not yet included in the main
pypi.org index}
\end{sphinxadmonition}

\sphinxAtStartPar
Once \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}} is installed, depending on your platform, you may also
need to {\hyperref[\detokenize{docs/guide-chapter-install:post-install-path-adjustments}]{\sphinxcrossref{adjust}}} your \sphinxcode{\sphinxupquote{PATH}}
environment variable.

\sphinxAtStartPar
\sphinxstylestrong{Upgrading} \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }\PYGZhy{}U\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}\PYGZhy{}extra\PYGZhy{}index\PYGZhy{}url\PYG{+w}{ }https://test.pypi.org/simple/
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxstylestrong{Uninstalling} \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }uninstall\PYG{+w}{ }pypop\PYGZhy{}genomics
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxstylestrong{For more, including handling common installation issues, see the} {\hyperref[\detokenize{docs/guide-chapter-install:detailed-installation-instructions}]{\sphinxcrossref{detailed installation instructions}}} \sphinxstylestrong{.}

\sphinxAtStartPar
Once you have installed \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}, you can move on to try some
{\hyperref[\detokenize{docs/guide-chapter-install:examples}]{\sphinxcrossref{example runs}}}.


\section{Examples}
\label{\detokenize{docs/guide-chapter-install:examples}}\label{\detokenize{docs/guide-chapter-install:guide-readme-examples}}
\sphinxAtStartPar
These are examples of how to check that the program is installed and
some minimal use cases.


\subsection{Checking version and installation}
\label{\detokenize{docs/guide-chapter-install:checking-version-and-installation}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop\PYG{+w}{ }\PYGZhy{}\PYGZhy{}version
\end{sphinxVerbatim}

\sphinxAtStartPar
This simply reports the version number and other information about
PyPop, and indirectly checks that the program is installed. If all is
well, you should see something like:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop 1.0.0a23
Copyright (C) 2003\PYGZhy{}2006 Regents of the University of California.
Copyright (C) 2007\PYGZhy{}2023 PyPop team.
This is free software.  There is NO warranty; not even for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
\end{sphinxVerbatim}

\sphinxAtStartPar
You can also run \sphinxcode{\sphinxupquote{pypop \sphinxhyphen{}\sphinxhyphen{}help}} to see a full list and explanation
of all the options available.


\subsection{Run a minimal dataset:}
\label{\detokenize{docs/guide-chapter-install:run-a-minimal-dataset}}
\sphinxAtStartPar
Download test \sphinxcode{\sphinxupquote{.ini}} and \sphinxcode{\sphinxupquote{.pop}} files: \sphinxhref{https://github.com/alexlancaster/pypop/blob/main/tests/data/minimal.ini}{minimal.ini} (https://github.com/alexlancaster/pypop/blob/main/tests/data/minimal.ini)
and \sphinxhref{https://github.com/alexlancaster/pypop/blob/main/tests/data/USAFEL-UchiTelle-small.pop}{USAFEL\sphinxhyphen{}UchiTelle\sphinxhyphen{}small.pop} (https://github.com/alexlancaster/pypop/blob/main/tests/data/USAFEL\sphinxhyphen{}UchiTelle\sphinxhyphen{}small.pop).
You can then run them

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop\PYG{+w}{ }\PYGZhy{}c\PYG{+w}{  }minimal.ini\PYG{+w}{ }USAFEL\PYGZhy{}UchiTelle\PYGZhy{}small.pop
\end{sphinxVerbatim}

\sphinxAtStartPar
If you have already cloned the git repository and it is your working
directory, you can simply run

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop\PYG{+w}{ }\PYGZhy{}c\PYG{+w}{  }tests/data/minimal.ini\PYG{+w}{ }tests/data/USAFEL\PYGZhy{}UchiTelle\PYGZhy{}small.pop
\end{sphinxVerbatim}

\sphinxAtStartPar
This will generate the following two files, an XML output file and a
plain text version:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{USAFEL}\PYG{o}{\PYGZhy{}}\PYG{n}{UchiTelle}\PYG{o}{\PYGZhy{}}\PYG{n}{small}\PYG{o}{\PYGZhy{}}\PYG{n}{out}\PYG{o}{.}\PYG{n}{xml}
\PYG{n}{USAFEL}\PYG{o}{\PYGZhy{}}\PYG{n}{UchiTelle}\PYG{o}{\PYGZhy{}}\PYG{n}{small}\PYG{o}{\PYGZhy{}}\PYG{n}{out}\PYG{o}{.}\PYG{n}{txt}
\end{sphinxVerbatim}


\section{Detailed installation instructions}
\label{\detokenize{docs/guide-chapter-install:detailed-installation-instructions}}
\sphinxAtStartPar
There are three main steps:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
install Python and \sphinxcode{\sphinxupquote{pip}}

\item {} 
\sphinxAtStartPar
install package from Test PyPI

\item {} 
\sphinxAtStartPar
adjusting your \sphinxcode{\sphinxupquote{PATH}} variable after installation

\end{enumerate}


\subsection{Install Python 3 and \sphinxstyleliteralintitle{\sphinxupquote{pip}}}
\label{\detokenize{docs/guide-chapter-install:install-python-3-and-pip}}
\sphinxAtStartPar
A full description of installing Python and \sphinxcode{\sphinxupquote{pip}} on your system is
beyond the scope of this guide, we recommend starting here:
\begin{quote}

\sphinxAtStartPar
\sphinxurl{https://wiki.python.org/moin/BeginnersGuide/Download}
\end{quote}

\sphinxAtStartPar
Here are some additional platform\sphinxhyphen{}specific notes that may be helpful:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Most Linux distributions come with Python 3 preinstalled. On most
modern systems, \sphinxcode{\sphinxupquote{pip}} and \sphinxcode{\sphinxupquote{python}} will default to Python 3.

\item {} 
\sphinxAtStartPar
MacOS 10.9 (Jaguar) up until 12.3 (Catalina), used to ship with
Python 2 pre\sphinxhyphen{}installed, but it now has to be manually installed.
See the \sphinxhref{https://docs.python.org/3/using/mac.html}{MacOS quick\sphinxhyphen{}start guide} (https://docs.python.org/3/using/mac.html) in the official
documentation for how to install Python 3. (Note that if Python is
installed on Mac via the MacOS developer tools, it may include the
version \sphinxcode{\sphinxupquote{3}} suffix on commands, e.g. \sphinxcode{\sphinxupquote{python3}} and \sphinxcode{\sphinxupquote{pip3}}, so
modify the below, accordingly).

\item {} 
\sphinxAtStartPar
For Windows, see also the \sphinxhref{https://docs.python.org/3/using/windows.html}{Windows quick\sphinxhyphen{}start guide} (https://docs.python.org/3/using/windows.html) in the official
documentation. Running \sphinxcode{\sphinxupquote{python}} in the Windows command terminal
in Windows 11 and later will launch the installer for the
Microsoft\sphinxhyphen{}maintained Windows package of Python 3.

\end{itemize}


\subsection{Install package from PyPI}
\label{\detokenize{docs/guide-chapter-install:install-package-from-pypi}}
\sphinxAtStartPar
Once you have both python and \sphinxcode{\sphinxupquote{pip}} installed, you can use \sphinxcode{\sphinxupquote{pip}}
to install pre\sphinxhyphen{}compiled binary “wheels” of \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}
pre\sphinxhyphen{}releases, test packages for PyPop available directly on the \sphinxhref{https://test.pypi.org/}{Test
PyPI} (https://test.pypi.org/).

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxAtStartPar
\sphinxstylestrong{These pre\sphinxhyphen{}release versions are being made available for initial
testing, they are not intended to be used for production
applications or analysis, and are not yet included in the main
pypi.org index}
\end{sphinxadmonition}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}\PYGZhy{}extra\PYGZhy{}index\PYGZhy{}url\PYG{+w}{ }https://test.pypi.org/simple/
\end{sphinxVerbatim}

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
If, for whatever reason, you cannot use the these binaries
(e.g. the pre\sphinxhyphen{}compiled binaries are not available for your
platform), you may need to follow the \sphinxhref{http://pypop.org/docs/guide-chapter-contributing.html\#installation-for-developers}{developer installation
instructions} (http://pypop.org/docs/guide\sphinxhyphen{}chapter\sphinxhyphen{}contributing.html\#installation\sphinxhyphen{}for\sphinxhyphen{}developers) in the contributors
guide.
\end{sphinxadmonition}

\sphinxAtStartPar
\sphinxstylestrong{Upgrade an existing PyPop installation}

\sphinxAtStartPar
To update an existing installation to a newer version, use the same
command as above, but add the \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}upgrade}} (short version: \sphinxcode{\sphinxupquote{\sphinxhyphen{}U}})
flag, i.e.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }\PYGZhy{}U\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}\PYGZhy{}extra\PYGZhy{}index\PYGZhy{}url\PYG{+w}{ }https://test.pypi.org/simple/
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxstylestrong{Issues with installation permission}

\sphinxAtStartPar
By default, \sphinxcode{\sphinxupquote{pip}} will attempt to install the \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}
package wherever the current Python installation is installed.  This
location may be a user\sphinxhyphen{}specific virtual environment (like \sphinxcode{\sphinxupquote{conda}},
see below), or a system\sphinxhyphen{}wide installation. On many Unix\sphinxhyphen{}based systems,
Python will generally already be pre\sphinxhyphen{}installed in a “system\sphinxhyphen{}wide”
location (e.g. under \sphinxcode{\sphinxupquote{/usr/lib}}) which is read\sphinxhyphen{}only for regular
users. (This can also be true for system\sphinxhyphen{}installed versions of Python
on Windows and MacOS.)

\sphinxAtStartPar
When \sphinxcode{\sphinxupquote{pip install}} cannot install in a read\sphinxhyphen{}only system\sphinxhyphen{}wide
location , \sphinxcode{\sphinxupquote{pip}} will gracefully “fall\sphinxhyphen{}back” to installing just for
you in your home directory (typically \sphinxcode{\sphinxupquote{\textasciitilde{}/.local/lib/python\textless{}VER\textgreater{}}}
where \sphinxcode{\sphinxupquote{\textless{}VER\textgreater{}}} is the version number of your current Python). In
general, this is what is wanted, so the above instructions are
normally sufficient.

\sphinxAtStartPar
However, you can also explicitly set installation to be in the user
directory, by adding the \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}user}} command\sphinxhyphen{}line option to the \sphinxcode{\sphinxupquote{pip
install}} command, i.e.:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}\PYGZhy{}user\PYG{+w}{ }\PYGZhy{}\PYGZhy{}extra\PYGZhy{}index\PYGZhy{}url\PYG{+w}{ }https://test.pypi.org/simple/
\end{sphinxVerbatim}

\sphinxAtStartPar
This may be necessary in certain cases where \sphinxcode{\sphinxupquote{pip install}} doesn’t
install into the expected user directory.

\begin{sphinxadmonition}{note}{Installing within a \sphinxstyleliteralintitle{\sphinxupquote{conda}} environment}

\sphinxAtStartPar
In the special case that you installing from within an activated
user\sphinxhyphen{}specific \sphinxcode{\sphinxupquote{conda}} virtual environment that provides Python,
then you should \sphinxstylestrong{not} add the \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}user}} because it will install
it in \sphinxcode{\sphinxupquote{\textasciitilde{}/.local/lib/}} rather than under the user\sphinxhyphen{}specific conda
virtual environment in \sphinxcode{\sphinxupquote{\textasciitilde{}/.conda/envs/}}.
\end{sphinxadmonition}


\subsection{Install package from GitHub Releases (advanced)}
\label{\detokenize{docs/guide-chapter-install:install-package-from-github-releases-advanced}}
\sphinxAtStartPar
We also sometimes make binary packages also available from the GitHub
release page:
\begin{quote}

\sphinxAtStartPar
\sphinxurl{https://github.com/alexlancaster/pypop/releases}
\end{quote}

\sphinxAtStartPar
To install these is similar to installing via PyPI above, except that
you need to explicitly provide a URL to the release page.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
First, visit the release page, and choose the release version you
wish to install (usually the most recent), and note the release tag
(e.g. \sphinxcode{\sphinxupquote{v1.0.0\sphinxhyphen{}a23}}).

\begin{sphinxadmonition}{note}{Release version numbers}

\sphinxAtStartPar
Note that version of the release is slightly different to the
\sphinxcode{\sphinxupquote{git}} tag.  This is because the \sphinxcode{\sphinxupquote{git}} tag follows \sphinxhref{https://semver.org/}{Semantic
Versioning} (https://semver.org/), which Python internally
normalizes and abbreviates.  So the release with the \sphinxcode{\sphinxupquote{git}} tag
\sphinxcode{\sphinxupquote{v1.0.0\sphinxhyphen{}a23}} is actually version \sphinxcode{\sphinxupquote{1.0.0a23}} of the
\sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}} package, and the version that \sphinxcode{\sphinxupquote{pip}} “sees”.
\end{sphinxadmonition}

\item {} 
\sphinxAtStartPar
Next, use \sphinxcode{\sphinxupquote{pip}} to install the package by running a command of
the form (this will select and install the correct wheel for your
Python version and operating system automatically):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}f\PYG{+w}{ }https://github.com/alexlancaster/pypop/releases/expanded\PYGZus{}assets/\PYGZlt{}TAG\PYGZus{}NAME\PYGZgt{}
\end{sphinxVerbatim}

\sphinxAtStartPar
where \sphinxstyleemphasis{\textless{}TAG\_NAME\textgreater{}} is replaced with a specific tag, e.g. for the example given above, you would run:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pypop\PYGZhy{}genomics\PYG{+w}{ }\PYGZhy{}f\PYG{+w}{ }https://github.com/alexlancaster/pypop/releases/expanded\PYGZus{}assets/v1.0.0\PYGZhy{}a23
\end{sphinxVerbatim}

\sphinxAtStartPar
You can also manually download the specific wheel from the github
release webpage and install directly, e.g.:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pypop\PYGZhy{}genomics\PYGZhy{}1.0.0a23\PYGZhy{}cp311\PYGZhy{}cp311\PYGZhy{}manylinux\PYGZus{}2\PYGZus{}17\PYGZus{}x86\PYGZus{}64.manylinux2014\PYGZus{}x86\PYGZus{}64.whl
\end{sphinxVerbatim}

\end{enumerate}


\subsection{Post\sphinxhyphen{}install \sphinxstyleliteralintitle{\sphinxupquote{PATH}} adjustments}
\label{\detokenize{docs/guide-chapter-install:post-install-path-adjustments}}
\sphinxAtStartPar
You may need to adjust the \sphinxcode{\sphinxupquote{PATH}} settings (especially on Windows)
for the \sphinxcode{\sphinxupquote{pypop}} scripts to be visible when run from your console
application, without having to supply the full path to the \sphinxcode{\sphinxupquote{pypop}}
executable file.

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxAtStartPar
Pay close attention to the “WARNINGS” that are shown during the
\sphinxcode{\sphinxupquote{pip}} installation, they will often note which directories need to
be added to the \sphinxcode{\sphinxupquote{PATH}}.
\end{sphinxadmonition}
\begin{itemize}
\item {} 
\sphinxAtStartPar
On Linux and MacOS, systems this is normally fairly simple and only
requires edit of the shell \sphinxcode{\sphinxupquote{.profile}}, or similar and addition of
the \sphinxcode{\sphinxupquote{\$HOME/.local/bin}} to the \sphinxcode{\sphinxupquote{PATH}} variable, followed by a
restart of the terminal.

\item {} 
\sphinxAtStartPar
For Windows, however, as noted in most online \sphinxhref{https://www.computerhope.com/issues/ch000549.htm}{instructions} (https://www.computerhope.com/issues/ch000549.htm), this may need
additional help from your system administrator if your user doesn’t
have the right permissions, and also require a system reboot.

\end{itemize}


\subsection{Uninstalling PyPop}
\label{\detokenize{docs/guide-chapter-install:uninstalling-pypop}}
\sphinxAtStartPar
To uninstall the current version of \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}genomics}}:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }uninstall\PYG{+w}{ }pypop\PYGZhy{}genomics
\end{sphinxVerbatim}


\section{Support and development}
\label{\detokenize{docs/guide-chapter-install:support-and-development}}
\sphinxAtStartPar
Please submit any bug reports, feature requests or questions, via our
GitHub issue tracker (see our \sphinxhref{http://pypop.org/docs/guide-chapter-contributing.html\#reporting-and-requesting}{bug reporting guidelines} (http://pypop.org/docs/guide\sphinxhyphen{}chapter\sphinxhyphen{}contributing.html\#reporting\sphinxhyphen{}and\sphinxhyphen{}requesting)
for more details on how to file a good bug report):
\begin{quote}

\sphinxAtStartPar
\sphinxurl{https://github.com/alexlancaster/pypop/issues}
\end{quote}

\sphinxAtStartPar
\sphinxstylestrong{Please do not report bugs via private email to developers.}

\sphinxAtStartPar
The development of the code for PyPop is via our GitHub project:
\begin{quote}

\sphinxAtStartPar
\sphinxurl{https://github.com/alexlancaster/pypop}
\end{quote}

\sphinxstepscope


\chapter{Getting started with PyPop}
\label{\detokenize{docs/guide-chapter-usage:getting-started-with-pypop}}\label{\detokenize{docs/guide-chapter-usage::doc}}

\section{Introduction}
\label{\detokenize{docs/guide-chapter-usage:introduction}}
\sphinxAtStartPar
You may use \sphinxstyleliteralstrong{\sphinxupquote{PyPop}} to analyze many different kinds of data, including
allele\sphinxhyphen{}level genotype data (as in \hyperref[\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}}}), allele\sphinxhyphen{}level
frequency data (as in \hyperref[\detokenize{docs/guide-chapter-usage:data-allelecount}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-allelecount}}}),
microsatellite data, SNP data, and nucleotide and amino acid sequence
data.

\sphinxAtStartPar
As mentioned in the installation chapter, a minimal working example
of a \sphinxhref{https://github.com/alexlancaster/pypop/blob/main/tests/data/USAFEL-UchiTelle-small.pop}{configuration file (.ini)} (https://github.com/alexlancaster/pypop/blob/main/tests/data/USAFEL\sphinxhyphen{}UchiTelle\sphinxhyphen{}small.pop),
and a \sphinxhref{https://github.com/alexlancaster/pypop/blob/main/tests/data/minimal.ini}{population file (.pop)} (https://github.com/alexlancaster/pypop/blob/main/tests/data/minimal.ini),
can be found by clicking the respective links.

\sphinxAtStartPar
There are two ways to run PyPop:
\begin{itemize}
\item {} 
\sphinxAtStartPar
interactive mode (where the program will prompt you to directly type
the input it needs); and

\item {} 
\sphinxAtStartPar
batch mode (where you supply all the command line options the program
needs).

\end{itemize}

\sphinxAtStartPar
For the most simplest application of PyPop, where you wish to analyze
a single population, the interactive mode is the simplest to use. We
will describe this mode first then describe batch mode.

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
The following assumes you have already {\hyperref[\detokenize{docs/guide-chapter-install:installing-pypop}]{\sphinxcrossref{\DUrole{std,std-ref}{installed PyPop}}}}, done any {\hyperref[\detokenize{docs/guide-chapter-install:post-install-path-adjustments}]{\sphinxcrossref{\DUrole{std,std-ref}{post\sphinxhyphen{}install adjustments}}}} needed for your platform, and
verified that you can run the main commands (see the
{\hyperref[\detokenize{docs/guide-chapter-install:examples}]{\sphinxcrossref{\DUrole{std,std-ref}{Examples}}}} section).
\end{sphinxadmonition}


\subsection{Interactive mode}
\label{\detokenize{docs/guide-chapter-usage:interactive-mode}}
\sphinxAtStartPar
To run PyPop in interactive mode, with a minimal “GUI”, on Windows or
MacOS, you can directly click on the \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}interactive}} file in the
directory where the scripts were installed (see {\hyperref[\detokenize{docs/guide-chapter-install:post-install-path-adjustments}]{\sphinxcrossref{\DUrole{std,std-ref}{post\sphinxhyphen{}install
adjustments}}}}).

\sphinxAtStartPar
You can also type \sphinxcode{\sphinxupquote{pypop\sphinxhyphen{}interactive}} after starting a console
application on all platforms (on MacOS and GNU/Linux, this is normally
the \sphinxstyleliteralstrong{\sphinxupquote{Terminal}} program, on Windows, it’s \sphinxstyleliteralstrong{\sphinxupquote{Command
prompt}}).

\sphinxAtStartPar
In most cases, this will launch a console with the following:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
PyPop: Python for Population Genomics (1.0.0a15)
Copyright (C) 2003\PYGZhy{}2006 Regents of the University of California
Copyright (C) 2007\PYGZhy{}2023 PyPop team.
This is free software.  There is NO warranty; not even for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You may redistribute copies of PyPop under the terms of the GNU
General Public License.  For more information about these
matters, see the file named COPYING.

Select both an \PYGZsq{}.ini\PYGZsq{} configuration file and a \PYGZsq{}.pop\PYGZsq{} file via the
system file dialog.
\end{sphinxVerbatim}

\sphinxAtStartPar
Following this:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
the system file dialog will appear prompting you to select an
\sphinxcode{\sphinxupquote{.ini}} {\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-configfile}]{\sphinxcrossref{\DUrole{std,std-ref}{configuration file}}}}.

\item {} 
\sphinxAtStartPar
a second system file dialog will prompt you for a \sphinxcode{\sphinxupquote{.pop}}
{\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-datafile}]{\sphinxcrossref{\DUrole{std,std-ref}{data file}}}}.

\item {} 
\sphinxAtStartPar
after both files are selected the console will display the
processing of the file:

\fvset{hllines={, 5,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
PyPop is processing sample.pop ...
PyPop run complete!
XML output(s) can be found in: [\PYGZsq{}sample\PYGZhy{}out.xml\PYGZsq{}]
Plain text output(s) can be found in: [\PYGZsq{}sample\PYGZhy{}out.txt\PYGZsq{}]
Press Enter to continue...
\end{sphinxVerbatim}
\sphinxresetverbatimhllines

\item {} 
\sphinxAtStartPar
when the run is completed, the last line will prompt you to press
\sphinxcode{\sphinxupquote{Enter}} to leave the console window (highlighted above).

\end{enumerate}

\sphinxAtStartPar
If the system file GUI dialog does not appear (e.g. if you are running
on a terminal without a display), it will fall\sphinxhyphen{}back to text\sphinxhyphen{}mode entry
for the files, where you need to type the full (either relative or
absolute) paths to the files. The output should resemble:

\fvset{hllines={, 14, 15,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
PyPop: Python for Population Genomics (1.0.0a15)
Copyright (C) 2003\PYGZhy{}2006 Regents of the University of California
Copyright (C) 2007\PYGZhy{}2023 PyPop team.
This is free software.  There is NO warranty; not even for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You may redistribute copies of PyPop under the terms of the GNU
General Public License.  For more information about these
matters, see the file named COPYING.

To accept the default in brackets for each filename, simply press
return for each prompt.

Please enter config filename [config.ini]: sample.ini
Please enter population filename [no default]: sample.pop
PyPop is processing sample.pop ...
PyPop run complete!
XML output(s) can be found in: [\PYGZsq{}sample\PYGZhy{}out.xml\PYGZsq{}]
Plain text output(s) can be found in: [\PYGZsq{}sample\PYGZhy{}out.txt\PYGZsq{}]
Press Enter to continue...
\end{sphinxVerbatim}
\sphinxresetverbatimhllines

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
Some messages with the prefix “LOG:” may appear during the console
operation.  They are informational only and do not indicate
improper operation of the program.
\end{sphinxadmonition}

\sphinxAtStartPar
In both cases you should substitute the names of your own
configuration (e.g., \sphinxcode{\sphinxupquote{config.ini}}) and population file (e.g.,
\sphinxcode{\sphinxupquote{Guatemalan.pop}}) for \sphinxcode{\sphinxupquote{sample.ini}} and \sphinxcode{\sphinxupquote{sample.pop}}
(highlighted above). The formats for these files are described in the
sections on the {\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-datafile}]{\sphinxcrossref{\DUrole{std,std-ref}{data file}}}} and
{\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-configfile}]{\sphinxcrossref{\DUrole{std,std-ref}{configuration file}}}}, below.


\subsection{Batch mode}
\label{\detokenize{docs/guide-chapter-usage:batch-mode}}
\sphinxAtStartPar
To run PyPop in the more common “batch mode”, you can run PyPop from
the console (as noted above, on Windows: open \sphinxstyleliteralstrong{\sphinxupquote{Command
prompt}}, aka a “DOS shell”; on MacOS or GNU/Linux: open the
\sphinxstyleliteralstrong{\sphinxupquote{Terminal}} application). Change to a directory where your
\sphinxcode{\sphinxupquote{.pop}} file is located, and type the command:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop Guatemalan.pop
\end{sphinxVerbatim}

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
If your system administrator has installed PyPop the name of the
script may be renamed to something different.
\end{sphinxadmonition}

\sphinxAtStartPar
Batch mode assumes two things: that you have a file called
\sphinxcode{\sphinxupquote{config.ini}} in your current folder and that you also have your
population file is in the current folder, otherwise you will need to
supply the full path to the file. You can specify a particular
configuration file for PyPop to use, by supplying the \sphinxcode{\sphinxupquote{\sphinxhyphen{}c}} option as
follows:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop \PYGZhy{}c newconfig.ini Guatemalan.pop
\end{sphinxVerbatim}

\sphinxAtStartPar
You may also redirect the output to a different directory (which must
already exist) by using the \sphinxcode{\sphinxupquote{\sphinxhyphen{}o}} option:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop \PYGZhy{}c newconfig.ini \PYGZhy{}o altdir Guatemalan.pop
\end{sphinxVerbatim}

\sphinxAtStartPar
Please see {\hyperref[\detokenize{docs/guide-chapter-usage:guide-pypop-cli}]{\sphinxcrossref{\DUrole{std,std-ref}{pypop usage}}}} for the full list of command\sphinxhyphen{}line
options.


\subsection{What happens when you run PyPop?}
\label{\detokenize{docs/guide-chapter-usage:what-happens-when-you-run-pypop}}\label{\detokenize{docs/guide-chapter-usage:guide-usage-intro-run-details}}
\sphinxAtStartPar
The most common types of analysis will involve the editing of your
\sphinxcode{\sphinxupquote{config.ini}} file to suit your data (see \sphinxhref{guide-usage-configfile}{The configuration
file}) followed by the selection of either
the interactive or batch mode described above. If your input
configuration file is \sphinxcode{\sphinxupquote{\sphinxstyleemphasis{configfilename}}} and your population file name
is \sphinxcode{\sphinxupquote{\sphinxstyleemphasis{popfilename}.txt}} the initial output will be generated quickly, but
your the PyPop execution will not be finished until the text output file
named \sphinxcode{\sphinxupquote{\sphinxstyleemphasis{popfilename}\sphinxhyphen{}out.txt}} has been created. A successful run will
produce two output files: \sphinxcode{\sphinxupquote{\sphinxstyleemphasis{popfilename}\sphinxhyphen{}out.xml}},
\sphinxcode{\sphinxupquote{\sphinxstyleemphasis{popfilename}\sphinxhyphen{}out.txt}}. A third output file will be created if you are
using the Anthony Nolan HLA filter option for HLA data to check your
input for valid/known HLA alleles: \sphinxcode{\sphinxupquote{popfilename\sphinxhyphen{}filter.xml}}).

\sphinxAtStartPar
The \sphinxcode{\sphinxupquote{popfilename\sphinxhyphen{}out.xml}} file is the primary output created by
PyPop and the human\sphinxhyphen{}readable \sphinxcode{\sphinxupquote{popfilename\sphinxhyphen{}out.txt}} file is a
summary of the complete XML output. The XML output can be further
transformed into plain text TSV files, either directly via \sphinxcode{\sphinxupquote{pypop}}
if invoked on multiple input files (using the \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}tsv}} option,
see {\hyperref[\detokenize{docs/guide-chapter-usage:guide-pypop-cli}]{\sphinxcrossref{\DUrole{std,std-ref}{pypop usage}}}}), or via the \sphinxcode{\sphinxupquote{popmeta}} tool that
aggregates results from different \sphinxcode{\sphinxupquote{pypop}} runs (see
{\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-popmeta}]{\sphinxcrossref{\DUrole{std,std-ref}{Using popmeta to aggregate results}}}}).

\sphinxAtStartPar
A typical PyPop run might take anywhere from a few of minutes to a few
hours, depending on how large your data set is and who else is using the
system at the same time. Note that performing the
\sphinxcode{\sphinxupquote{allPairwiseLDWithPermu}} test may take several \sphinxstylestrong{days} if you have
highly polymorphic loci in your data set.


\section{Using \sphinxstyleliteralintitle{\sphinxupquote{popmeta}} to aggregate results}
\label{\detokenize{docs/guide-chapter-usage:using-popmeta-to-aggregate-results}}\label{\detokenize{docs/guide-chapter-usage:guide-usage-popmeta}}
\sphinxAtStartPar
The \sphinxcode{\sphinxupquote{popmeta}} script (\sphinxcode{\sphinxupquote{popmeta.bat}} on Windows, \sphinxcode{\sphinxupquote{popmeta}} on
GNU/Linux) can aggregate results from a number of output XML files
from individual populations into a set of tab\sphinxhyphen{}separated (TSV) files
containing summary statistics via customized XSLT (eXtensible
Stylesheet Language for Transformations) stylesheets.  These TSV files
can be directly imported into a spreadsheet or statistical software
(e.g., \sphinxstyleliteralstrong{\sphinxupquote{R}}, \sphinxstyleliteralstrong{\sphinxupquote{SAS}}).  In addition, there is some
preliminary support for export into other formats, such as the
population genetic software (e.g., \sphinxstyleliteralstrong{\sphinxupquote{PHYLIP}}).

\sphinxAtStartPar
Here is an example of a \sphinxcode{\sphinxupquote{popmeta}} run, following on from the XML outputs
generated in similar fashion in the previous \sphinxcode{\sphinxupquote{pypop}} runs:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
popmeta \PYGZhy{}o altdir Guatemalan\PYGZhy{}out.xml NorthAmerican\PYGZhy{}out.xml
\end{sphinxVerbatim}

\sphinxAtStartPar
This will generate a number of \sphinxcode{\sphinxupquote{.dat}} files, including
\sphinxcode{\sphinxupquote{1\sphinxhyphen{}locus\sphinxhyphen{}allele.dat}}.

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
It’s highly recommended to use the \sphinxcode{\sphinxupquote{\sphinxhyphen{}o}} option to save the output
in a separate subdirectory, as the output \sphinxcode{\sphinxupquote{.dat}} files have
fixed names, and will overwrite any files in the local directory with the
same name).  See {\hyperref[\detokenize{docs/guide-chapter-usage:guide-popmeta-cli}]{\sphinxcrossref{\DUrole{std,std-ref}{popmeta usage}}}} for the full list of
options.
\end{sphinxadmonition}

\sphinxAtStartPar
Note that a similar effect can be achieved directly from a \sphinxcode{\sphinxupquote{pypop}}
run (assuming that the configuration file can be used for both
\sphinxcode{\sphinxupquote{.pop}} population files), by invoking \sphinxcode{\sphinxupquote{pypop}} with the
\sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}tsv}} option:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pypop \PYGZhy{}c newconfig.ini \PYGZhy{}o altdir Guatemalan.pop NorthAmerican.pop \PYGZhy{}\PYGZhy{}enable\PYGZhy{}tsv
\end{sphinxVerbatim}


\section{Command\sphinxhyphen{}line interfaces}
\label{\detokenize{docs/guide-chapter-usage:command-line-interfaces}}
\sphinxAtStartPar
Described below is the usage for both programs, including a full list
of the current command\sphinxhyphen{}line options and arguments.  Note that you can
also view this full list of options from the program itself by
supplying the \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}help}} option, i.e. \sphinxcode{\sphinxupquote{pypop \sphinxhyphen{}\sphinxhyphen{}help}}, or \sphinxcode{\sphinxupquote{popmeta
\sphinxhyphen{}\sphinxhyphen{}help}}, respectively.


\subsection{\sphinxstyleliteralintitle{\sphinxupquote{pypop}} usage}
\label{\detokenize{docs/guide-chapter-usage:pypop-usage}}\label{\detokenize{docs/guide-chapter-usage:guide-pypop-cli}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{pypop} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{o} \PYG{n}{OUTPUTDIR}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{V}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{c} \PYG{n}{CONFIG}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{m}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{d}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{x} \PYG{n}{XSLFILE}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{t}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{enable}\PYG{o}{\PYGZhy{}}\PYG{n}{ihwg}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{enable}\PYG{o}{\PYGZhy{}}\PYG{n}{phylip}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{i}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{f} \PYG{n}{FILELIST}\PYG{p}{]}
             \PYG{p}{[}\PYG{n}{POPFILE} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{p}{]}
\end{sphinxVerbatim}


\subsubsection{Options for pypop}
\label{\detokenize{docs/guide-chapter-usage:options-for-pypop}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}o, \sphinxhyphen{}\sphinxhyphen{}outputdir]  
\sphinxAtStartPar
put output in directory OUTPUTDIR
\item [\sphinxhyphen{}V, \sphinxhyphen{}\sphinxhyphen{}version]  
\sphinxAtStartPar
show program’s version number and exit
\item [\sphinxhyphen{}c, \sphinxhyphen{}\sphinxhyphen{}config]  
\sphinxAtStartPar
select config file

\sphinxAtStartPar
Default: “config.ini”
\item [\sphinxhyphen{}m, \sphinxhyphen{}\sphinxhyphen{}testmode]  
\sphinxAtStartPar
run PyPop in test mode for unit testing
\item [\sphinxhyphen{}d, \sphinxhyphen{}\sphinxhyphen{}debug]  
\sphinxAtStartPar
enable debugging output (overrides config file setting)
\item [\sphinxhyphen{}x, \sphinxhyphen{}\sphinxhyphen{}xsl]  
\sphinxAtStartPar
override the default XSLT translation with XSLFILE
\end{optionlist}


\subsubsection{TSV output options}
\label{\detokenize{docs/guide-chapter-usage:tsv-output-options}}
\sphinxAtStartPar
Note that \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}}} flags only valid if \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}tsv}}/\sphinxcode{\sphinxupquote{\sphinxhyphen{}t}} selected
\begin{optionlist}{3cm}
\item [\sphinxhyphen{}t, \sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}tsv]  
\sphinxAtStartPar
generate TSV output files (aka run ‘popmeta’)
\item [\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}ihwg]  
\sphinxAtStartPar
enable 13th IWHG workshop populationdata default headers
\item [\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}phylip]  
\sphinxAtStartPar
enable generation of PHYLIP \sphinxcode{\sphinxupquote{.phy}} files
\end{optionlist}


\subsubsection{Mutually exclusive input options}
\label{\detokenize{docs/guide-chapter-usage:mutually-exclusive-input-options}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}i, \sphinxhyphen{}\sphinxhyphen{}interactive]  
\sphinxAtStartPar
run in interactive mode, prompting user for file names
\item [\sphinxhyphen{}f, \sphinxhyphen{}\sphinxhyphen{}filelist]  
\sphinxAtStartPar
file containing list of files (one per line) to process
(mutually exclusive with supplying POPFILEs)
\item [POPFILE]  
\sphinxAtStartPar
input population (\sphinxcode{\sphinxupquote{.pop}}) file(s)

\sphinxAtStartPar
Default: {[}{]}
\end{optionlist}


\subsection{\sphinxstyleliteralintitle{\sphinxupquote{popmeta}} usage}
\label{\detokenize{docs/guide-chapter-usage:popmeta-usage}}\label{\detokenize{docs/guide-chapter-usage:guide-popmeta-cli}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{usage}\PYG{p}{:} \PYG{n}{popmeta} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{h}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{o} \PYG{n}{OUTPUTDIR}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{V}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{disable}\PYG{o}{\PYGZhy{}}\PYG{n}{tsv}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{output}\PYG{o}{\PYGZhy{}}\PYG{n}{meta}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{n}{x} \PYG{n}{XSLDIR}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{enable}\PYG{o}{\PYGZhy{}}\PYG{n}{ihwg}\PYG{p}{]} \PYG{p}{[}\PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZhy{}}\PYG{n}{enable}\PYG{o}{\PYGZhy{}}\PYG{n}{phylip} \PYG{o}{|} \PYG{o}{\PYGZhy{}}\PYG{n}{b} \PYG{n}{FACTOR}\PYG{p}{]}
               \PYG{n}{XMLFILE} \PYG{p}{[}\PYG{n}{XMLFILE} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{p}{]}
\end{sphinxVerbatim}


\subsubsection{Positional Arguments}
\label{\detokenize{docs/guide-chapter-usage:positional-arguments}}\begin{optionlist}{3cm}
\item [XMLFILE]  
\sphinxAtStartPar
XML (\sphinxcode{\sphinxupquote{.xml}}) file(s) generated by pypop runs

\sphinxAtStartPar
Default: {[}{]}
\end{optionlist}


\subsubsection{Options for popmeta}
\label{\detokenize{docs/guide-chapter-usage:options-for-popmeta}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}o, \sphinxhyphen{}\sphinxhyphen{}outputdir]  
\sphinxAtStartPar
put output in directory OUTPUTDIR
\item [\sphinxhyphen{}V, \sphinxhyphen{}\sphinxhyphen{}version]  
\sphinxAtStartPar
show program’s version number and exit
\item [\sphinxhyphen{}\sphinxhyphen{}disable\sphinxhyphen{}tsv]  
\sphinxAtStartPar
disable generation of \sphinxcode{\sphinxupquote{.dat}} TSV files
\item [\sphinxhyphen{}\sphinxhyphen{}output\sphinxhyphen{}meta]  
\sphinxAtStartPar
dump the meta output file to stdout, ignore xslt file
\item [\sphinxhyphen{}x, \sphinxhyphen{}\sphinxhyphen{}xsldir]  
\sphinxAtStartPar
use specified directory to find meta XSLT
\item [\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}ihwg]  
\sphinxAtStartPar
enable 13th IWHG workshop populationdata default headers
\end{optionlist}


\subsubsection{Mutually exclusive popmeta options}
\label{\detokenize{docs/guide-chapter-usage:mutually-exclusive-popmeta-options}}\begin{optionlist}{3cm}
\item [\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}phylip]  
\sphinxAtStartPar
enable generation of PHYLIP \sphinxcode{\sphinxupquote{.phy}} files
\item [\sphinxhyphen{}b, \sphinxhyphen{}\sphinxhyphen{}batchsize]  
\sphinxAtStartPar
process in batches of size total/FACTOR rather than all at once, by default do separately (batchsize=0)

\sphinxAtStartPar
Default: 0
\end{optionlist}


\section{The data file}
\label{\detokenize{docs/guide-chapter-usage:the-data-file}}\label{\detokenize{docs/guide-chapter-usage:guide-usage-datafile}}

\subsection{Sample files}
\label{\detokenize{docs/guide-chapter-usage:sample-files}}
\sphinxAtStartPar
Data can be input either as genotypes, or in an allele count format,
depending on the format of your data.

\sphinxAtStartPar
As you will see in the following examples, population files begin with
header information. In the simplest case, the first line contains the
column headers for the genotype, allele count, or, sequence information
from the population. If the file contains a population data\sphinxhyphen{}block, then
the first line consists of headers identifying the data on the second
line, and the third line contains the column headers for the genotype or
allele count information.

\sphinxAtStartPar
Note that for genotype data, each locus corresponds to two columns in
the population file. The locus name must repeated, with a suffix such as
\sphinxcode{\sphinxupquote{\_1}}, \sphinxcode{\sphinxupquote{\_2}} (the default) or \sphinxcode{\sphinxupquote{\_a}}, \sphinxcode{\sphinxupquote{\_b}} and must match the format
defined in the \sphinxcode{\sphinxupquote{config.ini}} (see
{\hyperref[\detokenize{docs/guide-chapter-usage:validsamplefields}]{\sphinxcrossref{\DUrole{std,std-ref}{validSampleFields}}}}). Although PyPop needs this
distinction to be made, phase is NOT assumed, and if known it is
ignored.

\sphinxAtStartPar
\hyperref[\detokenize{docs/guide-chapter-usage:config-minimal-example}]{Listing \ref{\detokenize{docs/guide-chapter-usage:config-minimal-example}}} shows the relevant lines for the
configuration to read in the data shown in
\hyperref[\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}}} through to \hyperref[\detokenize{docs/guide-chapter-usage:data-allelecount}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-allelecount}}}.
\sphinxSetupCaptionForVerbatim{Multi\sphinxhyphen{}locus allele\sphinxhyphen{}level genotype data}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
a\PYGZus{}1   a\PYGZus{}2   c\PYGZus{}1   c\PYGZus{}2   b\PYGZus{}1   b\PYGZus{}2
****  ****  0102  02025 1301  18012
0101  0201  0307  0605  1401  39021
0210  03012 0712  0102  1520  1301
0101  0218  0804  1202  35091 4005
2501  0201  1507  0307  51013 1401
0210  3204  1801  0102  78021 1301
03012 3204  1507  0605  51013 39021
\end{sphinxVerbatim}

\sphinxAtStartPar
This is an example of the simplest kind of data file.
\sphinxSetupCaptionForVerbatim{Multi\sphinxhyphen{}locus allele\sphinxhyphen{}level HLA genotype data with sample information}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:data-minimal-noheader}}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
populat    id        a\PYGZus{}1   a\PYGZus{}2   c\PYGZus{}1   c\PYGZus{}2   b\PYGZus{}1   b\PYGZus{}2
UchiTelle  UT900\PYGZhy{}23  ****  ****  0102  02025 1301  18012
UchiTelle  UT900\PYGZhy{}24  0101  0201  0307  0605  1401  39021
UchiTelle  UT900\PYGZhy{}25  0210  03012 0712  0102  1520  1301
UchiTelle  UT900\PYGZhy{}26  0101  0218  0804  1202  35091 4005
UchiTelle  UT910\PYGZhy{}01  2501  0201  1507  0307  51013 1401
UchiTelle  UT910\PYGZhy{}02  0210  3204  1801  0102  78021 1301
UchiTelle  UT910\PYGZhy{}03  03012 3204  1507  0605  51013 39021
\end{sphinxVerbatim}

\sphinxAtStartPar
This example shows a data file which has non\sphinxhyphen{}allele data in some
columns, here we have population (\sphinxcode{\sphinxupquote{populat}}) and sample identifiers
(\sphinxcode{\sphinxupquote{id}}).
\sphinxSetupCaptionForVerbatim{Multi\sphinxhyphen{}locus allele\sphinxhyphen{}level HLA genotype data with sample and header information}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:data-hla}}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
labcode method              ethnic  contin  collect        latit           longit
USAFEL  12th Workshop SSOP  Telle   NW Asia Targen Village 41 deg 12 min N 94 deg 7 min E
populat     id         a\PYGZus{}1     a\PYGZus{}2     c\PYGZus{}1     c\PYGZus{}2     b\PYGZus{}1     b\PYGZus{}2
UchiTelle   UT900\PYGZhy{}23   ****    ****    0102    02025   1301    18012
UchiTelle   UT900\PYGZhy{}24   0101    0201    0307    0605    1401    39021
UchiTelle   UT900\PYGZhy{}25   0210    03012   0712    0102    1520    1301
UchiTelle   UT900\PYGZhy{}26   0101    0218    0804    1202    35091   4005
UchiTelle   UT910\PYGZhy{}01   2501    0201    1507    0307    51013   1401
UchiTelle   UT910\PYGZhy{}02   0210    3204    1801    0102    78021   1301
UchiTelle   UT910\PYGZhy{}03   03012   3204    1507    0605    51013   39021
\end{sphinxVerbatim}

\sphinxAtStartPar
This is an example of a data file which is identical to
\hyperref[\detokenize{docs/guide-chapter-usage:data-minimal-noheader}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-minimal-noheader}}}, but which includes population level
information.
\sphinxSetupCaptionForVerbatim{Multi\sphinxhyphen{}locus allele\sphinxhyphen{}level HLA genotype and microsatellite genotype data with header information}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:data-hla-microsat}}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
labcode ethnic  complex
USAFEL  ****    0
populat    id      drb1\PYGZus{}1  drb1\PYGZus{}2  dqb1\PYGZus{}1  dqb1\PYGZus{}2  d6s2222\PYGZus{}1  d6s2222\PYGZus{}2
UchiTelle  HJK\PYGZus{}2   01      0301    0201     0501    249        249
UchiTelle  HJK\PYGZus{}1   0301    0301    0201     0201    249        249
UchiTelle  HJK\PYGZus{}3   01      0301    0201     0501    249        249
UchiTelle  HJK\PYGZus{}4   01      0301    0201     0501    249        249
UchiTelle  MYU\PYGZus{}2   02      0401    0302     0602    247        249
UchiTelle  MYU\PYGZus{}1   0301    0301    0201     0201    247        249
UchiTelle  MYU\PYGZus{}3   0301    0401    0201     0302    249        249
UchiTelle  MYU\PYGZus{}4   0301    0401    0201     0302    247        249
\end{sphinxVerbatim}

\sphinxAtStartPar
This example mixes different kinds of data: HLA allele data (from DRB1
and DQB1 loci) with microsatellite data (locus D6S2222).
\sphinxSetupCaptionForVerbatim{Sequence genotype data with header information}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:data-nucleotide}}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
labcode file
BLOGGS  C\PYGZus{}New
popName ID       TGFB1cdn10(1) TGFB1cdn10(2) TGFBhapl(1) TGFBhapl(2)
Urboro  XQ\PYGZhy{}1     C             T             CG          TG
Urboro  XQ\PYGZhy{}2     C             C             CG          CG
Urboro  XQ\PYGZhy{}5     C             T             CG          TG
Urboro  XQ\PYGZhy{}21    C             T             CG          TG
Urboro  XQ\PYGZhy{}7     C             T             CG          TG
Urboro  XQ\PYGZhy{}20    C             T             CG          TG
Urboro  XQ\PYGZhy{}6     T             T             TG          TG
Urboro  XQ\PYGZhy{}8     C             T             CG          TG
Urboro  XQ\PYGZhy{}9     T             T             TG          TG
Urboro  XQ\PYGZhy{}10    C             T             CG          TG
\end{sphinxVerbatim}

\sphinxAtStartPar
This example includes nucleotide sequence data: the TGFB1CDN10 locus
consists of one nucleotide, the TGFBhapl locus is actually haplotype
data, but PyPop simply treats each combination as a separate “allele”
for subsequent analysis.
\sphinxSetupCaptionForVerbatim{Allele count data}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:data-allelecount}}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
populat    method  ethnic     country    latit   longit
UchiTelle  PCR\PYGZhy{}SSO Klingon    QZ         052.81N 100.25E
dqa1  count
0101  31
0102  37
0103  17
0201  21
0301  32
0401  9
0501  35
\end{sphinxVerbatim}

\sphinxAtStartPar
PyPop can also process allele count data. However, you cannot mix allele
count data and genotype data together in the one file.

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
Currently each \sphinxcode{\sphinxupquote{.pop}} file can only contain allele count data for
\sphinxstyleemphasis{one locus}. In order to process multiple loci for one population you
must create a separate \sphinxcode{\sphinxupquote{.pop}} for each locus.
\end{sphinxadmonition}

\sphinxAtStartPar
These population files are plain text files, such as you might save
out of the \sphinxstyleliteralstrong{\sphinxupquote{Notepad}} application on Windows (or
\sphinxstyleliteralstrong{\sphinxupquote{Emacs}}). The columns are all tab\sphinxhyphen{}delimited, so you can
include spaces in your labels. If you have your data in a spreadsheet
application, such as \sphinxstyleliteralstrong{\sphinxupquote{Excel}} or \sphinxstyleliteralstrong{\sphinxupquote{LibreOffice}}, export the file as
tab\sphinxhyphen{}delimited text, in order to use it as PyPop data file.


\subsection{Missing data}
\label{\detokenize{docs/guide-chapter-usage:missing-data}}
\sphinxAtStartPar
Untyped or missing data may be represented in a variety of ways. The
default value for untyped or missing data is a series of four asterisks
(\sphinxcode{\sphinxupquote{****}}) as specified by the \sphinxcode{\sphinxupquote{config.ini}}. You may not “represent”
untyped data by leaving a column blank, nor may you represent a
homozygote by leaving the second column blank. All cells for which you
have data must include data, and all cells for which you do not have
data must also be filled in, using a missing data value.

\sphinxAtStartPar
For individuals who were not typed at all loci, the data in loci for
which they are typed will be used on all single\sphinxhyphen{}locus analyses for that
individual and locus, so that you see the value of the number of
individuals (\sphinxcode{\sphinxupquote{n}}) vary from locus to locus in the output. These
individuals’ data will also be used for multi\sphinxhyphen{}locus analyses. Only the
loci that contain no missing data will be included in any multi\sphinxhyphen{}locus
analysis.

\sphinxAtStartPar
If an individual is only partially typed at a locus, it will be treated
as if it were completely untyped, and data for that individual for that
locus will be dropped from ALL analyses.

\begin{sphinxadmonition}{warning}{Warning:}\begin{itemize}
\item {} 
\sphinxAtStartPar
Do not leave trailing blank lines at the end of your data file, as
this currently causes PyPop to terminate with an error message
that takes experience to diagnose.

\item {} 
\sphinxAtStartPar
For haplotype estimation and linkage disequilibrium calculations
(i.e., the emhaplofreq part of the program) you are currently
restricted to a maximum of seven loci per haplotype request. For
haplotype estimation there is a limit of 5000 for the number of
individuals (\sphinxcode{\sphinxupquote{n}}) %
\begin{footnote}[1]\sphinxAtStartFootnote
These hardcoded numbers can be changed if you obtain the source code
yourself and change the appropriate \#define \sphinxcode{\sphinxupquote{emhaplofreq.h}} and
recompile the program.
%
\end{footnote}

\end{itemize}
\end{sphinxadmonition}


\section{The configuration file}
\label{\detokenize{docs/guide-chapter-usage:the-configuration-file}}\label{\detokenize{docs/guide-chapter-usage:guide-usage-configfile}}
\sphinxAtStartPar
The sets of population genetic analyses that are run on your population
data file and the manner in which the data file is interpreted by PyPop
is controlled by a configuration file, the default name for which is
\sphinxcode{\sphinxupquote{config.ini}}. This is another plain text file consisting of comments
(which are lines that start with a semi\sphinxhyphen{}colon), sections (which are
lines with labels in square brackets), and options (which are lines
specifying settings relevant to that section in the \sphinxcode{\sphinxupquote{option=value}}
format).

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
If any option runs over one line (such as \sphinxcode{\sphinxupquote{validSampleFields}}) then
the second and subsequent lines must be indented by exactly \sphinxstylestrong{one
space}.
\end{sphinxadmonition}


\subsection{A minimal configuration file}
\label{\detokenize{docs/guide-chapter-usage:a-minimal-configuration-file}}\label{\detokenize{docs/guide-chapter-usage:config-minimal}}
\sphinxAtStartPar
Here we present a minimal \sphinxcode{\sphinxupquote{.ini}} file corresponding to
\hyperref[\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-minimal-noheader-noids}}} A section by section
review of this file follows. (Note comment lines have been omitted in
the above example for clarity). A description of more advanced options
is contained in {\hyperref[\detokenize{docs/guide-chapter-usage:config-advanced}]{\sphinxcrossref{\DUrole{std,std-ref}{Advanced options}}}}.
\sphinxSetupCaptionForVerbatim{Minimal config.ini file}
\def\sphinxLiteralBlockLabel{\label{\detokenize{docs/guide-chapter-usage:config-minimal-example}}}
\fvset{hllines={, 1, 4, 14, 17, 22, 25,}}%
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{k}{[General]}
\PYG{n+na}{debug}\PYG{o}{=}\PYG{l+s}{0}

\PYG{k}{[ParseGenotypeFile]}
\PYG{n+na}{untypedAllele}\PYG{o}{=}\PYG{l+s}{****}
\PYG{n+na}{alleleDesignator}\PYG{o}{=}\PYG{l+s}{*}
\PYG{n+na}{validSampleFields}\PYG{o}{=}\PYG{l+s}{*a\PYGZus{}1}
\PYG{+w}{ }\PYG{n+na}{*a\PYGZus{}2}
\PYG{+w}{ }\PYG{n+na}{*c\PYGZus{}1}
\PYG{+w}{ }\PYG{n+na}{*c\PYGZus{}2}
\PYG{+w}{ }\PYG{n+na}{*b\PYGZus{}1}
\PYG{+w}{ }\PYG{n+na}{*b\PYGZus{}2}

\PYG{k}{[HardyWeinberg]}
\PYG{n+na}{lumpBelow}\PYG{o}{=}\PYG{l+s}{5}

\PYG{k}{[HardyWeinbergGuoThompson]}
\PYG{n+na}{dememorizationSteps}\PYG{o}{=}\PYG{l+s}{2000}
\PYG{n+na}{samplingNum}\PYG{o}{=}\PYG{l+s}{1000}
\PYG{n+na}{samplingSize}\PYG{o}{=}\PYG{l+s}{1000}

\PYG{k}{[HomozygosityEWSlatkinExact]}
\PYG{n+na}{numReplicates}\PYG{o}{=}\PYG{l+s}{10000}

\PYG{k}{[Emhaplofreq]}
\PYG{n+na}{allPairwiseLD}\PYG{o}{=}\PYG{l+s}{1}
\PYG{n+na}{allPairwiseLDWithPermu}\PYG{o}{=}\PYG{l+s}{0}
\PYG{c+c1}{;;numPermuInitCond=5}
\end{sphinxVerbatim}
\sphinxresetverbatimhllines

\sphinxAtStartPar
\sphinxstylestrong{Configuration file sections} (highlighted above)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}General{]}}}

\sphinxAtStartPar
This section contains variables that control the overall behavior of
PyPop.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{debug=0}}.

\sphinxAtStartPar
This setting is for debugging. Setting it to 1 will set off a
large amount of output of no interest to the general user. It
should not be used unless you are running into trouble and need to
communicate with the PyPop developers about the problems.

\end{itemize}

\item {} 
\sphinxAtStartPar
Specifying data formats

\sphinxAtStartPar
There are two possible formats: \sphinxcode{\sphinxupquote{{[}ParseGenotypeFile{]}}} and
\sphinxcode{\sphinxupquote{{[}ParseAlleleCountFile{]}}}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}ParseGenotypeFile{]}}}.

\sphinxAtStartPar
If your data is genotype data, you will want a section labeled:
\sphinxcode{\sphinxupquote{{[}ParseGenotypeFile{]}}}.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{alleleDesignator}}.

\sphinxAtStartPar
This option is used to tell PyPop what is allele data and what
isn’t. You must use this symbol in :ref:\sphinxcode{\sphinxupquote{\textasciigrave{}validSampleFields}}
option. The default is \sphinxcode{\sphinxupquote{*}}. In general, you won’t need to
change it. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{*}} \sphinxstylestrong{{]}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{untypedAllele}}.

\sphinxAtStartPar
This option is used to tell PyPop what symbol you have used in
your data files to represent untyped or unknown data
fields. These fields MAY NOT BE LEFT BLANK. You must use
something consistent that cannot be confused with real data
here. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{****}} \sphinxstylestrong{{]}}

\end{itemize}

\end{itemize}
\phantomsection\label{\detokenize{docs/guide-chapter-usage:validsamplefields}}\begin{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{validSampleFields}}.

\sphinxAtStartPar
This option should contain the names of the loci immediately
preceding your genotype data (if it has three header lines, this
information will be on the third line, otherwise it will be the
first line of the file).\sphinxstylestrong{{[}There is no default, this option must
always be present{]}}

\sphinxAtStartPar
The format is as follows, for each sample field (which may either
be an identifying field for the sample such as \sphinxcode{\sphinxupquote{populat}}, or
contain allele data) create a new line where:
\begin{itemize}
\item {} 
\sphinxAtStartPar
The first line (\sphinxcode{\sphinxupquote{validSampleFields=}}) consists of the name of
your sample field (if it contains allele data, the name of the
field should be preceded by the character designated in the
\sphinxcode{\sphinxupquote{alleleDesignator}} option above).

\item {} 
\sphinxAtStartPar
All subsequent lines after the first \sphinxstyleemphasis{must} be preceded by \sphinxstyleemphasis{one
space} (again if it contains allele data, the name of the field
should be preceded by the character designated in the
\sphinxcode{\sphinxupquote{alleleDesignator}} option above).

\end{itemize}

\sphinxAtStartPar
Here is an example:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
validSampleFields=*a\PYGZus{}1
 *a\PYGZus{}2
 *c\PYGZus{}1
 *c\PYGZus{}2
 *b\PYGZus{}1
 *b\PYGZus{}2    Note initial space at start of line.
\end{sphinxVerbatim}

\sphinxAtStartPar
Here is example that includes identifying (non\sphinxhyphen{}allele data)
information such as sample id (\sphinxcode{\sphinxupquote{id}}) and population name
(\sphinxcode{\sphinxupquote{populat}}):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
validSampleFields=populat
 id
 *a\PYGZus{}1
 *a\PYGZus{}2
 *c\PYGZus{}1
 *c\PYGZus{}2
 *b\PYGZus{}1
 *b\PYGZus{}2
\end{sphinxVerbatim}

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}ParseAlleleCountFile{]}}}.

\sphinxAtStartPar
If your data is not genotype data, but rather, data of the
allele\sphinxhyphen{}name count format, then you will want to use the
\sphinxcode{\sphinxupquote{{[}ParseAlleleCountFile{]}}} section INSTEAD of the
\sphinxcode{\sphinxupquote{{[}ParseGenotypeFile{]}}} section. The \sphinxcode{\sphinxupquote{alleleDesignator}} and
\sphinxcode{\sphinxupquote{untypedAllele}} options work identically to that described for
\sphinxcode{\sphinxupquote{{[}ParseGenotypeFile{]}}}.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{validSampleFields}}.

\sphinxAtStartPar
This option should contain either a single locus name or a
colon\sphinxhyphen{}separated list of all loci that will be in the data files
you intend to analyze using a specific \sphinxcode{\sphinxupquote{.ini}} file. The
colon\sphinxhyphen{}separated list allows you to avoid changing the \sphinxcode{\sphinxupquote{.ini}}
file when running over a collection of data files containing
different loci. e.g.,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
validSampleFields=A:B:C:DQA1:DQB1:DRB1:DPB1:DPA1
 count
\end{sphinxVerbatim}

\sphinxAtStartPar
Note that each \sphinxcode{\sphinxupquote{.pop}} file must contain only one locus (see
\sphinxhref{data-allelecount-note}{note\_title} in
\hyperref[\detokenize{docs/guide-chapter-usage:data-allelecount}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-allelecount}}}). Listing multiple loci
simply permits the same \sphinxcode{\sphinxupquote{.ini}} file to be reused for each data
file.

\end{itemize}
\end{quote}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}HardyWeinberg{]}}}

\sphinxAtStartPar
Hardy\sphinxhyphen{}Weinberg analysis is enabled by the presence of this section.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{lumpBelow}}.

\sphinxAtStartPar
This option value represents a cut\sphinxhyphen{}off value. Alleles with an
expected value equal to or less than \sphinxcode{\sphinxupquote{lumpBelow}} will be lumped
together into a single category for the purpose of calculating the
degrees of freedom and overall \sphinxcode{\sphinxupquote{p}}\sphinxhyphen{}value for the chi\sphinxhyphen{}squared
Hardy\sphinxhyphen{}Weinberg test.

\end{itemize}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}HardyWeinbergGuoThompson{]}}}

\sphinxAtStartPar
When this section is present, an implementation of the
Hardy\sphinxhyphen{}Weinberg exact test is run using the original
\sphinxcite{docs/biblio:guo-thompson-1992} code, using a Monte\sphinxhyphen{}Carlo Markov chain (MCMC). In
addition, two measures (Chen and Diff) of the goodness of it of
individual genotypes are reported under this option \sphinxcite{docs/biblio:chen-etal-1999}
By default this section is not enabled. This is a different
implementation to the \sphinxstyleliteralstrong{\sphinxupquote{Arlequin}} version listed in
{\hyperref[\detokenize{docs/guide-chapter-usage:config-advanced}]{\sphinxcrossref{\DUrole{std,std-ref}{Advanced options}}}}, below.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{dememorizationSteps}}.

\sphinxAtStartPar
Number of steps of to “burn\sphinxhyphen{}in” the Markov chain before statistics
are collected.\sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{2000}} \sphinxstylestrong{{]}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{samplingNum}}.

\sphinxAtStartPar
Number of Markov chain samples \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{1000}} \sphinxstylestrong{{]}}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{samplingSize}}.

\sphinxAtStartPar
Markov chain sample size\sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{1000}} \sphinxstylestrong{{]}}.

\end{itemize}

\sphinxAtStartPar
Note that the \sphinxstylestrong{total} number of steps in the Monte\sphinxhyphen{}Carlo Markov
chain is the product of \sphinxcode{\sphinxupquote{samplingNum}} and \sphinxcode{\sphinxupquote{samplingSize}}, so the
default values described above would contain 1,000,000 (= 1000 x
1000) steps in the MCMC chain.

\sphinxAtStartPar
The default values for options described above have proved to be
optimal for us and if the options are not provided these defaults
will be used. If you change the values and have problems, please let
us \sphinxstylestrong{know}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}HomozygosityEWSlatkinExact{]}}}

\sphinxAtStartPar
The presence of this section enables Slatkin’s \sphinxcite{docs/biblio:slatkin-1994}
implementation of the Ewens\sphinxhyphen{}Watterson exact test of neutrality.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{numReplicates}}.

\sphinxAtStartPar
The default values have proved to be optimal for us. There is no
reason to change them unless you are particularly curious. If you
change the default values and have problems, please let us know.

\end{itemize}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}Emhaplofreq{]}}}

\sphinxAtStartPar
The presence of this section enables haplotype estimation and
calculation of linkage disequilibrium (LD) measures.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{lociToEstHaplo}}.

\sphinxAtStartPar
In this option you can list the multi\sphinxhyphen{}locus haplotypes for which
you wish the program to estimate and to calculate the LD. It
should be a comma\sphinxhyphen{}separated list of colon\sphinxhyphen{}joined loci. e.g.,

\begin{sphinxVerbatim}[commandchars=\\\{\}]
lociToEstHaplo=a:b:drb1,a:b:c,drb1:dqa1:dpb1,drb1:dqb1:dpb1
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{allPairwiseLD}}.

\sphinxAtStartPar
Set this to \sphinxcode{\sphinxupquote{1}} (one) if you want the program to calculate all
pairwise LD for your data, otherwise set this to \sphinxcode{\sphinxupquote{0}} (zero).

\end{itemize}

\end{itemize}
\phantomsection\label{\detokenize{docs/guide-chapter-usage:config-allpairwiseldwithpermu}}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{allPairwiseLDWithPermu}}.

\sphinxAtStartPar
Set this to a positive integer greater than 1 if you need to
determine the significance of the pairwise LD measures in the
previous section. The number you use is the number of permutations
that will be run to ascertain the significance (this should be at
least 1000 or greater). (Note this is done via permutation testing
performed after the pairwise LD test for all pairs of loci. Note
also that this test can take \sphinxstyleemphasis{DAYS} if your data is highly
polymorphic.)

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{numPermuInitCond}}.

\sphinxAtStartPar
Set this to change the number of initial conditions used per
permutation. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{5}} \sphinxstylestrong{{]}}. (\sphinxstyleemphasis{Note: this parameter is only used
if} \sphinxcode{\sphinxupquote{allPairwiseLDWithPermu}} \sphinxstyleemphasis{is set and nonzero}).

\end{itemize}


\subsection{Advanced options}
\label{\detokenize{docs/guide-chapter-usage:advanced-options}}\label{\detokenize{docs/guide-chapter-usage:config-advanced}}
\sphinxAtStartPar
The following section describes additional options to previously
described sections. Most of the time these options can be omitted and
PyPop will choose defaults, however these advanced options do offer
greater control over the application. In particular, customization will
be required for data that has sample identifiers as in
\hyperref[\detokenize{docs/guide-chapter-usage:data-minimal-noheader}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-minimal-noheader}}} or header data block as in
\hyperref[\detokenize{docs/guide-chapter-usage:data-hla}]{Listing \ref{\detokenize{docs/guide-chapter-usage:data-hla}}} and both \sphinxcode{\sphinxupquote{validSampleFields}} (described
above) and \sphinxcode{\sphinxupquote{validPopFields}} (described below) will need to be
modified.

\sphinxAtStartPar
It also describes two extra sections related to using PyPop in
conjunction with \sphinxstyleliteralstrong{\sphinxupquote{Arlequin}}: \sphinxcode{\sphinxupquote{{[}Arlequin{]}}} and
\sphinxcode{\sphinxupquote{{[}HardyWeinbergGuoThompsonArlequin{]}}}.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}General{]}}} \sphinxstylestrong{advanced options}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{txtOutFilename}} and \sphinxcode{\sphinxupquote{xmlOutFilename}}.

\sphinxAtStartPar
If you wish to specify a particular name for the output file, which
you want to remain identical over several runs, you can set these
two items to particular values. The default is to have the program
select the output filename, which can be controlled by the next
variable. \sphinxstylestrong{{[}Default: not used{]}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{outFilePrefixType}}.

\sphinxAtStartPar
This option can either be omitted entirely (in which case the
default will be \sphinxcode{\sphinxupquote{filename}}) or be set in several ways. The
default is set as \sphinxcode{\sphinxupquote{filename}}, which will result in three output
files named \sphinxcode{\sphinxupquote{original\sphinxhyphen{}filename\sphinxhyphen{}minus\sphinxhyphen{}suffix\sphinxhyphen{}out.xml}},
\sphinxcode{\sphinxupquote{original\sphinxhyphen{}filename\sphinxhyphen{}minus\sphinxhyphen{}suffix\sphinxhyphen{}out.txt}}, and
\sphinxcode{\sphinxupquote{original\sphinxhyphen{}filename\sphinxhyphen{}minus\sphinxhyphen{}suffix\sphinxhyphen{}filter.xml}}. \sphinxstylestrong{{[}Default:}
\sphinxcode{\sphinxupquote{filename}} \sphinxstylestrong{{]}}

\sphinxAtStartPar
If you set the value to \sphinxcode{\sphinxupquote{date}} instead of filename, you’ll get the
date incorporated in the filename as follows:
\sphinxcode{\sphinxupquote{original\sphinxhyphen{}filename\sphinxhyphen{}minus\sphinxhyphen{}suffix\sphinxhyphen{}YYYY\sphinxhyphen{}nn\sphinxhyphen{}dd\sphinxhyphen{}HH\sphinxhyphen{}MM\sphinxhyphen{}SS\sphinxhyphen{}out.\sphinxstyleemphasis{xml,txt}}}.
e.g., \sphinxcode{\sphinxupquote{USAFEL\sphinxhyphen{}UchiTelle\sphinxhyphen{}2003\sphinxhyphen{}09\sphinxhyphen{}21\sphinxhyphen{}01\sphinxhyphen{}29\sphinxhyphen{}35\sphinxhyphen{}out.xml}} (where Y, n,
d, H, M, S refer to year, month, day, hour, minute and second,
respectively).

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{xslFilename}}.

\sphinxAtStartPar
This option specifies where to find the XSLT file to use for
transforming PyPop’s xml output into human\sphinxhyphen{}readable form. Most users
will not normally need to set this option, and the default is the
system\sphinxhyphen{}installed \sphinxcode{\sphinxupquote{text.xsl}} file.

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}ParseGenotypeFile{]}}} \sphinxstylestrong{advanced options}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{fieldPairDesignator}}.

\sphinxAtStartPar
This option allows you to override the coding for the headers for
each pair of alleles at each locus; it must match the entry in the
config file under \sphinxcode{\sphinxupquote{validSampleFields}} and the entries in your
population data file. If you want to use something other than \sphinxcode{\sphinxupquote{\_1}}
and \sphinxcode{\sphinxupquote{\_2}}, change this option, for instance, to use letters and
parentheses, change it as follows: \sphinxcode{\sphinxupquote{fieldPairDesignator=(a):(b)}}
\sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{\_1:\_2}} \sphinxstylestrong{{]}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{popNameDesignator}}.

\sphinxAtStartPar
There is a special designator to mark the population name field,
which is usually the first field in the data block. \sphinxstylestrong{{[}Default:}
\sphinxcode{\sphinxupquote{+}} \sphinxstylestrong{{]}}

\sphinxAtStartPar
If you are analyzing data that contains a population name for each
sample, then the first entry in your \sphinxcode{\sphinxupquote{validSampleFields}} section
should have a prefixed +, as below:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
validSampleFields=+populat
 *a\PYGZus{}1
 *a\PYGZus{}2
 ...
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{validPopFields}}.

\sphinxAtStartPar
If you are analyzing data with an initial two line population header
block information as in {\hyperref[\detokenize{docs/guide-chapter-usage:data-hla}]{\sphinxcrossref{\DUrole{std,std-ref}{Multi\sphinxhyphen{}locus allele\sphinxhyphen{}level HLA genotype data with sample and header information}}}}, then you will
need to set this option. In this case, it should contain the field
names in the first line of the header information of your file.
\sphinxstylestrong{{[}Default: required when a population data\sphinxhyphen{}block is present in data
file{]}}, e.g.:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
validPopFields=labcode
 method
 ethnic
 country
 latit
 longit
\end{sphinxVerbatim}

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}Emhaplofreq{]}}} \sphinxstylestrong{advanced options}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{permutationPrintFlag}}.

\sphinxAtStartPar
Determines whether the likelihood ratio for each permutation will be
logged to the XML output file, this is disabled by default.
\sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{0}} \sphinxstylestrong{(i.e. OFF){]}}.

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxAtStartPar
If this is enabled it can \sphinxstyleemphasis{drastically} increase the size of the
output XML file on the order of the product of the number of
possible pairwise comparisons and permutations. Machines with
lower RAM and disk space may have difficulty coping with this.
\end{sphinxadmonition}

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}Arlequin{]}}} \sphinxstylestrong{extra section}

\sphinxAtStartPar
This section sets characteristics of the \sphinxstyleliteralstrong{\sphinxupquote{Arlequin}}
application if it has been installed (it must be installed separately
from PyPop as we cannot distribute it). The options in this section
are only used when a test requiring \sphinxstyleliteralstrong{\sphinxupquote{Arlequin}}, such as it’s
implementation of Guo and Thompson’s \sphinxcite{docs/biblio:guo-thompson-1992} Hardy\sphinxhyphen{}Weinberg
exact test is invoked (see below).
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{arlequinExec}}.

\sphinxAtStartPar
This option specifies where to find the \sphinxstyleliteralstrong{\sphinxupquote{Arlequin}}
executable on your system. The default assumes it is on your system
path. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{arlecore.exe}} \sphinxstylestrong{{]}}

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}HardyWeinbergGuoThompsonArlequin{]}}} \sphinxstylestrong{extra section}

\sphinxAtStartPar
When this section is present, \sphinxstyleliteralstrong{\sphinxupquote{Arlequin}}’s implementation of the
Hardy\sphinxhyphen{}Weinberg exact test is run, using a Monte\sphinxhyphen{}Carlo Markov Chain
implementation. By default this section is not enabled.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{markovChainStepsHW}}.

\sphinxAtStartPar
Length of steps in the Markov chain \sphinxstylestrong{{[}Default: 2500000{]}}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{markovChainDememorisationStepsHW}}.

\sphinxAtStartPar
Number of steps of to “burn\sphinxhyphen{}in” the Markov chain before statistics
are collected.\sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{5000}} \sphinxstylestrong{{]}}

\end{itemize}

\sphinxAtStartPar
The default values for options described above have proved to be optimal
for us and if the options are not provided these defaults will be used.
If you change the values and have problems, please let us \sphinxstylestrong{know}.

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}Filters{]}}} \sphinxstylestrong{extra section}

\sphinxAtStartPar
When this section is present, it allows you to specify succesive filters
to the data.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{filtersToApply}}.

\sphinxAtStartPar
Here you specify which filters you want applied to the data and the
order in which you want them applied. Separate each filter name with
a colon (\sphinxcode{\sphinxupquote{:}}). Currently there are four predefined filter:
\sphinxcode{\sphinxupquote{AnthonyNolan}}, \sphinxcode{\sphinxupquote{Sequence}}, \sphinxcode{\sphinxupquote{DigitBinning}}, and
\sphinxcode{\sphinxupquote{CustomBinning}}. If you specify one or more of these filters, you
will get the default behavior of the filter. If you wish to modify
the default behavior, you should add a section with the same name as
the specified filter(s). See next section for more on this. Please
note that, while you are allowed to specify any ordering for the
filters, some orderings may not make sense. For example, the ordering
Sequence:AnthonyNolan would not make sense (because as far as PyPop
is concerned, your alleles are now amino acid residues.) However, the
reverse ordering, AnthonyNolan:Sequence, would be logical and perhaps
even advisable.

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}AnthonyNolan{]}}} \sphinxstylestrong{filter section}

\sphinxAtStartPar
This section is \sphinxstyleemphasis{only} useful for HLA data. Like all filter sections, it
will only be used if present in the \sphinxcode{\sphinxupquote{filtersToApply}} line specified
above. If so enabled, your data will be filtered through the Anthony
Nolan database of known HLA allele names before processing. The data
files this filter relies on are \sphinxstyleemphasis{not} currently distributed with PyPop
but can be obtained via the \sphinxhref{ftp://ftp.ebi.ac.uk/pub/databases/imgt/mhc/hla/}{IMGT ftp
site} (ftp://ftp.ebi.ac.uk/pub/databases/imgt/mhc/hla/). Invocation of
this filter will produce a \sphinxcode{\sphinxupquote{popfile\sphinxhyphen{}filter.xml}} file output showing
what was resolved and what could not be resolved.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{alleleFileFormat}}.

\sphinxAtStartPar
This options specifies which of the formats the Anthony Nolan
allele data will be used. The option can be set to either \sphinxcode{\sphinxupquote{txt}}
(for the plain free text format) or \sphinxcode{\sphinxupquote{msf}} (for the \sphinxhref{http://www.ebi.ac.uk/imgt/hla/download.html}{Multiple
Sequence Format} (http://www.ebi.ac.uk/imgt/hla/download.html))
\sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{msf}} \sphinxstylestrong{{]}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{directory}}.

\sphinxAtStartPar
Specifies the path to the root of the sequence files. For \sphinxcode{\sphinxupquote{txt}}:
\sphinxstylestrong{{[}Default:}
\sphinxcode{\sphinxupquote{\sphinxstyleemphasis{prefix}/share/PyPop/anthonynolan/HIG\sphinxhyphen{}seq\sphinxhyphen{}pep\sphinxhyphen{}text/}}
\sphinxstylestrong{{]}}.  For \sphinxcode{\sphinxupquote{msf}} files \sphinxstylestrong{{[}Default:}
\sphinxcode{\sphinxupquote{\sphinxstyleemphasis{prefix}/share/PyPop/anthonynolan/msf/}} \sphinxstylestrong{{]}}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{preserve\sphinxhyphen{}ambiguous}}.

\sphinxAtStartPar
The default behavior of the \sphinxcode{\sphinxupquote{AnthonyNolan}} filter is to ignore
allele ambiguity (“slash”) notation. This notation, common in the
literature, looks like: \sphinxcode{\sphinxupquote{010101/0102/010301}}. The default behavior
will simply truncate this to \sphinxcode{\sphinxupquote{0101}}. If you want to preserve the
notation, set the option to \sphinxcode{\sphinxupquote{1}}. This will result in a filtered
allele “name” of \sphinxcode{\sphinxupquote{0101/0102/0103}} in the above hypothetical
example. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{0}} \sphinxstylestrong{{]}}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{preserve\sphinxhyphen{}unknown}}.

\sphinxAtStartPar
The default behavior of the \sphinxcode{\sphinxupquote{AnthonyNolan}} filter is to replace
unknown alleles with the \sphinxcode{\sphinxupquote{untypedAllele}} designator. If you want
the filter to keep allele names it does not recognize, set the option
to \sphinxcode{\sphinxupquote{1}}. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{0}} \sphinxstylestrong{{]}}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{preserve\sphinxhyphen{}lowres}}.

\sphinxAtStartPar
This option is similar to \sphinxcode{\sphinxupquote{preserve\sphinxhyphen{}unknown}}, but only applies to
lowres alleles. If set to \sphinxcode{\sphinxupquote{1}}, PyPop will keep allele names that are
shorter than the default allele name length, usually 4 digits long.
But if the preserve\sphinxhyphen{}unknown flag is set, this one has no effect,
because all unknown alleles are preserved. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{0}} \sphinxstylestrong{{]}}.

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}Sequence{]}}} \sphinxstylestrong{filter section}

\sphinxAtStartPar
This section allows configuration of the sequence filter. Like all
filter sections, it will only will be used if present in the
\sphinxcode{\sphinxupquote{filtersToApply}} line specified above. If so enabled, your allele
names will be translated into sequences, and all ensuing analyses will
consider each position in the sequence to be a distinct locus. This
filter makes use of the same msf format alignment files as used above in
the AnthonyNolan filter. It does not work with the txt format alignment
files.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{sequenceFileSuffix}}.

\sphinxAtStartPar
Determines the files that will be examined in order to read in a
sequence for each allele. (ie, if the file for locus A is
\sphinxcode{\sphinxupquote{A\_prot.msf}}, the value would be \sphinxcode{\sphinxupquote{\_prot}} whereas if you
wanted to use the nucleotide sequence files, you might use
\sphinxcode{\sphinxupquote{\_nuc}}.) \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{\_prot}} \sphinxstylestrong{{]}}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{directory}}.

\sphinxAtStartPar
Specifies the path to the root of the sequence files, in the same
manner as in the AnthonyNolan section, above.

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}DigitBinning{]}}} \sphinxstylestrong{filter section}

\sphinxAtStartPar
This section allows configuration of the DigitBinning filter. Like all
filter sections, it will be used if present in the \sphinxcode{\sphinxupquote{filtersToApply}}
line specified above. If so enabled, your allele names will be truncated
after the nth digit.
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{binningDigits}}.

\sphinxAtStartPar
An integer that specifies how many digits to keep after the
truncation. \sphinxstylestrong{{[}Default:} \sphinxcode{\sphinxupquote{4}} \sphinxstylestrong{{]}}.

\end{itemize}

\sphinxAtStartPar
\sphinxcode{\sphinxupquote{{[}CustomBinning{]}}} \sphinxstylestrong{filter section}

\sphinxAtStartPar
This section allows configuration of the CustomBinning filter. Like all
filter sections, it will only be used if present in the
\sphinxcode{\sphinxupquote{filtersToApply}} line specified above.

\sphinxAtStartPar
You can provide a set of custom rules for replacing allele names. Allele
names should be separated by \sphinxcode{\sphinxupquote{/}} marks. This filter matches any allele
names that are exactly the same as the ones you list here, and will also
find “close matches” (but only if there are no exact matches.). Here is
an example:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
A=01/02/03
 04/05/0306
 !06/1201/1301
 !07/0805
\end{sphinxVerbatim}

\sphinxAtStartPar
In the example above, \sphinxcode{\sphinxupquote{A*03}} alleles will match to \sphinxcode{\sphinxupquote{01/02/03}},
except for \sphinxcode{\sphinxupquote{A*0306}}, which will match to \sphinxcode{\sphinxupquote{04/05/0306}}. If you place
a \sphinxcode{\sphinxupquote{!}} mark in front of the first allele name, that first name will be
used as the “new name” for the binned group (for example, \sphinxcode{\sphinxupquote{A*0805}}
will be called \sphinxcode{\sphinxupquote{07}} in the custom\sphinxhyphen{}binned data.) Note that the space at
the beginning of the lines (following the first line of each locus) is
important. The above rules are just dummy examples, provided to
illustrate how the filter works. PyPop is distributed with a
biologically relevant set of \sphinxcode{\sphinxupquote{CustomBinning}} rules that have been
compiled from several sources %
\begin{footnote}[2]\sphinxAtStartFootnote
\sphinxcite{docs/biblio:mack-etal-2007}; \sphinxcite{docs/biblio:cano-etal-2007}; The Anthony Nolan list of deleted
allele names
(\textasciigrave{} \textless{}\sphinxurl{http://www.anthonynolan.com/HIG/lists/delnames.html}\textgreater{}\textasciigrave{}\_\_); and the
Ambiguous Allele Combinations, release 2.18.0
(\textasciigrave{} \textless{}\sphinxurl{http://www.ebi.ac.uk/imgt/hla/ambig.html}\textgreater{}\textasciigrave{}\_\_).
%
\end{footnote}

\sphinxstepscope


\chapter{Interpreting PyPop output}
\label{\detokenize{docs/guide-chapter-instructions:interpreting-pypop-output}}\label{\detokenize{docs/guide-chapter-instructions::doc}}
\sphinxAtStartPar
As mentioned in {\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-intro-run-details}]{\sphinxcrossref{\DUrole{std,std-ref}{What happens when you run PyPop?}}}}, The XML file
is the primary output created by PyPop and contains the complete set of
results. The text output, generated from the XML file via XSLT, contains
a human\sphinxhyphen{}readable summary of the XML results. Below we discuss the output
contained in this text file.


\section{Population summary}
\label{\detokenize{docs/guide-chapter-instructions:population-summary}}\label{\detokenize{docs/guide-chapter-instructions:instructions-pop-summary}}
\sphinxAtStartPar
A \sphinxcode{\sphinxupquote{Population Summary}} is generated for each dataset analyzed. This
summary provides basic demographic information and summarizes
information about the sample size.

\sphinxAtStartPar
Sample output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Population Summary
==================
Population Name: UchiTelle
       Lab code: USAFEL
  Typing method: 12th Workshop SSOP
      Ethnicity: Telle
      Continent: NW Asia
Collection site: Targen Village
       Latitude: 41 deg 12 min N
      Longitude: 94 deg 7 min E

Population Totals
\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}
Sample Size (n): 47
Allele Count (2n): 94
Total Loci in file: 9
Total Loci with data: 8
\end{sphinxVerbatim}


\section{Single locus analyses}
\label{\detokenize{docs/guide-chapter-instructions:single-locus-analyses}}\label{\detokenize{docs/guide-chapter-instructions:instructions-locus-info}}

\subsection{Basic allele count information}
\label{\detokenize{docs/guide-chapter-instructions:basic-allele-count-information}}\label{\detokenize{docs/guide-chapter-instructions:instructions-allelecounts}}
\sphinxAtStartPar
Information relevant to individual loci is reported. Sample size and
allele counts will differ among loci if not all individuals were typed
at each locus. Untyped individuals are those for which one or two
alleles were not reported. The alleles are listed in descending
frequency (and count) in the left hand column, and are sorted
numerically in the right column. The number of distinct alleles \sphinxcode{\sphinxupquote{k}} is
reported.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
I. Single Locus Analyses
========================

1. Locus: A
\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}

1.1. Allele Counts [A]
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Untyped individuals: 2
Sample Size (n): 45
Allele Count (2n): 90
Distinct alleles (k): 10

Counts ordered by frequency   | Counts ordered by name
Name      Frequency (Count)   | Name      Frequency (Count)
0201      0.21111   19        | 0101      0.13333   12
0301      0.15556   14        | 0201      0.21111   19
0101      0.13333   12        | 0210      0.10000   9
2501      0.12222   11        | 0218      0.10000   9
0210      0.10000   9         | 0301      0.15556   14
0218      0.10000   9         | 2501      0.12222   11
3204      0.08889   8         | 3204      0.08889   8
6901      0.04444   4         | 6814      0.03333   3
6814      0.03333   3         | 6901      0.04444   4
7403      0.01111   1         | 7403      0.01111   1
Total     1.00000   90        | Total     1.00000   90
\end{sphinxVerbatim}

\sphinxAtStartPar
In the cases where there is no information for a locus, a message is
displayed indicating lack of data.

\sphinxAtStartPar
Sample output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
4. Locus: DRA
\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}
 No data for this locus!
\end{sphinxVerbatim}


\subsection{Chi\sphinxhyphen{}square test for deviation from Hardy\sphinxhyphen{}Weinberg proportions (HWP).}
\label{\detokenize{docs/guide-chapter-instructions:chi-square-test-for-deviation-from-hardy-weinberg-proportions-hwp}}\label{\detokenize{docs/guide-chapter-instructions:instructions-hardyweinberg}}
\sphinxAtStartPar
For each locus, the observed genotype counts are compared to those
expected under Hardy Weinberg proportions (HWP). A triangular matrix
reports observed and expected genotype counts. If the matrix is more
than 80 characters, the output is split into different sections. Each
cell contains the observed and expected number for a given genotype in
the format \sphinxcode{\sphinxupquote{observed/expected}}.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
6.2. HardyWeinberg [DQA1]
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Table of genotypes, format of each cell is: observed/expected.

0201 8/5.1
0301 4/4.0 1/0.8
0401 3/6.9 1/2.7 6/2.3
0501 8/9.9 5/3.8 5/6.7 6/4.8
      0201  0301  0401  0501
                             [Cols: 1 to 4]
\end{sphinxVerbatim}

\sphinxAtStartPar
The values in this matrix are used to test hypotheses of deviation from
HWP. The output also includes the chi\sphinxhyphen{}square statistic, the number of
degrees of freedom and associated \(p\)\sphinxhyphen{}value for a number of classes of
genotypes and is summarized in the following table:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
                      Observed    Expected  Chi\PYGZhy{}square   DoF   p\PYGZhy{}value
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
            Common         N/A         N/A        4.65     1  0.0310*
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
  Lumped genotypes         N/A         N/A        1.17     1  0.2797
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
   Common + lumped         N/A         N/A        5.82     1  0.0158*
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
   All homozygotes          21       13.01        4.91     1  0.0268*
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
 All heterozygotes          26       33.99        1.88     1  0.1706
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Common heterozygotes by allele
              0201          15       20.78        1.61        0.2050
              0301          10       10.47        0.02        0.8850
              0401           9       16.31        3.28        0.0703
              0501          18       20.43        0.29        0.5915

\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Common genotypes
         0201:0201           8        5.11        1.63        0.2014
         0201:0401           3        6.93        2.23        0.1358
         0201:0501           8        9.89        0.36        0.5472
         0401:0501           5        6.70        0.43        0.5109
             Total          24       28.63
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
\end{sphinxVerbatim}
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxstylestrong{Common.}

\sphinxAtStartPar
The result for goodness of fit to HWP using only the genotypes with
at least \sphinxcode{\sphinxupquote{lumpBelow}} expected counts (the common genotypes) (in the
output shown throughout this example \sphinxcode{\sphinxupquote{lumpBelow}} is equal to 5).

\sphinxAtStartPar
If the dataset contains no genotypes with expected counts equal or
greater than \sphinxcode{\sphinxupquote{lumpBelow}}, then there are no common genotypes and
the following message is reported:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
No common genotypes; chi\PYGZhy{}square cannot be calculated
\end{sphinxVerbatim}

\sphinxAtStartPar
The analysis of common genotypes may lead to a situtation where there
are fewer classes (genotypes) than allele frequencies to estimate.
This means that the analysis cannot be performed (degrees of freedom
\textless{} 1). In such a case the following message is reported, explaining
why the analysis could not be performed:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Too many parameters for chi\PYGZhy{}square test.
\end{sphinxVerbatim}

\sphinxAtStartPar
To obviate this as much as possible, only alleles which occur in
common genotypes are used in the calculation of degrees of freedom.

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{Lumped genotypes.}

\sphinxAtStartPar
The result for goodness of fit to HWP for the pooled set of genotypes
that individually have less than \sphinxcode{\sphinxupquote{lumpBelow}} expected counts.

\sphinxAtStartPar
The pooling procedure is designed to avoid carrying out the
chi\sphinxhyphen{}square goodness of fit test in cases where there are low expected
counts, which could lead to spurious rejection of HWP. However, in
certain cases it may not be possible to carry out this pooling
approach. The interpretation of results based on lumped genotypes
will depend on the particular genotypes that are combined in this
class.

\sphinxAtStartPar
If the sum of expected counts in the lumped class does not add up to
\sphinxcode{\sphinxupquote{lumpBelow}}, then the test for the lumped genotypes cannot be
calculated and the following message is reported:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
The total number of expected genotypes is less than 5
\end{sphinxVerbatim}

\sphinxAtStartPar
This may by remedied by combining rare alleles and recalculating
overall chi\sphinxhyphen{}square value and degrees of freedom. (This would require
appropriate manipulation of the data set by hand and is not a feature
of PyPop).

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{Common + lumped.}

\sphinxAtStartPar
The result for goodness of fit to HWP for both the common and the
lumped genotypes.

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{All homozygotes.}

\sphinxAtStartPar
The result for goodness of fit to HWP for the pooled set of
homozygous genotypes.

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{All heterozygotes.}

\sphinxAtStartPar
The result for goodness of fit to HWP for the pooled set of
heterozygous genotypes.

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{Common heterozygotes.}

\sphinxAtStartPar
The common heterozygotes by allele section summarizes the observed
and expected number of counts of all heterozygotes carrying a
specific allele with expected value GE \sphinxcode{\sphinxupquote{lumpBelow}}.

\item {} 
\sphinxAtStartPar
\sphinxstylestrong{Common genotypes.}

\sphinxAtStartPar
The common genotypes by genotype section lists observed, expected,
chi\sphinxhyphen{}square and \(p\)\sphinxhyphen{}values for all observed genotypes with expected
values GE \sphinxcode{\sphinxupquote{lumpBelow}}.

\end{itemize}


\subsection{Exact test for deviation from HWP}
\label{\detokenize{docs/guide-chapter-instructions:exact-test-for-deviation-from-hwp}}\label{\detokenize{docs/guide-chapter-instructions:instructions-hardyweinberg-exact}}
\sphinxAtStartPar
If enabled in the configuration file, the exact test for deviations from
HWP will be output. The exact test uses the method of \sphinxcite{docs/biblio:guo-thompson-1992}.
The \(p\)\sphinxhyphen{}value provided describes how probable the observed set of
genotypes is, with respect to a large sample of other genotypic
configurations (conditioned on the same allele frequencies and \(2n\)).
\(p\)\sphinxhyphen{}values lower than 0.05 can be interpreted as evidence that the
sample does not fit HWP. In addition, those individual genotypes
deviating significantly (\(p< 0.05\)) from expected HWP as
computed with the Chen and “diff” measures are reported.

\sphinxAtStartPar
There are two implementations for this test, the first using the gthwe
implementation originally due to Guo \& Thompson, but modified by John
Chen, the second being Arlequin’s \sphinxcite{docs/biblio:schneider-etal-2000} implementation.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
6.3. Guo and Thompson HardyWeinberg output [DQA1]
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Total steps in MCMC: 1000000
Dememorization steps: 2000
Number of Markov chain samples: 1000
Markov chain sample size: 1000
Std. error: 0.0009431
p\PYGZhy{}value (overall): 0.0537
\end{sphinxVerbatim}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
6.4. Guo and Thompson HardyWeinberg output(Arlequin\PYGZsq{}s implementation) [DQA1]
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Observed heterozygosity: 0.553190
Expected heterozygosity: 0.763900
Std. deviation: 0.000630
Dememorization steps: 100172
p\PYGZhy{}value: 0.0518
\end{sphinxVerbatim}

\sphinxAtStartPar
Note that in the Arlequin implementation, the output is slightly
different, and the only directly comparable value between the two
implementation is the \(p\)\sphinxhyphen{}value. These \(p\)\sphinxhyphen{}values may be slightly
different, but should agree to within one significant figure.


\subsection{The Ewens\sphinxhyphen{}Watterson homozygosity test of neutrality}
\label{\detokenize{docs/guide-chapter-instructions:the-ewens-watterson-homozygosity-test-of-neutrality}}\label{\detokenize{docs/guide-chapter-instructions:instructions-homozygosity}}
\sphinxAtStartPar
For each locus, we implement the Ewens\sphinxhyphen{}Watterson homozygosity test of
neutrality (\sphinxcite{docs/biblio:ewens-1972}; \sphinxcite{docs/biblio:watterson-1978}). We use the term
\sphinxstyleemphasis{observed homozygosity} to denote the homozygosity statistic
(\(F\)), computed as the sum of the squared allele
frequencies. This value is compared to the \sphinxstyleemphasis{expected homozygosity}
which is computed by simulation under neutrality/equilibrium
expectations, for the same sample size (\(2n\)) and number of
unique alleles (\(k\)). Note that the homozygosity \sphinxcode{\sphinxupquote{F}}
statistic, \(F=\sum_{i=1}^{k}p_{i}^{2}\), is often referred to as
the \sphinxstyleemphasis{expected homozygosity} (with \sphinxstyleemphasis{expectation} referring to HWP) to
distinguish it from the observed proportion of homozygotes. We avoid
referring to the observed \(F\) statistic as the “\sphinxstyleemphasis{observed
expected homozygosity}” (to simplify and hopefully avoid confusion)
since the homozygosity test of neutrality is concerned with
comparisons of observed results to expectations under neutrality. Both
the \sphinxstyleemphasis{observed} statistic (based on the actual data) and \sphinxstyleemphasis{expected}
statistic (based on simulations under neutrality) used in this test
are computed as the sum of the squared allele frequencies.

\sphinxAtStartPar
The \sphinxstyleemphasis{normalized deviate of the homozygosity} (\(F_{nd}\)) is the
difference between the \sphinxstyleemphasis{observed homozygosity} and \sphinxstyleemphasis{expected
homozygosity}, divided by the square root of the variance of the
expected homozygosity (also obtained by simulations; \sphinxcite{docs/biblio:salamon-etal-1999}).
Significant negative normalized deviates imply \sphinxstyleemphasis{observed homozygosity}
values lower than \sphinxstyleemphasis{expected homozygosity}, in the direction of balancing
selection. Significant positive values are in the direction of
directional selection.

\sphinxAtStartPar
The \(p\)\sphinxhyphen{}value in the last row of the output is the probability of
obtaining a homozygosity \(F\) statistic under neutral evolution that is
less than or equal to the observed \(F\) statistic. It is computed based
on the null distribution of homozygosity \(F\) values simulated under
neutrality/equilibrium conditions for the same sample size (\(2n\)) and
number of unique alleles (\(k\)). For a one\sphinxhyphen{}tailed test of the null
hypothesis of neutrality against the alternative of balancing selection,
\(p\)\sphinxhyphen{}values less than 0.05 are considered significant at the 0.05
level. For a two\sphinxhyphen{}tailed test against the alternative of either balancing
or directional selection, \(p\)\sphinxhyphen{}values less than 0.025 or greater than
0.975 can be considered significant at the 0.05 level.

\sphinxAtStartPar
The standard implementation of the test uses a Monte\sphinxhyphen{}Carlo
implementation of the exact test written by Slatkin (\sphinxcite{docs/biblio:slatkin-1994};
\sphinxcite{docs/biblio:slatkin-1996}). A Markov\sphinxhyphen{}chain Monte Carlo method is used to obtain the
null distribution of the homozygosity statistic under neutrality. The
reported \(p\)\sphinxhyphen{}values are one\sphinxhyphen{}tailed (against the alternative of
balancing selection), but can be interpreted for a two\sphinxhyphen{}tailed test by
considering either extreme of the distribution (\textless{} 0.025 or \textgreater{} 0.975) at
the 0.05 level.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
1.6. Slatkin\PYGZsq{}s implementation of EW homozygosity test of neutrality [A]
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Observed F: 0.1326, Expected F: 0.2654, Variance in F: 0.0083
Normalized deviate of F (Fnd): \PYGZhy{}1.4603, p\PYGZhy{}value of F: 0.0029**
\end{sphinxVerbatim}

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxAtStartPar
The version of this test based on tables of simulated percentiles of
the Ewens\sphinxhyphen{}Watterson statistics is now disabled by default and its use
is deprecated in preference to the Slatkin exact test described
above, however some older PyPop runs may include output, so it is
documented here for completeness. This version differs from the
Monte\sphinxhyphen{}Carlo Markov Chain version described above in that the data is
simulated under neutrality to obtain the required statistics.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
1.4. Ewens\PYGZhy{}Watterson homozygosity test of neutrality [A]
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Observed F: 0.1326, Expected F: 0.2651, Normalized deviate (Fnd): \PYGZhy{}1.4506
p\PYGZhy{}value range: 0.0000 \PYGZlt{} p \PYGZlt{}= 0.0100 *
\end{sphinxVerbatim}
\end{sphinxadmonition}


\section{Multi\sphinxhyphen{}locus analyses}
\label{\detokenize{docs/guide-chapter-instructions:multi-locus-analyses}}\label{\detokenize{docs/guide-chapter-instructions:instructions-haplo}}
\sphinxAtStartPar
Haplotype frequencies are estimated using the iterative
Expectation\sphinxhyphen{}Maximization (EM) algorithm (\sphinxcite{docs/biblio:dempster-1977};
\sphinxcite{docs/biblio:excoffier-slatkin-1995}). Multiple starting conditions are used to
minimize the possibility of local maxima being reached by the EM
iterations. The haplotype frequencies reported are those that correspond
to the highest logarithm of the sample likelihood found over the
different starting conditions and are labeled as the maximum likelihood
estimates (MLE).

\sphinxAtStartPar
The output provides the names of loci for which haplotype frequencies
were estimated, the number of individual genotypes in the dataset
(\sphinxcode{\sphinxupquote{before\sphinxhyphen{}filtering}}), the number of genotypes that have data for all
loci for which haplotype estimation will be performed
(\sphinxcode{\sphinxupquote{after\sphinxhyphen{}filtering}}), the number of unique phenotypes (unphased
genotypes), the number of unique phased genotypes, the total number of
possible haplotypes that are compatible with the genotypic data (many of
these will have an estimated frequency of zero), and the log\sphinxhyphen{}likelihood
of the observed genotypes under the assumption of linkage equilibrium.


\subsection{All pairwise LD}
\label{\detokenize{docs/guide-chapter-instructions:all-pairwise-ld}}\label{\detokenize{docs/guide-chapter-instructions:instructions-pairwise-ld}}
\sphinxAtStartPar
A series of linkage disequilibrium (LD) measures are provided for each
pair of loci, as shown in the sample output below.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
II. Multi\PYGZhy{}locus Analyses
========================

Haplotype/ linkage disequlibrium (LD) statistics
\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}\PYGZus{}

Pairwise LD estimates
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Locus pair        D      D\PYGZsq{}      Wn  ln(L\PYGZus{}1) ln(L\PYGZus{}0)      S  ALD\PYGZus{}1\PYGZus{}2  ALD\PYGZus{}2\PYGZus{}1
A:C         0.01465 0.49229 0.39472  \PYGZhy{}289.09 \PYGZhy{}326.81  75.44  0.41435  0.37525
A:B         0.01491 0.50895 0.40145  \PYGZhy{}293.47 \PYGZhy{}330.84  74.73  0.40726  0.39512
A:DRB1      0.01299 0.42896 0.38416  \PYGZhy{}282.00 \PYGZhy{}309.16  54.32  0.32934  0.38370
A:DQA1      0.01219 0.33413 0.36466  \PYGZhy{}269.57 \PYGZhy{}286.08  33.02  0.25803  0.34897
A:DQB1      0.01356 0.39266 0.37495  \PYGZhy{}275.58 \PYGZhy{}297.62  44.07  0.29931  0.37489
A:DPA1      0.01681 0.32397 0.36666  \PYGZhy{}219.78 \PYGZhy{}226.97  14.38  0.19446  0.35360
A:DPB1      0.01362 0.42240 0.40404  \PYGZhy{}237.85 \PYGZhy{}262.06  48.42  0.33848  0.41739
C:B         0.04125 0.88739 0.85752  \PYGZhy{}210.37 \PYGZhy{}342.68 264.63  0.84781  0.86104
C:DRB1      0.01698 0.48046 0.47513  \PYGZhy{}280.34 \PYGZhy{}317.66  74.62  0.32308  0.47691
C:DQA1      0.02072 0.47797 0.49368  \PYGZhy{}263.23 \PYGZhy{}293.74  61.01  0.31386  0.50338
C:DQB1      0.01766 0.45793 0.49879  \PYGZhy{}269.55 \PYGZhy{}305.28  71.46  0.30479  0.50122
C:DPA1      0.02039 0.41030 0.46438  \PYGZhy{}224.72 \PYGZhy{}236.52  23.61  0.21172  0.46433
C:DPB1      0.01898 0.46453 0.37002  \PYGZhy{}242.45 \PYGZhy{}268.46  52.01  0.33462  0.45327
B:DRB1      0.01723 0.50254 0.41712  \PYGZhy{}286.79 \PYGZhy{}320.50  67.42  0.32654  0.43913
B:DQA1      0.01845 0.44225 0.43582  \PYGZhy{}271.36 \PYGZhy{}296.59  50.45  0.28877  0.44993
B:DQB1      0.01958 0.49040 0.43654  \PYGZhy{}277.30 \PYGZhy{}308.13  61.65  0.31328  0.45679
B:DPA1      0.01875 0.37441 0.40117  \PYGZhy{}229.76 \PYGZhy{}239.16  18.80  0.20689  0.40443
B:DPB1      0.01898 0.46082 0.38001  \PYGZhy{}247.84 \PYGZhy{}272.77  49.86  0.32227  0.45680
DRB1:DQA1   0.06138 0.92556 0.92465  \PYGZhy{}164.06 \PYGZhy{}271.56 214.99  0.82051  0.93006
DRB1:DQB1   0.06058 1.00000 1.00000  \PYGZhy{}147.74 \PYGZhy{}283.10 270.72  0.93302  1.00000

...
\end{sphinxVerbatim}

\sphinxAtStartPar
For each locus pair, we report three measures of overall linkage
disequilibrium. \(D'\) \sphinxcite{docs/biblio:hedrick-1987} weights the contribution to
LD of specific allele pairs by the product of their allele frequencies
(\sphinxcode{\sphinxupquote{D\textquotesingle{}}} in the output); \(W_n\) \sphinxcite{docs/biblio:cramer-1946} is a re\sphinxhyphen{}expression
of the chi\sphinxhyphen{}square statistic for deviations between observed and
expected haplotype frequencies (\sphinxcode{\sphinxupquote{W\_n}} in the
output)). \(W_{A/B}\) and \(W_{B/A}\) (\sphinxcode{\sphinxupquote{ALD\_1\_2}} and
\sphinxcode{\sphinxupquote{ALD\_2\_1}}, respectively in the output) are extensions of \(W_n\)
that account for asymmetry when the number of alleles differs at two
loci \sphinxcite{docs/biblio:thomson-single-2014}. Below we describe the measures, each of
which is normalized to lie between zero and one.
\begin{description}
\sphinxlineitem{\(D'\)}
\sphinxAtStartPar
Overall LD, summing contributions (\(D'_{ij}=D_{ij} /D_{max}\)) of all the haplotypes in a
multi\sphinxhyphen{}allelic two\sphinxhyphen{}locus system, can be measured using Hedrick’s
\(D'\) statistic, using the products of allele frequencies at the
loci, \(p_i\) and \(q_j\), as weights.

\end{description}
\begin{equation*}
\begin{split}{D}' = \sum_{i=1}^{I} {\sum_{j=1}^{J} {p_i } } q_j \left|{{D}'_{ij} } \right|\end{split}
\end{equation*}\begin{description}
\sphinxlineitem{\(W_n\)}
\sphinxAtStartPar
Also known as Cramer’s V Statistic \sphinxcite{docs/biblio:cramer-1946}, \(W_n\), is a
second overall measure of LD between two loci. It is a re\sphinxhyphen{}expression
of the Chi\sphinxhyphen{}square statistic, \(X^2_{LD}\), normalized
to be between zero and one. When there are only two alleles per
locus, \(W_n\) is equivalent to the correlation coefficient
between the two loci, defined as:

\end{description}
\begin{equation*}
\begin{split}W_n = \left[ {\frac{\sum_{i=1}^{I} {\sum_{j=1}^{J}{D_{ij}^2 / p_i } q_j } }{\min (I - 1,J - 1)}} \right]^{\frac{1}{2}} = \left[ {\frac{X_{LD}^2 / 2N}{\min (I - 1,J - 1)}}\right]^{\frac{1}{2}}\end{split}
\end{equation*}\begin{description}
\sphinxlineitem{two alleles case}
\sphinxAtStartPar
When there are only two alleles per locus, \(W_n\) is equivalent
to the correlation coefficient between the two loci, defined as
\(r =\sqrt {D_{11} / p_1 p_2 q_1 q_2 }\).

\sphinxlineitem{\(W_{A/B}\) and \(W_{B/A}\)}
\sphinxAtStartPar
When there are different numbers of alleles at the two loci,
the direct correlation property for the \(r\) correlation
measure is not retained by \(W_n\), its multi\sphinxhyphen{}allelic extension.
The complementary pair of conditional asymmetric LD (ALD) measures,
\(W_{A/B}\) and \(W_{B/A}\), were developed to extend the \(W_n\) measure.
\(W_{A/B}\) is (inversely) related to the
degree of variation of A locus alleles on haplotypes conditioned
on B locus alleles. If there is no variation of A locus alleles
on haplotypes conditioned on B locus alleles, then \(W_{A/B} = 1\)
\(W_{A/B} = W_{B/A} = W_n\) when there is symmetry in the data and
thus for bi\sphinxhyphen{}allelic SNPs.

\end{description}
\begin{equation*}
\begin{split}W_{A/B} = \left[ {\frac{\sum_{i=1}^{I} {\sum_{j=1}^{J}{D_{ij}^2 / q_j } } }{ 1 - F_A }} \right]^{\frac{1}{2}}\end{split}
\end{equation*}\begin{equation*}
\begin{split}W_{B/A} = \left[ {\frac{\sum_{i=1}^{I} {\sum_{j=1}^{J}{D_{ij}^2 / p_i } } }{ 1 - F_B }} \right]^{\frac{1}{2}}\end{split}
\end{equation*}
\sphinxAtStartPar
In addition to the LD measures described above, for each locus pair,
we describe three additional measures related to the log\sphinxhyphen{}likelihood
that are displayed in the output above:
\begin{description}
\sphinxlineitem{\(\ln(L_1)\)}
\sphinxAtStartPar
the log\sphinxhyphen{}likelihood of obtaining the observed data given the inferred
haplotype frequencies (\sphinxcode{\sphinxupquote{ln(L\_1)}} in the output)

\sphinxlineitem{\(\ln(L_0)\)}
\sphinxAtStartPar
the log\sphinxhyphen{}likelihood of the data under the null hypothesis of linkage
equilibrium (\sphinxcode{\sphinxupquote{ln(L\_0)}} in the output)

\sphinxlineitem{\(S\)}
\sphinxAtStartPar
the statistic (\sphinxcode{\sphinxupquote{S}} in the output) is defined as twice the
difference between these likelihoods. \(S\) has an asymptotic
chi\sphinxhyphen{}square distribution, but the null distribution of \(S\) is
better approximated using a randomization procedure. If a
permutation test is requested (by setting the option
\sphinxcode{\sphinxupquote{allPairwiseLDWithPermu}} to a a number greater than zero in the
{\hyperref[\detokenize{docs/guide-chapter-usage:config-allpairwiseldwithpermu}]{\sphinxcrossref{\DUrole{std,std-ref}{.ini file}}}}), the empirical
distribution of \(S\) is generated by shuffling genotypes among
individuals, separately for each locus, thus creating linkage
equilibrium. The additional column \sphinxcode{\sphinxupquote{\# permu}} that will be
generated (not shown in the example output above) will indicate how
many permutations were carried out. The \(p\)\sphinxhyphen{}value (also not
shown) will be the fraction of permutations that results in values of
\sphinxtitleref{S} greater or equal to that observed. A \(p < 0.05\) is
indicative of overall significant LD.

\end{description}

\sphinxAtStartPar
Individual LD coefficients, \(D_{ij}\), are stored in the XML
output file, but are not printed in the default text output. They can
be accessed in the summary text files created by the \sphinxcode{\sphinxupquote{popmeta}}
script (see {\hyperref[\detokenize{docs/guide-chapter-usage:guide-usage-intro-run-details}]{\sphinxcrossref{\DUrole{std,std-ref}{What happens when you run PyPop?}}}}).


\subsection{Haplotype frequency estimation}
\label{\detokenize{docs/guide-chapter-instructions:haplotype-frequency-estimation}}\label{\detokenize{docs/guide-chapter-instructions:instructions-haplotype-freqs}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
Haplotype frequency est. for loci: A:B:DRB1
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
Number of individuals: 47 (before\PYGZhy{}filtering)
Number of individuals: 45 (after\PYGZhy{}filtering)
Unique phenotypes: 45
Unique genotypes: 113
Number of haplotypes: 188
Loglikelihood under linkage equilibrium [ln(L\PYGZus{}0)]: \PYGZhy{}472.700542
Loglikelihood obtained via the EM algorithm [ln(L\PYGZus{}1)]: \PYGZhy{}340.676530
Number of iterations before convergence: 67
\end{sphinxVerbatim}

\sphinxAtStartPar
The estimated haplotype frequencies are sorted alphanumerically by
haplotype name (left side), or in decreasing frequency (right side).
Only haplotypes estimated at a frequency of 0.00001 or larger are
reported. The first column gives the allele names in each of the three
loci, the second column provides the maximum likelihood estimate for
their frequencies, (\sphinxcode{\sphinxupquote{frequency}}), and the third column gives the
corresponding approximate number of haplotypes (\sphinxcode{\sphinxupquote{\# copies}}).

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Haplotypes sorted by name             | Haplotypes sorted by frequency
haplotype         frequency \PYGZsh{} copies  | haplotype         frequency \PYGZsh{} copies
0101:1301:0402:   0.02222   2.0       | 0201:1401:0402:   0.03335   3.0
0101:1301:1101:   0.01111   1.0       | 3204:1401:0802:   0.03333   3.0
0101:1401:0901:   0.01111   1.0       | 0301:1401:0407:   0.03333   3.0
0101:1520:0802:   0.01111   1.0       | 0301:1301:0402:   0.03333   3.0
0101:1801:0407:   0.01111   1.0       | 0201:1401:1101:   0.03332   3.0
0101:3902:0404:   0.01111   1.0       | 0301:1520:0802:   0.02222   2.0
0101:3902:1602:   0.01111   1.0       | 0101:4005:0802:   0.02222   2.0
0101:4005:0802:   0.02222   2.0       | 0301:3902:0402:   0.02222   2.0
0101:8101:0802:   0.01111   1.0       | 0201:1301:1602:   0.02222   2.0
0101:8101:1602:   0.01111   1.0       | 0218:1401:0404:   0.02222   2.0
0201:1301:1602:   0.02222   2.0       | 0210:5101:1602:   0.02222   2.0
0201:1401:0402:   0.03335   3.0       | 0218:1401:1602:   0.02222   2.0
0201:1401:0404:   0.01111   1.0       | 0101:1301:0402:   0.02222   2.0
0201:1401:0407:   0.02222   2.0       | 2501:4005:0802:   0.02222   2.0
0201:1401:0802:   0.01111   1.0       | 2501:1301:0802:   0.02222   2.0

...
\end{sphinxVerbatim}

\sphinxstepscope


\chapter{Contributing to PyPop}
\label{\detokenize{docs/guide-chapter-contributing:contributing-to-pypop}}\label{\detokenize{docs/guide-chapter-contributing::doc}}
\sphinxAtStartPar
Contributions to PyPop are welcome, and they are greatly appreciated!
Every little bit helps, and credit will always be given.


\section{Reporting and requesting}
\label{\detokenize{docs/guide-chapter-contributing:reporting-and-requesting}}

\subsection{Did you find a bug?}
\label{\detokenize{docs/guide-chapter-contributing:did-you-find-a-bug}}\label{\detokenize{docs/guide-chapter-contributing:guide-contributing-bug-report}}
\sphinxAtStartPar
When \sphinxhref{https://github.com/alexlancaster/pypop/issues}{reporting a bug} (https://github.com/alexlancaster/pypop/issues) please use one of
the provided issue templates if applicable, otherwise just start a
blank issue and describe your situation.
\begin{itemize}
\item {} 
\sphinxAtStartPar
Ensure the bug was not already reported by searching on GitHub under
\sphinxhref{https://github.com/alexlancaster/pypop/issues}{Issues} (https://github.com/alexlancaster/pypop/issues).

\item {} 
\sphinxAtStartPar
If you’re unable to find an open issue addressing the problem, open
a new one. Be sure to include a title and clear description, as much
relevant information as possible, and a code sample or an executable
test case demonstrating the expected behavior that is not occurring.

\item {} 
\sphinxAtStartPar
If possible, use the relevant bug report templates to create the issue.

\item {} 
\sphinxAtStartPar
When reporting bugs, especially during installation, please run the
following and include the output:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nb}{echo}\PYG{+w}{ }\PYG{n+nv}{\PYGZdl{}CPATH}
\PYG{n+nb}{echo}\PYG{+w}{ }\PYG{n+nv}{\PYGZdl{}LIBRARY\PYGZus{}PATH}
\PYG{n+nb}{echo}\PYG{+w}{ }\PYG{n+nv}{\PYGZdl{}PATH}
which\PYG{+w}{ }python
\end{sphinxVerbatim}

\sphinxAtStartPar
If you are running on MacOS, and you used the MacPorts installation
method, please also run and include the output of:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{port} \PYG{n}{installed}
\end{sphinxVerbatim}

\end{itemize}


\subsection{Documentation improvements}
\label{\detokenize{docs/guide-chapter-contributing:documentation-improvements}}
\sphinxAtStartPar
\sphinxstylestrong{pypop} could always use more documentation, whether as part of the
official docs, in docstrings, or even on the web in blog posts,
articles, and such. Write us a \sphinxhref{https://github.com/alexlancaster/pypop/issues/new}{documentation issue} (https://github.com/alexlancaster/pypop/issues/new) describing what
you would like to see improved in here.

\sphinxAtStartPar
If you are able to contribute directly (e.g., via a pull request), please read
our {\hyperref[\detokenize{docs/guide-chapter-contributing:making-a-documentation-or-website-contribution}]{\sphinxcrossref{website contribution guide}}}.


\subsection{Feature requests and feedback}
\label{\detokenize{docs/guide-chapter-contributing:feature-requests-and-feedback}}
\sphinxAtStartPar
The best way to send feedback is to file an issue using the \sphinxhref{https://github.com/alexlancaster/pypop/issues/new?assignees=\&labels=\&projects=\&template=feature\_request.md}{feature
template} (https://github.com/alexlancaster/pypop/issues/new?assignees=\&labels=\&projects=\&template=feature\_request.md).

\sphinxAtStartPar
If you are proposing a feature:
\begin{itemize}
\item {} 
\sphinxAtStartPar
Explain in detail how it would work.

\item {} 
\sphinxAtStartPar
Keep the scope as narrow as possible, to make it easier to implement.

\item {} 
\sphinxAtStartPar
Remember that this is a volunteer\sphinxhyphen{}driven project, and that code contributions are welcome

\end{itemize}


\section{Making a code contribution}
\label{\detokenize{docs/guide-chapter-contributing:making-a-code-contribution}}
\sphinxAtStartPar
To contribute new code that implement a feature, or fix a bug, this
section provides a step\sphinxhyphen{}by\sphinxhyphen{}step guide to getting you set\sphinxhyphen{}up.  The main
steps are:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
forking the repository (or “repo”)

\item {} 
\sphinxAtStartPar
cloning the main repo on to your local machine

\item {} 
\sphinxAtStartPar
making a new branch

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{docs/guide-chapter-contributing:installation-for-developers}]{\sphinxcrossref{installing a development version}}} on your machine

\item {} 
\sphinxAtStartPar
updating your branch when “upstream” (the main repository) has changes to include those changes in your local branch

\item {} 
\sphinxAtStartPar
updating the changelog in \sphinxcode{\sphinxupquote{NEWS.rst}}

\item {} 
\sphinxAtStartPar
checking unit tests pass

\item {} 
\sphinxAtStartPar
making a pull request

\end{enumerate}


\subsection{Fork this repository}
\label{\detokenize{docs/guide-chapter-contributing:fork-this-repository}}
\sphinxAtStartPar
\sphinxhref{https://github.com/alexlancaster/pypop/network/members}{Fork this repository before contributing} (https://github.com/alexlancaster/pypop/network/members). Forks creates a cleaner representation of the \sphinxhref{https://github.com/alexlancaster/pypop/network}{contributions to the
project} (https://github.com/alexlancaster/pypop/network).


\subsection{Clone the main repository}
\label{\detokenize{docs/guide-chapter-contributing:clone-the-main-repository}}
\sphinxAtStartPar
Next, clone the main repository to your local machine:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }clone\PYG{+w}{ }https://github.com/alexlancaster/pypop.git
\PYG{n+nb}{cd}\PYG{+w}{ }pypop
\end{sphinxVerbatim}

\sphinxAtStartPar
Add your fork as an upstream repository:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }remote\PYG{+w}{ }add\PYG{+w}{ }myfork\PYG{+w}{ }git://github.com/YOUR\PYGZhy{}USERNAME/pypop.git
git\PYG{+w}{ }fetch\PYG{+w}{ }myfork
\end{sphinxVerbatim}


\subsection{Make a new branch}
\label{\detokenize{docs/guide-chapter-contributing:make-a-new-branch}}
\sphinxAtStartPar
From the \sphinxcode{\sphinxupquote{main}} branch create a new branch where to develop the new code.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }checkout\PYG{+w}{ }main
git\PYG{+w}{ }checkout\PYG{+w}{ }\PYGZhy{}b\PYG{+w}{ }new\PYGZus{}branch
\end{sphinxVerbatim}

\sphinxAtStartPar
\sphinxstylestrong{Note} the \sphinxcode{\sphinxupquote{main}} branch is from the main repository.


\subsection{Build locally and make your changes}
\label{\detokenize{docs/guide-chapter-contributing:build-locally-and-make-your-changes}}
\sphinxAtStartPar
Now you are ready to make your changes.  First, you need to build
\sphinxcode{\sphinxupquote{pypop}} locally on your machine, and ensure it works, see the
separate section on {\hyperref[\detokenize{docs/guide-chapter-contributing:installation-for-developers}]{\sphinxcrossref{building and installing a development version}}}.

\sphinxAtStartPar
Once you have done the installation and have verified that it works,
you can start to develop the feature, or make the bug fix, and keep
regular pushes to your fork with comprehensible commit messages.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }status
git\PYG{+w}{ }add\PYG{+w}{ }\PYG{c+c1}{\PYGZsh{} (the files you want)}
git\PYG{+w}{ }commit\PYG{+w}{ }\PYG{c+c1}{\PYGZsh{} (add a nice commit message)}
git\PYG{+w}{ }push\PYG{+w}{ }myfork\PYG{+w}{ }new\PYGZus{}branch
\end{sphinxVerbatim}

\sphinxAtStartPar
While you are developing, you can execute \sphinxcode{\sphinxupquote{pytest}} as needed to run
your unit tests. See {\hyperref[\detokenize{docs/guide-chapter-contributing:run-unit-tests-with-pytest}]{\sphinxcrossref{run unit tests with pytest}}}.


\subsection{Keep your branch in sync with upstream}
\label{\detokenize{docs/guide-chapter-contributing:keep-your-branch-in-sync-with-upstream}}
\sphinxAtStartPar
You should keep your branch in sync with the upstream \sphinxcode{\sphinxupquote{main}}
branch. For that:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }checkout\PYG{+w}{ }main\PYG{+w}{  }\PYG{c+c1}{\PYGZsh{} return to the main branch}
git\PYG{+w}{ }pull\PYG{+w}{  }\PYG{c+c1}{\PYGZsh{} retrieve the latest source from the main repository}
git\PYG{+w}{ }checkout\PYG{+w}{ }new\PYGZus{}branch\PYG{+w}{  }\PYG{c+c1}{\PYGZsh{} return to your devel branch}
git\PYG{+w}{ }merge\PYG{+w}{ }\PYGZhy{}\PYGZhy{}no\PYGZhy{}ff\PYG{+w}{ }main\PYG{+w}{  }\PYG{c+c1}{\PYGZsh{} merge the new code to your branch}
\end{sphinxVerbatim}

\sphinxAtStartPar
At this point you may need to solve merge conflicts if they exist. If you don’t
know how to do this, I suggest you start by reading the \sphinxhref{https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github}{official docs} (https://docs.github.com/en/pull\sphinxhyphen{}requests/collaborating\sphinxhyphen{}with\sphinxhyphen{}pull\sphinxhyphen{}requests/addressing\sphinxhyphen{}merge\sphinxhyphen{}conflicts/resolving\sphinxhyphen{}a\sphinxhyphen{}merge\sphinxhyphen{}conflict\sphinxhyphen{}on\sphinxhyphen{}github)

\sphinxAtStartPar
You can push to your fork now if you wish:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }push\PYG{+w}{ }myfork\PYG{+w}{ }new\PYGZus{}branch
\end{sphinxVerbatim}

\sphinxAtStartPar
And, continue doing your developments are previously discussed.


\subsection{Update \sphinxstyleliteralintitle{\sphinxupquote{NEWS.rst}}}
\label{\detokenize{docs/guide-chapter-contributing:update-news-rst}}
\sphinxAtStartPar
Update the changelog file under \sphinxcode{\sphinxupquote{NEWS.rst}} with an explanatory
bullet list of your contribution. Add that list under the “Notes
towards the next release” under the appropriate category, e.g. for a
new feature you would add something like:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
Notes towards next release
\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}
(unreleased)

New features
\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}\PYGZca{}

* here goes my new additions
* explain them shortly and well
\end{sphinxVerbatim}

\sphinxAtStartPar
Also add your name to the authors list at \sphinxcode{\sphinxupquote{website/docs/AUTHORS.rst}}.


\subsection{Run unit tests with \sphinxstyleliteralintitle{\sphinxupquote{pytest}}}
\label{\detokenize{docs/guide-chapter-contributing:run-unit-tests-with-pytest}}
\sphinxAtStartPar
Once you have done your initial installation, you should first check
that the build worked, by running the test suite, via \sphinxcode{\sphinxupquote{pytest}}:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pytest\PYG{+w}{ }tests
\end{sphinxVerbatim}

\sphinxAtStartPar
If \sphinxcode{\sphinxupquote{pytest}} is not already installed, you can install via:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }pytest
\end{sphinxVerbatim}

\sphinxAtStartPar
If you run into errors during your initial installationg, please first
carefully repeat and/or check your installation. If you still get
errors, file a bug, and include the output of \sphinxcode{\sphinxupquote{pytest}} run in
verbose mode and capturing the output

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pytest\PYG{+w}{ }\PYGZhy{}s\PYG{+w}{ }\PYGZhy{}v\PYG{+w}{ }tests
\end{sphinxVerbatim}

\sphinxAtStartPar
You should also continuously run \sphinxcode{\sphinxupquote{pytest}} as you are developing your
code, to ensure that you don’t inadvertently break anything.

\sphinxAtStartPar
Also before creating a Pull Request from your branch, check that all
the tests pass correctly, using the above.

\sphinxAtStartPar
These are exactly the same tests that will be performed online via
Github Actions continuous integration (CI).  This project follows CI
good practices (let us know if something can be improved).


\subsection{Make a Pull Request}
\label{\detokenize{docs/guide-chapter-contributing:make-a-pull-request}}
\sphinxAtStartPar
Once you are finished, you can create a pull request to the main
repository and engage with the developers.  If you need some code
review or feedback while you’re developing the code just make a pull
request.

\sphinxAtStartPar
\sphinxstylestrong{However, before submitting a Pull Request, verify your development branch passes all
tests as} {\hyperref[\detokenize{docs/guide-chapter-contributing:run-unit-tests-with-pytest}]{\sphinxcrossref{described above}}} \sphinxstylestrong{. If you are
developing new code you should also implement new test cases.}

\sphinxAtStartPar
\sphinxstylestrong{Pull Request checklist}

\sphinxAtStartPar
Before requesting a finale merge, you should:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
Make sure your PR passes all \sphinxcode{\sphinxupquote{pytest}} tests.

\item {} 
\sphinxAtStartPar
Add unit tests if you are developing new features

\item {} 
\sphinxAtStartPar
Update documentation when there’s new API, functionality etc.

\item {} 
\sphinxAtStartPar
Add a note to \sphinxcode{\sphinxupquote{NEWS.rst}} about the changes.

\item {} 
\sphinxAtStartPar
Add yourself to \sphinxcode{\sphinxupquote{website/docs/AUTHORS.rst}}.

\end{enumerate}


\section{Installation for developers}
\label{\detokenize{docs/guide-chapter-contributing:installation-for-developers}}
\sphinxAtStartPar
Once you have setup your branch as described in {\hyperref[\detokenize{docs/guide-chapter-contributing:making-a-code-contribution}]{\sphinxcrossref{making a code
contribution}}}, above, you are ready for the four main steps of the
developer installation:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
install a build environment

\item {} 
\sphinxAtStartPar
build

\item {} 
\sphinxAtStartPar
run tests

\end{enumerate}

\begin{sphinxadmonition}{note}{Note:}
\sphinxAtStartPar
Note that you if you need to install PyPop from source, but do not
intend to contribute code, you can skip creating your own forking
and making an additional branch, and clone the main upstream
repository directly:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
git\PYG{+w}{ }clone\PYG{+w}{ }https://github.com/alexlancaster/pypop.git
\PYG{n+nb}{cd}\PYG{+w}{ }pypop
\end{sphinxVerbatim}
\end{sphinxadmonition}

\sphinxAtStartPar
For most developers, we recommend using the miniconda approach
described below.


\subsection{Install the build environment}
\label{\detokenize{docs/guide-chapter-contributing:install-the-build-environment}}
\sphinxAtStartPar
To install the build environment, you should choose either \sphinxcode{\sphinxupquote{conda}} or
system packages. Once you have chosen and installed the build
environment, you should follow the instructions related to the option
you chose here in all subsequent steps.


\subsubsection{Install build environment via miniconda (recommended)}
\label{\detokenize{docs/guide-chapter-contributing:install-build-environment-via-miniconda-recommended}}\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
Visit \sphinxurl{https://docs.conda.io/en/latest/miniconda.html} to download the
miniconda installer for your platform, and follow the instructions to
install.
\begin{quote}

\sphinxAtStartPar
In principle, the rest of the PyPop miniconda installation process
should work on any platform that is supported by miniconda, but
only Linux and MacOS have been tested in standalone mode, at this
time.
\end{quote}

\item {} 
\sphinxAtStartPar
Once miniconda is installed, create a new conda environment, using
the following commands:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
conda\PYG{+w}{ }create\PYG{+w}{ }\PYGZhy{}n\PYG{+w}{ }pypop3\PYG{+w}{ }gsl\PYG{+w}{ }swig\PYG{+w}{ }\PYG{n+nv}{python}\PYG{o}{=}\PYG{l+m}{3}
\end{sphinxVerbatim}

\sphinxAtStartPar
This will download and create a self\sphinxhyphen{}contained build\sphinxhyphen{}environment that
uses of Python to the system\sphinxhyphen{}installed one, along with other
requirements. You will need to use this this environment for the
build, installation and running of PyPop. The conda environment name,
above, \sphinxcode{\sphinxupquote{pypop3}}, can be replaced with your own name.
\begin{quote}

\sphinxAtStartPar
When installing on MacOS, before installing \sphinxcode{\sphinxupquote{conda}}, you should
first to ensure that the Apple Command Line Developer Tools
(XCode) are
\sphinxhref{https://mac.install.guide/commandlinetools/4.html}{installed} (https://mac.install.guide/commandlinetools/4.html),
so you have the compiler (\sphinxcode{\sphinxupquote{clang}}, the drop\sphinxhyphen{}in replacement for
\sphinxcode{\sphinxupquote{gcc}}), \sphinxcode{\sphinxupquote{git}} etc. \sphinxcode{\sphinxupquote{conda}} is unable to include the full
development environment for \sphinxcode{\sphinxupquote{clang}} as a conda package for legal
reasons.
\end{quote}

\item {} 
\sphinxAtStartPar
Activate the environment, and set environments variables needed for
compilation:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
conda\PYG{+w}{ }activate\PYG{+w}{ }pypop3
conda\PYG{+w}{ }env\PYG{+w}{ }config\PYG{+w}{ }vars\PYG{+w}{ }\PYG{n+nb}{set}\PYG{+w}{ }\PYG{n+nv}{CPATH}\PYG{o}{=}\PYG{l+s+si}{\PYGZdl{}\PYGZob{}}\PYG{n+nv}{CONDA\PYGZus{}PREFIX}\PYG{l+s+si}{\PYGZcb{}}/include:\PYG{l+s+si}{\PYGZdl{}\PYGZob{}}\PYG{n+nv}{CPATH}\PYG{l+s+si}{\PYGZcb{}}
conda\PYG{+w}{ }env\PYG{+w}{ }config\PYG{+w}{ }vars\PYG{+w}{ }\PYG{n+nb}{set}\PYG{+w}{ }\PYG{n+nv}{LIBRARY\PYGZus{}PATH}\PYG{o}{=}\PYG{l+s+si}{\PYGZdl{}\PYGZob{}}\PYG{n+nv}{CONDA\PYGZus{}PREFIX}\PYG{l+s+si}{\PYGZcb{}}/lib:\PYG{l+s+si}{\PYGZdl{}\PYGZob{}}\PYG{n+nv}{LIBRARY\PYGZus{}PATH}\PYG{l+s+si}{\PYGZcb{}}
conda\PYG{+w}{ }env\PYG{+w}{ }config\PYG{+w}{ }vars\PYG{+w}{ }\PYG{n+nb}{set}\PYG{+w}{ }\PYG{n+nv}{LD\PYGZus{}LIBRARY\PYGZus{}PATH}\PYG{o}{=}\PYG{l+s+si}{\PYGZdl{}\PYGZob{}}\PYG{n+nv}{CONDA\PYGZus{}PREFIX}\PYG{l+s+si}{\PYGZcb{}}/lib:\PYG{l+s+si}{\PYGZdl{}\PYGZob{}}\PYG{n+nv}{LD\PYGZus{}LIBRARY\PYGZus{}PATH}\PYG{l+s+si}{\PYGZcb{}}
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
To ensure that the environment variables are saved, reactivate the
environment:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
conda\PYG{+w}{ }activate\PYG{+w}{ }pypop3
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
Skip ahead to {\hyperref[\detokenize{docs/guide-chapter-contributing:build-pypop}]{\sphinxcrossref{Build PyPop}}}.

\end{enumerate}


\subsubsection{Install build environment via system packages (advanced)}
\label{\detokenize{docs/guide-chapter-contributing:install-build-environment-via-system-packages-advanced}}

\paragraph{Unix/Linux:}
\label{\detokenize{docs/guide-chapter-contributing:unix-linux}}\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
Ensure Python 3 version of \sphinxcode{\sphinxupquote{pip}} is installed:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
python3\PYG{+w}{ }\PYGZhy{}m\PYG{+w}{ }ensurepip\PYG{+w}{ }\PYGZhy{}\PYGZhy{}user\PYG{+w}{ }\PYGZhy{}\PYGZhy{}no\PYGZhy{}default\PYGZhy{}pip
\end{sphinxVerbatim}
\begin{quote}

\sphinxAtStartPar
Note the use of the \sphinxcode{\sphinxupquote{python3}} \sphinxhyphen{} you may find this to be
necessary on systems which parallel\sphinxhyphen{}install both Python 2 and 3,
which is typically the case. On newer systems you may find that
\sphinxcode{\sphinxupquote{python}} and \sphinxcode{\sphinxupquote{pip}} are, by default, the Python 3 version of
those tools.
\end{quote}

\item {} 
\sphinxAtStartPar
Install packages system\sphinxhyphen{}wide:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumii}{enumiii}{}{.}%
\item {} 
\sphinxAtStartPar
Fedora/Centos/RHEL

\begin{sphinxVerbatim}[commandchars=\\\{\}]
sudo\PYG{+w}{ }dnf\PYG{+w}{ }install\PYG{+w}{ }git\PYG{+w}{ }swig\PYG{+w}{ }gsl\PYGZhy{}devel\PYG{+w}{ }python3\PYGZhy{}devel
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
Ubuntu

\begin{sphinxVerbatim}[commandchars=\\\{\}]
sudo\PYG{+w}{ }apt\PYG{+w}{ }install\PYG{+w}{ }git\PYG{+w}{ }swig\PYG{+w}{ }libgsl\PYGZhy{}dev\PYG{+w}{ }python\PYGZhy{}setuptools
\end{sphinxVerbatim}

\end{enumerate}

\end{enumerate}


\paragraph{MacOS X}
\label{\detokenize{docs/guide-chapter-contributing:macos-x}}\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
Install developer command\sphinxhyphen{}line tools:
\sphinxurl{https://developer.apple.com/downloads/} (includes \sphinxcode{\sphinxupquote{git}}, \sphinxcode{\sphinxupquote{gcc}})

\item {} 
\sphinxAtStartPar
Visit \sphinxurl{http://macports.org} and follow the instructions there to
install the latest version of MacPorts for your version of MacOS X.

\item {} 
\sphinxAtStartPar
Set environment variables to use macports version of Python and other
packages, packages add the following to \sphinxcode{\sphinxupquote{\textasciitilde{}/.bash\_profile}}

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nb}{export}\PYG{+w}{ }\PYG{n+nv}{PATH}\PYG{o}{=}/opt/local/bin:\PYG{n+nv}{\PYGZdl{}PATH}
\PYG{n+nb}{export}\PYG{+w}{ }\PYG{n+nv}{LIBRARY\PYGZus{}PATH}\PYG{o}{=}/opt/local/lib/:\PYG{n+nv}{\PYGZdl{}LIBRARY\PYGZus{}PATH}
\PYG{n+nb}{export}\PYG{+w}{ }\PYG{n+nv}{CPATH}\PYG{o}{=}/opt/local/include:\PYG{n+nv}{\PYGZdl{}CPATH}
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
Rerun your bash shell login in order to make these new exports active
in your environment. At the command line type:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n+nb}{exec}\PYG{+w}{ }bash\PYG{+w}{ }\PYGZhy{}login
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
Install dependencies via MacPorts and set Python version to use
(FIXME: currently untested!)

\begin{sphinxVerbatim}[commandchars=\\\{\}]
sudo\PYG{+w}{ }port\PYG{+w}{ }install\PYG{+w}{ }swig\PYGZhy{}python\PYG{+w}{ }gsl\PYG{+w}{ }py39\PYGZhy{}numpy\PYG{+w}{ }py39\PYGZhy{}lxml\PYG{+w}{ }py39\PYGZhy{}setuptools\PYG{+w}{ }py39\PYGZhy{}pip\PYG{+w}{ }py39\PYGZhy{}pytest
sudo\PYG{+w}{ }port\PYG{+w}{ }\PYG{k}{select}\PYG{+w}{ }\PYGZhy{}\PYGZhy{}set\PYG{+w}{ }python\PYG{+w}{ }python39
sudo\PYG{+w}{ }port\PYG{+w}{ }\PYG{k}{select}\PYG{+w}{ }\PYGZhy{}\PYGZhy{}set\PYG{+w}{ }pip\PYG{+w}{ }pip39
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
Check that the MacPorts version of Python is active by typing:
\sphinxcode{\sphinxupquote{which python}}, if it is working correctly you should see
\sphinxcode{\sphinxupquote{/opt/local/bin/python}}.

\end{enumerate}


\subsubsection{Windows}
\label{\detokenize{docs/guide-chapter-contributing:windows}}
\sphinxAtStartPar
(Currently untested in standalone\sphinxhyphen{}mode)


\subsection{Build PyPop}
\label{\detokenize{docs/guide-chapter-contributing:build-pypop}}
\sphinxAtStartPar
You should choose \sphinxstyleemphasis{either} of the following two approaches. Don’t try
to mix\sphinxhyphen{}and\sphinxhyphen{}match the two. The build\sphinxhyphen{}and\sphinxhyphen{}install approach is only
recommended if don’t plan to make any modifications to the code
locally.


\subsubsection{Build\sphinxhyphen{}and\sphinxhyphen{}install (not recommended for developers)}
\label{\detokenize{docs/guide-chapter-contributing:build-and-install-not-recommended-for-developers}}
\sphinxAtStartPar
Once you have setup your environment and cloned the repo, you can use
the following one\sphinxhyphen{}liner to examine the \sphinxcode{\sphinxupquote{setup.py}} and pull all the
required dependencies from \sphinxcode{\sphinxupquote{pypi.org}} and build and install the
package.
\begin{quote}

\sphinxAtStartPar
Note that if you use this method and install the package, it will be
available to run anywhere on your system, by running \sphinxcode{\sphinxupquote{pypop}}.
\end{quote}
\begin{quote}

\sphinxAtStartPar
If you use this installation method, changes you make to the code,
locally, or via subsequent \sphinxcode{\sphinxupquote{git pull}} requests will not be
available in the installed version until you repeat the
\sphinxcode{\sphinxupquote{pip install}} command.
\end{quote}
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
if you installed the conda development environment, use:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }.\PYG{o}{[}test\PYG{o}{]}
\end{sphinxVerbatim}
\begin{quote}

\sphinxAtStartPar
(the \sphinxcode{\sphinxupquote{{[}test{]}}} keyword is included to make sure that any package
requirements for the test suite are installed as well).
\end{quote}

\item {} 
\sphinxAtStartPar
if you installed a system\sphinxhyphen{}wide environment, the process is slightly
different, because we install into the user’s \sphinxcode{\sphinxupquote{\$HOME/.local}} rather
than the conda environment:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }\PYGZhy{}\PYGZhy{}user\PYG{+w}{ }.\PYG{o}{[}test\PYG{o}{]}
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
PyPop is ready\sphinxhyphen{}to\sphinxhyphen{}use, you should {\hyperref[\detokenize{docs/guide-chapter-contributing:run-unit-tests-with-pytest}]{\sphinxcrossref{run unit tests with pytest}}}.

\item {} 
\sphinxAtStartPar
if you later decide you want to switch to using the developer
approach, below, follow the {\hyperref[\detokenize{docs/guide-chapter-contributing:cleaning-up-build}]{\sphinxcrossref{cleaning up build}}} before
starting.

\end{enumerate}


\subsubsection{Build\sphinxhyphen{}and\sphinxhyphen{}run\sphinxhyphen{}from\sphinxhyphen{}checkout (recommended for developers)}
\label{\detokenize{docs/guide-chapter-contributing:build-and-run-from-checkout-recommended-for-developers}}\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
First manually install the dependencies via \sphinxcode{\sphinxupquote{pip}}, note that if you
are running on Python \textless{}= 3.8, you will need to also add
\sphinxcode{\sphinxupquote{importlib\sphinxhyphen{}resources}} to the list of packages, below.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumii}{enumiii}{}{.}%
\item {} 
\sphinxAtStartPar
conda

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }numpy\PYG{+w}{ }lxml\PYG{+w}{ }psutil\PYG{+w}{ }pytest
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
system\sphinxhyphen{}wide

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }\PYGZhy{}\PYGZhy{}user\PYG{+w}{ }numpy\PYG{+w}{ }lxml\PYG{+w}{ }psutil\PYG{+w}{ }pytest
\end{sphinxVerbatim}

\end{enumerate}

\item {} 
\sphinxAtStartPar
Run the build

\begin{sphinxVerbatim}[commandchars=\\\{\}]
./setup.py\PYG{+w}{ }build
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
You will be runnning PyPop, directly out of the \sphinxcode{\sphinxupquote{src/PyPop}}
subdirectory (e.g. \sphinxcode{\sphinxupquote{./src/PyPop/pypop.py}} and
\sphinxcode{\sphinxupquote{./src/PyPop/popmeta.py}}). Note that you have to include the
\sphinxcode{\sphinxupquote{.py}} extension when you run from an uninstalled checkout,
because the script is not installed.

\end{enumerate}


\subsubsection{Cleaning up build}
\label{\detokenize{docs/guide-chapter-contributing:cleaning-up-build}}
\sphinxAtStartPar
If you installed using the approach in {\hyperref[\detokenize{docs/guide-chapter-contributing:build-and-install-not-recommended-for-developers}]{\sphinxcrossref{Build\sphinxhyphen{}and\sphinxhyphen{}install (not recommended
for developers)}}}, above, follow the end\sphinxhyphen{}user instructions on
{\hyperref[\detokenize{docs/guide-chapter-install:uninstalling-pypop}]{\sphinxcrossref{\DUrole{std,std-ref}{Uninstalling PyPop}}}}.  In addition, to clean\sphinxhyphen{}up any compiled
files and force a recompilation from scratch, run the \sphinxcode{\sphinxupquote{clean}}
command:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
./setup\PYG{+w}{ }clean\PYG{+w}{ }\PYGZhy{}\PYGZhy{}all
\end{sphinxVerbatim}


\section{Making a documentation or website contribution}
\label{\detokenize{docs/guide-chapter-contributing:making-a-documentation-or-website-contribution}}
\sphinxAtStartPar
Interested in maintaining the PyPop website and/or documentation, such
as the \sphinxstyleemphasis{PyPop User Guide}? Here are ways to help.


\subsection{Overview}
\label{\detokenize{docs/guide-chapter-contributing:overview}}
\sphinxAtStartPar
All the documentation (including the website homepage) are maintained in
this directory (and subdirectories) as
\sphinxhref{https://docutils.sourceforge.io/rst.html}{reStructuredText} (https://docutils.sourceforge.io/rst.html)
(\sphinxcode{\sphinxupquote{.rst}}) documents. reStructuredText is very similar to GitHub
markdown (\sphinxcode{\sphinxupquote{.md}}) and should be fairly self\sphinxhyphen{}explanatory to edit
(especially for pure text changes). From the .rst “source” files which
are maintained here on github, we use
\sphinxhref{https://www.sphinx-doc.org/en/master/}{sphinx} (https://www.sphinx\sphinxhyphen{}doc.org/en/master/) to generate (aka
“compile”) the HTML for both the pypop.org user guide and and PDF (via
LaTeX) output. We have setup a GitHub action, so that as soon as a
documentation source file is changed, it will automatically recompile
all the documentation, update the \sphinxcode{\sphinxupquote{gh\sphinxhyphen{}pages}} branch (which is synced
to the GitHub pages) and update the files on the website.

\sphinxAtStartPar
Here’s an overview of the process:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{o}{.}\PYG{n}{rst} \PYG{n}{files} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}} \PYG{n}{sphinx} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}} \PYG{n}{HTML} \PYG{o}{/} \PYG{n}{PDF} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}} \PYG{n}{push} \PYG{n}{to} \PYG{n}{gh}\PYG{o}{\PYGZhy{}}\PYG{n}{pages} \PYG{n}{branch} \PYG{o}{\PYGZhy{}}\PYG{o}{\PYGZgt{}} \PYG{n}{publish} \PYG{n}{on} \PYG{n}{pypop}\PYG{o}{.}\PYG{n}{org}
\end{sphinxVerbatim}

\sphinxAtStartPar
This means that any changes to the source will automatically update both
website home page the documentation.

\sphinxAtStartPar
Once any changes are pushed to a branch (as described below), the GitHub
action will automatically rebuild the website, and the results will be
synced to a “staging” version of the website at:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxurl{https://alexlancaster.github.io/beta.pypop.org/}

\end{itemize}


\subsection{Structure}
\label{\detokenize{docs/guide-chapter-contributing:structure}}
\sphinxAtStartPar
Here’s an overview of the source files for the website/documentation
located in the \sphinxcode{\sphinxupquote{website}} subdirectory at the time of writing.  Note
that some of the documentation and website files, use the
\sphinxcode{\sphinxupquote{include::}} directive to include some “top\sphinxhyphen{}level” files, located
outside \sphinxcode{\sphinxupquote{website}} like \sphinxcode{\sphinxupquote{README.rst}} and \sphinxcode{\sphinxupquote{CONTRIBUTING.rst}}:
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{index.rst}} (this is the source for the homepage at
\sphinxurl{http://pypop.org/})

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{conf.py}} (Sphinx configuration file \sphinxhyphen{} project name and other
global settings are stored here)

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{docs}} (directory containing the source for the \sphinxstyleemphasis{PyPop User Guide}, which will eventually live at \sphinxurl{http://pypop.org/docs}).
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{index.rst}} (source for the top\sphinxhyphen{}level of the \sphinxstyleemphasis{PyPop User Guide})

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{guide\sphinxhyphen{}chapter\sphinxhyphen{}install.rst}} (pulls in parts of the top\sphinxhyphen{}level \sphinxcode{\sphinxupquote{README.rst}})

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{guide\sphinxhyphen{}chapter\sphinxhyphen{}usage.rst}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{guide\sphinxhyphen{}chapter\sphinxhyphen{}instructions.rst}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{guide\sphinxhyphen{}chapter\sphinxhyphen{}contributing.rst}} (pulls in top\sphinxhyphen{}level
\sphinxcode{\sphinxupquote{CONTRIBUTING.rst}} that contains the source of the text that you are reading right now)

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{guide\sphinxhyphen{}chapter\sphinxhyphen{}changes.rst}} (pulls in top\sphinxhyphen{}level \sphinxcode{\sphinxupquote{NEWS.rst}} and \sphinxcode{\sphinxupquote{AUTHORS.rst}}, which is local to \sphinxcode{\sphinxupquote{website}})

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{AUTHORS.rst}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{licenses.rst}} (pulls in top\sphinxhyphen{}level \sphinxcode{\sphinxupquote{LICENSE}})

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{biblio.rst}}

\end{itemize}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{html\_root}} (any files or directories commited in this directory
will appear at the top\sphinxhyphen{}level of the website)
\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{psb\sphinxhyphen{}pypop.pdf}} (e.g. this resides at
\sphinxurl{http://pypop.org/psb-pypop.pdf})

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{tissue\sphinxhyphen{}antigens\sphinxhyphen{}lancaster\sphinxhyphen{}2007.pdf}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{PyPopLinux\sphinxhyphen{}0.7.0.tar.gz}} (old binaries \sphinxhyphen{} will be removed soon)

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{PyPopWin32\sphinxhyphen{}0.7.0.zip}}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{popdata}} (directory \sphinxhyphen{} Suppl. data for Solberg et. al 2018 \sphinxhyphen{}
\sphinxurl{https://pypop.org/popdata/})

\end{itemize}

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{reference}} (directory containing the old DocBook\sphinxhyphen{}based
documentation, preserved to allow for unconverted files to be
converted later, this directory is ignored by the build process)

\end{itemize}


\subsection{Modifying documentation}
\label{\detokenize{docs/guide-chapter-contributing:modifying-documentation}}

\subsubsection{Minor modifications}
\label{\detokenize{docs/guide-chapter-contributing:minor-modifications}}
\sphinxAtStartPar
For small typo fixes, moderate copyedits at the paragraph level
(e.g. adding or modifying paragraphs with little or no embedded markup),
you can make changes directly on the github website.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
navigate to the \sphinxcode{\sphinxupquote{.rst}} file you want to modify in the GitHub code
directory, you’ll see a preview of how most of the \sphinxcode{\sphinxupquote{.rst}} will be
rendered

\item {} 
\sphinxAtStartPar
hover over the edit button \sphinxhyphen{} you’ll see an “\sphinxstylestrong{Edit the file in a
fork in your project}” (if you are already a project collaborator,
you may also have the optional of creating a branch directly in the
main repository).

\item {} 
\sphinxAtStartPar
click it and it will open up a window where you can make your changes

\item {} 
\sphinxAtStartPar
make your edits (it’s a good idea to look at the preview tab
periodically as you make modifications)

\item {} 
\sphinxAtStartPar
once you’ve finished with the modifications, click “\sphinxstylestrong{Commit
changes}”

\item {} 
\sphinxAtStartPar
put in an a commit message, and click “\sphinxstylestrong{Propose changes}”

\item {} 
\sphinxAtStartPar
this will automatically create a new branch in your local fork, and
you can immediately open up a pull\sphinxhyphen{}request by clicking “\sphinxstylestrong{Create pull
request}”

\item {} 
\sphinxAtStartPar
open up a pull\sphinxhyphen{}request and submit \sphinxhyphen{} new documentation will be
automatically built and reviewed. if all is good, it will be merged
by the maintainer and made live on the site.

\end{enumerate}


\subsubsection{Major modifications}
\label{\detokenize{docs/guide-chapter-contributing:major-modifications}}
\sphinxAtStartPar
For larger structural changes involving restructuring documentation or
other major changes across multiple \sphinxcode{\sphinxupquote{.rst}} files, \sphinxstylestrong{it is highly
recommended} that you should make all changes in your own local fork,
by cloning the repository on your computer and then building the
documentation locally. Here’s an overview of how to do that:
\begin{quote}

\sphinxAtStartPar
The commands in the “Sphinx build” section of the workflow
\sphinxhref{https://github.com/alexlancaster/pypop/blob/main/.github/workflows/documentation.yaml}{.github/workflows/documentation.yaml} (https://github.com/alexlancaster/pypop/blob/main/.github/workflows/documentation.yaml)
which are used to run the GitHub Action that builds the documentation
when it it deployed, is the best source for the most update\sphinxhyphen{}to\sphinxhyphen{}date
commands to run, and should be consulted if the instructions in this
document become out of date.
\end{quote}
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
install sphinx and sphinx extensions

\begin{sphinxVerbatim}[commandchars=\\\{\}]
pip\PYG{+w}{ }install\PYG{+w}{ }setuptools\PYGZus{}scm\PYG{+w}{ }sphinx\PYG{+w}{ }piccolo\PYGZhy{}theme\PYG{+w}{ }sphinx\PYGZus{}rtd\PYGZus{}theme\PYG{+w}{ }myst\PYGZus{}parser\PYG{+w}{ }rst2pdf\PYG{+w}{ }sphinx\PYGZus{}togglebutton\PYG{+w}{ }sphinx\PYGZhy{}argparse
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
make a fork of pypop if you haven’t already (see {\hyperref[\detokenize{docs/guide-chapter-contributing:fork-this-repository}]{\sphinxcrossref{previous section}}})

\item {} 
\sphinxAtStartPar
{\hyperref[\detokenize{docs/guide-chapter-contributing:clone-the-main-repository}]{\sphinxcrossref{clone the fork and add your fork as an upstream repository}}} on your local computer, and {\hyperref[\detokenize{docs/guide-chapter-contributing:make-a-new-branch}]{\sphinxcrossref{make a new
branch}}}.

\item {} 
\sphinxAtStartPar
make your changes to your \sphinxcode{\sphinxupquote{.rst}} files and/or \sphinxcode{\sphinxupquote{conf.py}}

\item {} 
\sphinxAtStartPar
build the HTML documentation:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
sphinx\PYGZhy{}build\PYG{+w}{ }website\PYG{+w}{ }\PYGZus{}build
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
view the local documention: you can open up browser and navigate to
the \sphinxcode{\sphinxupquote{index.html}} in the top\sphinxhyphen{}level of the newly\sphinxhyphen{}created \sphinxcode{\sphinxupquote{\_build}}
directory

\item {} 
\sphinxAtStartPar
use \sphinxcode{\sphinxupquote{git commit}} to commit your changes to your local fork.

\item {} 
\sphinxAtStartPar
open up a pull\sphinxhyphen{}request against the upstream repository

\end{enumerate}

\sphinxAtStartPar
Building the PDF for the \sphinxstyleemphasis{PyPop User Guide} is a bit more involved, as
you will need to have various TeX packages installed.
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
install the LaTeX packages (these are packages needed for Ubuntu,
they may be different on your distribution):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
sudo\PYG{+w}{ }apt\PYGZhy{}get\PYG{+w}{ }install\PYG{+w}{ }\PYGZhy{}y\PYG{+w}{ }latexmk\PYG{+w}{ }texlive\PYGZhy{}latex\PYGZhy{}recommended\PYG{+w}{ }texlive\PYGZhy{}latex\PYGZhy{}extra\PYG{+w}{ }texlive\PYGZhy{}fonts\PYGZhy{}recommended\PYG{+w}{ }texlive\PYGZhy{}fonts\PYGZhy{}extra\PYG{+w}{ }texlive\PYGZhy{}luatex\PYG{+w}{ }texlive\PYGZhy{}xetex
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
build the LaTeX and then compile the PDF:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
sphinx\PYGZhy{}build\PYG{+w}{ }\PYGZhy{}b\PYG{+w}{ }latex\PYG{+w}{ }website\PYG{+w}{ }\PYGZus{}latexbuild
make\PYG{+w}{ }\PYGZhy{}C\PYG{+w}{ }\PYGZus{}latexbuild
\end{sphinxVerbatim}

\item {} 
\sphinxAtStartPar
the user guide will be generated in \sphinxcode{\sphinxupquote{\_latexbuild/pypop\sphinxhyphen{}guide.pdf}}

\end{enumerate}

\sphinxstepscope


\chapter{Authors and history}
\label{\detokenize{docs/guide-chapter-changes:authors-and-history}}\label{\detokenize{docs/guide-chapter-changes:guide-changes-authors}}\label{\detokenize{docs/guide-chapter-changes::doc}}

\section{Authors of software components}
\label{\detokenize{docs/guide-chapter-changes:authors-of-software-components}}\label{\detokenize{docs/guide-chapter-changes:guide-preface-authors}}\begin{description}
\sphinxlineitem{Alex Lancaster}
\sphinxAtStartPar
Co\sphinxhyphen{}designer of Python framework: author of main engine, text file
parser, Python extension module framework using SWIG, XML output and
XSLT post\sphinxhyphen{}processing framework (to generate plain text and HTML
output).

\sphinxlineitem{Mark P. Nelson}
\sphinxAtStartPar
Co\sphinxhyphen{}designer of Python framework: implemented and maintained Python
modules, particularly the module for Hardy\sphinxhyphen{}Weinberg analysis. Updated
and maintained XSLT code.

\sphinxlineitem{Richard M. Single}
\sphinxAtStartPar
Author of haplotype frequency and linkage disequilibrium analysis
module “emhaplofreq”, author of R programs to do further statistical
analysis and generate graphs and figures in PostScript.

\sphinxlineitem{Diogo Meyer}
\sphinxAtStartPar
Contributed further statistical analysis code for the R programs.

\sphinxlineitem{Owen Solberg}
\sphinxAtStartPar
Implemented filter modules, including conversion to allele name
information to sequence data.

\sphinxlineitem{Yingssu Tsai}
\sphinxAtStartPar
Implemented prototype of the allele names to sequence conversion
filter module.

\sphinxlineitem{Glenys Thomson}
\sphinxAtStartPar
Principal investigator and project lead.

\sphinxlineitem{gthwe}
\sphinxAtStartPar
The Hardy\sphinxhyphen{}Weinberg “exact test” implementation is a modified version
of Guo \& Thompson’s Guo:Thompson:1992 code. Dr. Sun\sphinxhyphen{}Wei Guo has
kindly allowed us to release the code under the \sphinxhref{http://www.gnu.org/licenses/gpl.html}{GNU General Public
License} (http://www.gnu.org/licenses/gpl.html). Original code
available at
\sphinxurl{http://www.stat.washington.edu/thompson/Genepi/Hardy.shtml}.

\sphinxlineitem{\sphinxcode{\sphinxupquote{slatkin\sphinxhyphen{}exact/monte\sphinxhyphen{}carlo.c}}}
\sphinxAtStartPar
Montgomery Slatkin’s implementation of a Monte Carlo approximation of
the Ewens\sphinxhyphen{}Watterson exact test of neutrality (Slatkin:1994,
Slatkin:1996). Original code can be found at:
http://ib.berkeley.edu/labs/slatkin/monty/Ewens\_exact.program.

\sphinxlineitem{\sphinxcode{\sphinxupquote{pval}}}
\sphinxAtStartPar
The code in the ‘\sphinxcode{\sphinxupquote{pval}}’ directory (with the exception of
‘\sphinxcode{\sphinxupquote{pval.c}}’ the SWIG wrapper, \sphinxcode{\sphinxupquote{\textquotesingle{}pval\_wrap.i}}’ and the Makefile) is
part of the R project’s ‘\sphinxcode{\sphinxupquote{nmath}}’ numerical library
\sphinxurl{http://www.r-project.org/} and is also licensed under the GNU General
Public License (GPL). Minor modifications have been made to allow the
module to build correctly.

\end{description}


\section{PyPop Release History}
\label{\detokenize{docs/guide-chapter-changes:pypop-release-history}}

\subsection{Notes towards next release}
\label{\detokenize{docs/guide-chapter-changes:notes-towards-next-release}}\label{\detokenize{docs/guide-chapter-changes:news-start}}
\sphinxAtStartPar
(unreleased)


\subsubsection{New features}
\label{\detokenize{docs/guide-chapter-changes:new-features}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Updated to Python 3

\item {} 
\sphinxAtStartPar
Implement new assymetric LD (ALD) measure

\item {} 
\sphinxAtStartPar
New wrapper module \sphinxcode{\sphinxupquote{Haplostats}}. This wraps a portion of the
\sphinxcode{\sphinxupquote{haplo.stats}} R package \sphinxcode{\sphinxupquote{haplo\sphinxhyphen{}stats}} for haplotype
estimation. {[}Implementation in alpha\sphinxhyphen{}phase{]}.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{popmeta}}: now accepts the \sphinxcode{\sphinxupquote{\sphinxhyphen{}o}}/\sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}outputdir}} option for saving
generated files.

\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{pypop}}: renamed \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}generate\sphinxhyphen{}tsv}} to \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}tsv}}

\end{itemize}


\subsection{Release Notes for PyPop 0.7.0}
\label{\detokenize{docs/guide-chapter-changes:release-notes-for-pypop-0-7-0}}
\sphinxAtStartPar
(2008\sphinxhyphen{}09\sphinxhyphen{}09)


\subsubsection{New features}
\label{\detokenize{docs/guide-chapter-changes:id1}}\begin{itemize}
\item {} 
\sphinxAtStartPar
\sphinxcode{\sphinxupquote{makeNewPopFile}} option has been changed.  This option allows user to
generate intermediate output of filtered files. Now option should
be of the format: \sphinxcode{\sphinxupquote{type:order}} where \sphinxcode{\sphinxupquote{type}} is one of
\sphinxcode{\sphinxupquote{separate\sphinxhyphen{}loci}} or \sphinxcode{\sphinxupquote{all\sphinxhyphen{}loci}} so that the user can specify whether
a separate file should be generated for each locus
(\sphinxcode{\sphinxupquote{separate\sphinxhyphen{}loci}}) or a single file with all loci (\sphinxcode{\sphinxupquote{all\sphinxhyphen{}loci}}).
\sphinxcode{\sphinxupquote{order}} should be the order in the filtering chain where the
matrix is generated, there is no default, for example, for
generating files after the first filter operation use \sphinxcode{\sphinxupquote{1}}.

\item {} 
\sphinxAtStartPar
New command\sphinxhyphen{}line option \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}generate\sphinxhyphen{}tsv}}, will generate the \sphinxcode{\sphinxupquote{.dat}}
tab\sphinxhyphen{}separated values (TSV) files on the the generated \sphinxhyphen{}out.xml
files (aka “popmeta”) directly from pypop without needing to run
additional script.  Now output from pypop can be directly read
into spreadsheet.

\item {} 
\sphinxAtStartPar
New feature: add individual genotype tests to Hardy\sphinxhyphen{}Weinberg module
(gthwe), now computes statistics based on individual genotypes in
the HWP table. The \sphinxcode{\sphinxupquote{{[}HardyWeinbergGuoThompson{]}}} or
\sphinxcode{\sphinxupquote{{[}HardyWeinbergGuoThompsonMonteCarlo{]}}} options must be enabled in the
configuration “.ini” file in order for these tests to be carried out.

\item {} 
\sphinxAtStartPar
Major improvements to custom and random binning filters (Owen Solberg).

\item {} 
\sphinxAtStartPar
New feature: generate homozygosity values using the Ewens\sphinxhyphen{}Watterson test for
all pairwise loci, or all sites within a gene for sequence data
(\sphinxcode{\sphinxupquote{{[}homozygosityEWSlatkinExactPairwise{]}}} in .ini file).  Note: this
really only works for sequence data where the phase for sites
within an allele are known.

\item {} 
\sphinxAtStartPar
Haplotype and LD estimation module \sphinxcode{\sphinxupquote{emhaplofreq}} improvements
\begin{itemize}
\item {} 
\sphinxAtStartPar
improved memory usage and speed for emhaplofreq module.

\item {} 
\sphinxAtStartPar
maximum sample size for emhaplofreq module increased from 1023 to
5000 individuals.

\item {} 
\sphinxAtStartPar
maximum length of allele names increased to 20

\end{itemize}

\end{itemize}


\subsubsection{Bug fixes}
\label{\detokenize{docs/guide-chapter-changes:bug-fixes}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Support Python 2.4 on GCC 4.0 platforms.

\item {} 
\sphinxAtStartPar
Add missing initialisation for non\sphinxhyphen{}sequence data when processing
haplotypes.  Thanks to Jill Hollenbach for the report.

\item {} 
\sphinxAtStartPar
Fix memory leak in xslt translation.

\item {} 
\sphinxAtStartPar
Various fixes relating to parsing XML output.

\item {} 
\sphinxAtStartPar
Fixed an incorrect parameter name.

\item {} 
\sphinxAtStartPar
Handle some missing sections in .ini better. Thanks to
Owen Solberg for report.

\item {} 
\sphinxAtStartPar
Various build and installation fixes (SWIG, compilation flags)

\item {} 
\sphinxAtStartPar
Make name of source package be lowercase “pypop”.

\item {} 
\sphinxAtStartPar
Change data directory: /usr/share/pypop/ to /usr/share/PyPop/

\item {} 
\sphinxAtStartPar
Print out warning when maximum length of allele exceeded, rather than
crashing.  Thanks to Steve Mack for report.

\end{itemize}


\subsubsection{Other issues}
\label{\detokenize{docs/guide-chapter-changes:other-issues}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Sequence filter
\begin{itemize}
\item {} 
\sphinxAtStartPar
In the Sequence filter, add special case for Anthony Nolan HLA data:
mark null alleles ending in “N” (e.g. HLA\sphinxhyphen{}B*5127N) as “missing
data” (\sphinxcode{\sphinxupquote{****}}).

\item {} 
\sphinxAtStartPar
Also in Sequence, keep track of unsequenced sites separately
(via unsequencedSites variable) from “untyped” (aka “missing
data”). Treat unsequencedSite as a unique allele to make sure that
those sites don’t get treated as having a consensus sequence if
only one of the sequences in the the set of matches is typed.

\item {} 
\sphinxAtStartPar
If no matching sequence is found in the MSF files, then return a
sequence of * symbols (ie, will be treated as truly missing data,
not untyped alleles.

\item {} 
\sphinxAtStartPar
Add another special case for HLA data: test for 7 digits in allele names
(e.g. if 2402101 is not found insert a zero after the first 4
digits to form 24020101, and check for that).  This is to cope
with yet\sphinxhyphen{}another HLA nomenclature change.

\end{itemize}

\item {} 
\sphinxAtStartPar
Change semantics of batchsize, make “0” (default) process files separately
if only R dat files is enabled.  If batchsize not set explicitly
(and therefore 0) set batchsize to \sphinxcode{\sphinxupquote{1}} is PHYLIP mode is enabled.

\end{itemize}


\subsection{Release Notes for PyPop 0.6.0}
\label{\detokenize{docs/guide-chapter-changes:release-notes-for-pypop-0-6-0}}
\sphinxAtStartPar
(2005\sphinxhyphen{}04\sphinxhyphen{}13)


\subsubsection{New features}
\label{\detokenize{docs/guide-chapter-changes:id2}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Allow for odd allele counts when processing an allele count data
(i.e “semi”\sphinxhyphen{}typing).  When PyPop is dealing with data that is
originally genotyped, the current default is preserved i.e.  we
dis\sphinxhyphen{}allow individuals that are typed at only allele, and set
allowSemiTyped to false.

\item {} 
\sphinxAtStartPar
New command\sphinxhyphen{}line option \sphinxcode{\sphinxupquote{\sphinxhyphen{}f}} (long version \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}filelist}}) which
accepts a file containing a list of files (one per line) to
process (note that this is mutually exclusive with supplying
INPUTFILEs, and will abort with an error message if you supply
both simultaneously).

\item {} 
\sphinxAtStartPar
In batch version, handle multiple INPUTFILEs supplied as command\sphinxhyphen{}line
arguments and support Unix shell\sphinxhyphen{}globbing syntax (e.g. \sphinxcode{\sphinxupquote{pypop.py
\sphinxhyphen{}c config.ini *.pop}}). (NOTE: This is supported \sphinxstyleemphasis{only} in
batch version, not in the interactive version, which expects one
and only one file supplied by user.

\item {} 
\sphinxAtStartPar
Allele count files can now be filtered through the filter apparatus
(particularly the Sequence and AnthonyNolan) in the same was as
genotype files transparently.  {[}This has been enabled via a code
refactor that treats allele count files as pseudo\sphinxhyphen{}genotype files
for the purpose of filtering{]}.  This change also resulted in the
removal of the obsolete lookup\sphinxhyphen{}table\sphinxhyphen{}based homozygosity test.

\item {} 
\sphinxAtStartPar
Add \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}disable\sphinxhyphen{}ihwg}} option to popmeta script to disable hardcoded
generation of the IHWG header output, and use the output as
defined in the header in the original .pop input text file.  This
is disabled by default to preserve backwards compatibility.

\item {} 
\sphinxAtStartPar
Add \sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}batchsize}} (\sphinxcode{\sphinxupquote{\sphinxhyphen{}b}} short version) option  for popmeta.  Does the
processing in “batches”.  If set and greater than one, list of XML
files is split into batchsize group.  For example, if there are 20
XML files and option is via using (“\sphinxhyphen{}b 2” or “\textendash{}batchsize=2”) then
the files will be processed in two batches, each consisting of 10
files.  If the number does not divide evenly, the last list will
contain all the “left\sphinxhyphen{}over” files.  This option is particularly
useful with large XML files that may not fit in memory all at
once.  Note this option is mutually exclusive with the
\sphinxcode{\sphinxupquote{\sphinxhyphen{}\sphinxhyphen{}enable\sphinxhyphen{}PHYLIP}} option because the PHYLIP output needs to
calculate allele frequencies across all populations before
generating files.

\item {} 
\sphinxAtStartPar
New .ini file option: \sphinxcode{\sphinxupquote{{[}HardyWeinbergGuoThompsonMonteCarlo{]}}}: add a plain
Monte\sphinxhyphen{}Carlo (randomization, without the Markov chain test) test
for the HardyWeinberg “exact test”.  Add code for Guo \& Thompson
test to distribution (now under GNU GPL).

\end{itemize}


\subsubsection{Bug fixes}
\label{\detokenize{docs/guide-chapter-changes:id3}}\begin{itemize}
\item {} 
\sphinxAtStartPar
HardyWeinbergGuoThompson overall p\sphinxhyphen{}value test was numerically unstable
because it attempted to check for equality in greater than or
equal to constructs (“\textless{}=”) which is not reliable in C.  Replaced
this with a GNU Scientific Library (GSL) function gsl\_fcmp() which
compares floats to within an EPSILON (defaults to 1e\sphinxhyphen{}6).

\item {} 
\sphinxAtStartPar
Allow \sphinxcode{\sphinxupquote{HardyWeinbergGuoThompson\textasciigrave{} test to be run if at least two alleles
present (test was originally failing with a \textasciigrave{}\textasciigrave{}too\sphinxhyphen{}few\sphinxhyphen{}alleles}}
message if there were not at least 3 alleles).  Thanks to Kristie
Mather for the report.

\item {} 
\sphinxAtStartPar
Checks to see if a locus is monomorphic, if it is, it generates an
allele summary report, but skips the rest of the single locus
analyses which do not make sense for monomorphic locus.  Thanks to
Steve Mack and Owen Solberg for the bug report(s).

\item {} 
\sphinxAtStartPar
Now builds against recent versions of SWIG (no longer stuck at version
1.3.9), should be compatible with versions of SWIG \textgreater{} 1.3.10.
(Tested against SWIG 1.3.21).

\item {} 
\sphinxAtStartPar
Homozygosity module: Prevent math errors by in Slatkin’s exact test by
forcing the homozygosity to be positive (only a problem for rare
cases, when the result is so close to zero that the floating point
algorithms cause a negative result.)

\end{itemize}


\subsection{Release Notes for PyPop 0.5.2 (public beta)}
\label{\detokenize{docs/guide-chapter-changes:release-notes-for-pypop-0-5-2-public-beta}}
\sphinxAtStartPar
(2004\sphinxhyphen{}03\sphinxhyphen{}09)


\subsubsection{Bug fixes}
\label{\detokenize{docs/guide-chapter-changes:id4}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Add missing RandomBinning.py file to source distribution
Thanks to Hazael Maldonado Torres for the bug report.

\item {} 
\sphinxAtStartPar
Fixed line endings for .bat scripts for Win32 so they work under
Windows 98 thanks to Wendy Hartogensis for the bug report.

\end{itemize}


\subsection{Release Notes for PyPop 0.5.1 (public beta)}
\label{\detokenize{docs/guide-chapter-changes:release-notes-for-pypop-0-5-1-public-beta}}
\sphinxAtStartPar
(2004\sphinxhyphen{}02\sphinxhyphen{}26)


\subsubsection{Changes}
\label{\detokenize{docs/guide-chapter-changes:changes}}\begin{itemize}
\item {} 
\sphinxAtStartPar
New parameter \sphinxcode{\sphinxupquote{numInitCond}}, number of initial conditions by the
haplotype estimation and LD algorithm used before performing
permutations. Defaults to 50.

\item {} 
\sphinxAtStartPar
Remove some LOG messages/diagnostics that were erroneously implying
an error to the user (if nothing is wrong, don’t say anything).  Add
some more useful messages for what is being done in haplo/LD
estimation step.

\item {} 
\sphinxAtStartPar
Add popmeta.py to the distribution: this is undocumented and unsupported
as yet, it is at alpha stage only, use at your own risk!

\end{itemize}


\subsubsection{Bug fixes}
\label{\detokenize{docs/guide-chapter-changes:id5}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Remember to output plaintext version of LD for specified loci.

\end{itemize}


\subsection{Release Notes for PyPop 0.5 (public beta)}
\label{\detokenize{docs/guide-chapter-changes:release-notes-for-pypop-0-5-public-beta}}
\sphinxAtStartPar
(2003\sphinxhyphen{}12\sphinxhyphen{}31)


\subsubsection{Changes}
\label{\detokenize{docs/guide-chapter-changes:id6}}\begin{itemize}
\item {} 
\sphinxAtStartPar
All Linux wrapper scripts no longer have .sh file suffixes for
consistency with DOS (all DOS bat files can be executed without
specifying the .bat extension).

\end{itemize}


\subsubsection{Bug fixes}
\label{\detokenize{docs/guide-chapter-changes:id7}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Add wrapper scripts for interactive and batch mode for
both DOS and Linux so that correct shared libraries are called.

\item {} 
\sphinxAtStartPar
Pause and wait for user to press a key at end of DOS .bat file
so that output can be viewed before window close.

\item {} 
\sphinxAtStartPar
Set PYTHONHOME in wrapper scripts to prevent messages about
missing \textless{}prefix\textgreater{} being displayed.

\end{itemize}


\subsection{Release Notes for PyPop 0.4.3beta}
\label{\detokenize{docs/guide-chapter-changes:release-notes-for-pypop-0-4-3beta}}

\subsubsection{Bug fixes}
\label{\detokenize{docs/guide-chapter-changes:id8}}\begin{itemize}
\item {} 
\sphinxAtStartPar
Fixed bug in processing of \sphinxcode{\sphinxupquote{popname}} field.
Thanks to Richard Single for the report.

\end{itemize}

\sphinxstepscope


\chapter{Licenses}
\label{\detokenize{docs/licenses:licenses}}\label{\detokenize{docs/licenses:license}}\label{\detokenize{docs/licenses::doc}}
\begingroup
\footnotesize
\sphinxsetup{%
      %TitleColor={named}{blue},
}


\section{License terms for PyPop}
\label{\detokenize{docs/licenses:license-terms-for-pypop}}
\sphinxAtStartPar
Except where noted in the source, PyPop is distributed under the terms
of the \sphinxhref{http://www.gnu.org/licenses/gpl.html}{GNU General Public License} (http://www.gnu.org/licenses/gpl.html) ({\hyperref[\detokenize{docs/licenses:gpl}]{\sphinxcrossref{\DUrole{std,std-ref}{GNU General Public License}}}}). The list of authors and
contributors is located in {\hyperref[\detokenize{docs/guide-chapter-changes:guide-preface-authors}]{\sphinxcrossref{\DUrole{std,std-ref,std,std-ref}{Authors of software components}}}}.

\sphinxAtStartPar
PyPop: Python for Population Genomics

\sphinxAtStartPar
Copyright © 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009 by The
Regents of the University of California (Regents). All rights reserved.

\sphinxAtStartPar
All code in this program is copyright the Regents (except where
otherwise explicitly mentioned).

\sphinxAtStartPar
This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

\sphinxAtStartPar
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
Public License ({\hyperref[\detokenize{docs/licenses:gpl}]{\sphinxcrossref{\DUrole{std,std-ref}{GNU General Public License}}}}) for more details.


\subsection{GNU General Public License}
\label{\detokenize{docs/licenses:gpl}}\label{\detokenize{docs/licenses:id1}}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
                    GNU GENERAL PUBLIC LICENSE
                       Version 2, June 1991

 Copyright (C) 1989, 1991 Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110\PYGZhy{}1301 USA
 Everyone is permitted to copy and distribute verbatim copies
 of this license document, but changing it is not allowed.

                            Preamble

  The licenses for most software are designed to take away your
freedom to share and change it.  By contrast, the GNU General Public
License is intended to guarantee your freedom to share and change free
software\PYGZhy{}\PYGZhy{}to make sure the software is free for all its users.  This
General Public License applies to most of the Free Software
Foundation\PYGZsq{}s software and to any other program whose authors commit to
using it.  (Some other Free Software Foundation software is covered by
the GNU Lesser General Public License instead.)  You can apply it to
your programs, too.

  When we speak of free software, we are referring to freedom, not
price.  Our General Public Licenses are designed to make sure that you
have the freedom to distribute copies of free software (and charge for
this service if you wish), that you receive source code or can get it
if you want it, that you can change the software or use pieces of it
in new free programs; and that you know you can do these things.

  To protect your rights, we need to make restrictions that forbid
anyone to deny you these rights or to ask you to surrender the rights.
These restrictions translate to certain responsibilities for you if you
distribute copies of the software, or if you modify it.

  For example, if you distribute copies of such a program, whether
gratis or for a fee, you must give the recipients all the rights that
you have.  You must make sure that they, too, receive or can get the
source code.  And you must show them these terms so they know their
rights.

  We protect your rights with two steps: (1) copyright the software, and
(2) offer you this license which gives you legal permission to copy,
distribute and/or modify the software.

  Also, for each author\PYGZsq{}s protection and ours, we want to make certain
that everyone understands that there is no warranty for this free
software.  If the software is modified by someone else and passed on, we
want its recipients to know that what they have is not the original, so
that any problems introduced by others will not reflect on the original
authors\PYGZsq{} reputations.

  Finally, any free program is threatened constantly by software
patents.  We wish to avoid the danger that redistributors of a free
program will individually obtain patent licenses, in effect making the
program proprietary.  To prevent this, we have made it clear that any
patent must be licensed for everyone\PYGZsq{}s free use or not licensed at all.

  The precise terms and conditions for copying, distribution and
modification follow.

                    GNU GENERAL PUBLIC LICENSE
   TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION

  0. This License applies to any program or other work which contains
a notice placed by the copyright holder saying it may be distributed
under the terms of this General Public License.  The \PYGZdq{}Program\PYGZdq{}, below,
refers to any such program or work, and a \PYGZdq{}work based on the Program\PYGZdq{}
means either the Program or any derivative work under copyright law:
that is to say, a work containing the Program or a portion of it,
either verbatim or with modifications and/or translated into another
language.  (Hereinafter, translation is included without limitation in
the term \PYGZdq{}modification\PYGZdq{}.)  Each licensee is addressed as \PYGZdq{}you\PYGZdq{}.

Activities other than copying, distribution and modification are not
covered by this License; they are outside its scope.  The act of
running the Program is not restricted, and the output from the Program
is covered only if its contents constitute a work based on the
Program (independent of having been made by running the Program).
Whether that is true depends on what the Program does.

  1. You may copy and distribute verbatim copies of the Program\PYGZsq{}s
source code as you receive it, in any medium, provided that you
conspicuously and appropriately publish on each copy an appropriate
copyright notice and disclaimer of warranty; keep intact all the
notices that refer to this License and to the absence of any warranty;
and give any other recipients of the Program a copy of this License
along with the Program.

You may charge a fee for the physical act of transferring a copy, and
you may at your option offer warranty protection in exchange for a fee.

  2. You may modify your copy or copies of the Program or any portion
of it, thus forming a work based on the Program, and copy and
distribute such modifications or work under the terms of Section 1
above, provided that you also meet all of these conditions:

    a) You must cause the modified files to carry prominent notices
    stating that you changed the files and the date of any change.

    b) You must cause any work that you distribute or publish, that in
    whole or in part contains or is derived from the Program or any
    part thereof, to be licensed as a whole at no charge to all third
    parties under the terms of this License.

    c) If the modified program normally reads commands interactively
    when run, you must cause it, when started running for such
    interactive use in the most ordinary way, to print or display an
    announcement including an appropriate copyright notice and a
    notice that there is no warranty (or else, saying that you provide
    a warranty) and that users may redistribute the program under
    these conditions, and telling the user how to view a copy of this
    License.  (Exception: if the Program itself is interactive but
    does not normally print such an announcement, your work based on
    the Program is not required to print an announcement.)

These requirements apply to the modified work as a whole.  If
identifiable sections of that work are not derived from the Program,
and can be reasonably considered independent and separate works in
themselves, then this License, and its terms, do not apply to those
sections when you distribute them as separate works.  But when you
distribute the same sections as part of a whole which is a work based
on the Program, the distribution of the whole must be on the terms of
this License, whose permissions for other licensees extend to the
entire whole, and thus to each and every part regardless of who wrote it.

Thus, it is not the intent of this section to claim rights or contest
your rights to work written entirely by you; rather, the intent is to
exercise the right to control the distribution of derivative or
collective works based on the Program.

In addition, mere aggregation of another work not based on the Program
with the Program (or with a work based on the Program) on a volume of
a storage or distribution medium does not bring the other work under
the scope of this License.

  3. You may copy and distribute the Program (or a work based on it,
under Section 2) in object code or executable form under the terms of
Sections 1 and 2 above provided that you also do one of the following:

    a) Accompany it with the complete corresponding machine\PYGZhy{}readable
    source code, which must be distributed under the terms of Sections
    1 and 2 above on a medium customarily used for software interchange; or,

    b) Accompany it with a written offer, valid for at least three
    years, to give any third party, for a charge no more than your
    cost of physically performing source distribution, a complete
    machine\PYGZhy{}readable copy of the corresponding source code, to be
    distributed under the terms of Sections 1 and 2 above on a medium
    customarily used for software interchange; or,

    c) Accompany it with the information you received as to the offer
    to distribute corresponding source code.  (This alternative is
    allowed only for noncommercial distribution and only if you
    received the program in object code or executable form with such
    an offer, in accord with Subsection b above.)

The source code for a work means the preferred form of the work for
making modifications to it.  For an executable work, complete source
code means all the source code for all modules it contains, plus any
associated interface definition files, plus the scripts used to
control compilation and installation of the executable.  However, as a
special exception, the source code distributed need not include
anything that is normally distributed (in either source or binary
form) with the major components (compiler, kernel, and so on) of the
operating system on which the executable runs, unless that component
itself accompanies the executable.

If distribution of executable or object code is made by offering
access to copy from a designated place, then offering equivalent
access to copy the source code from the same place counts as
distribution of the source code, even though third parties are not
compelled to copy the source along with the object code.

  4. You may not copy, modify, sublicense, or distribute the Program
except as expressly provided under this License.  Any attempt
otherwise to copy, modify, sublicense or distribute the Program is
void, and will automatically terminate your rights under this License.
However, parties who have received copies, or rights, from you under
this License will not have their licenses terminated so long as such
parties remain in full compliance.

  5. You are not required to accept this License, since you have not
signed it.  However, nothing else grants you permission to modify or
distribute the Program or its derivative works.  These actions are
prohibited by law if you do not accept this License.  Therefore, by
modifying or distributing the Program (or any work based on the
Program), you indicate your acceptance of this License to do so, and
all its terms and conditions for copying, distributing or modifying
the Program or works based on it.

  6. Each time you redistribute the Program (or any work based on the
Program), the recipient automatically receives a license from the
original licensor to copy, distribute or modify the Program subject to
these terms and conditions.  You may not impose any further
restrictions on the recipients\PYGZsq{} exercise of the rights granted herein.
You are not responsible for enforcing compliance by third parties to
this License.

  7. If, as a consequence of a court judgment or allegation of patent
infringement or for any other reason (not limited to patent issues),
conditions are imposed on you (whether by court order, agreement or
otherwise) that contradict the conditions of this License, they do not
excuse you from the conditions of this License.  If you cannot
distribute so as to satisfy simultaneously your obligations under this
License and any other pertinent obligations, then as a consequence you
may not distribute the Program at all.  For example, if a patent
license would not permit royalty\PYGZhy{}free redistribution of the Program by
all those who receive copies directly or indirectly through you, then
the only way you could satisfy both it and this License would be to
refrain entirely from distribution of the Program.

If any portion of this section is held invalid or unenforceable under
any particular circumstance, the balance of the section is intended to
apply and the section as a whole is intended to apply in other
circumstances.

It is not the purpose of this section to induce you to infringe any
patents or other property right claims or to contest validity of any
such claims; this section has the sole purpose of protecting the
integrity of the free software distribution system, which is
implemented by public license practices.  Many people have made
generous contributions to the wide range of software distributed
through that system in reliance on consistent application of that
system; it is up to the author/donor to decide if he or she is willing
to distribute software through any other system and a licensee cannot
impose that choice.

This section is intended to make thoroughly clear what is believed to
be a consequence of the rest of this License.

  8. If the distribution and/or use of the Program is restricted in
certain countries either by patents or by copyrighted interfaces, the
original copyright holder who places the Program under this License
may add an explicit geographical distribution limitation excluding
those countries, so that distribution is permitted only in or among
countries not thus excluded.  In such case, this License incorporates
the limitation as if written in the body of this License.

  9. The Free Software Foundation may publish revised and/or new versions
of the General Public License from time to time.  Such new versions will
be similar in spirit to the present version, but may differ in detail to
address new problems or concerns.

Each version is given a distinguishing version number.  If the Program
specifies a version number of this License which applies to it and \PYGZdq{}any
later version\PYGZdq{}, you have the option of following the terms and conditions
either of that version or of any later version published by the Free
Software Foundation.  If the Program does not specify a version number of
this License, you may choose any version ever published by the Free Software
Foundation.

  10. If you wish to incorporate parts of the Program into other free
programs whose distribution conditions are different, write to the author
to ask for permission.  For software which is copyrighted by the Free
Software Foundation, write to the Free Software Foundation; we sometimes
make exceptions for this.  Our decision will be guided by the two goals
of preserving the free status of all derivatives of our free software and
of promoting the sharing and reuse of software generally.

                            NO WARRANTY

  11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW.  EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE PROGRAM \PYGZdq{}AS IS\PYGZdq{} WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED
OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  THE ENTIRE RISK AS
TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU.  SHOULD THE
PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING,
REPAIR OR CORRECTION.

  12. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED
TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY
YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER
PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

                     END OF TERMS AND CONDITIONS

            How to Apply These Terms to Your New Programs

  If you develop a new program, and you want it to be of the greatest
possible use to the public, the best way to achieve this is to make it
free software which everyone can redistribute and change under these terms.

  To do so, attach the following notices to the program.  It is safest
to attach them to the start of each source file to most effectively
convey the exclusion of warranty; and each file should have at least
the \PYGZdq{}copyright\PYGZdq{} line and a pointer to where the full notice is found.

    \PYGZlt{}one line to give the program\PYGZsq{}s name and a brief idea of what it does.\PYGZgt{}
    Copyright (C) \PYGZlt{}year\PYGZgt{}  \PYGZlt{}name of author\PYGZgt{}

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110\PYGZhy{}1301 USA.

Also add information on how to contact you by electronic and paper mail.

If the program is interactive, make it output a short notice like this
when it starts in an interactive mode:

    Gnomovision version 69, Copyright (C) year name of author
    Gnomovision comes with ABSOLUTELY NO WARRANTY; for details type `show w\PYGZsq{}.
    This is free software, and you are welcome to redistribute it
    under certain conditions; type `show c\PYGZsq{} for details.

The hypothetical commands `show w\PYGZsq{} and `show c\PYGZsq{} should show the appropriate
parts of the General Public License.  Of course, the commands you use may
be called something other than `show w\PYGZsq{} and `show c\PYGZsq{}; they could even be
mouse\PYGZhy{}clicks or menu items\PYGZhy{}\PYGZhy{}whatever suits your program.

You should also get your employer (if you work as a programmer) or your
school, if any, to sign a \PYGZdq{}copyright disclaimer\PYGZdq{} for the program, if
necessary.  Here is a sample; alter the names:

  Yoyodyne, Inc., hereby disclaims all copyright interest in the program
  `Gnomovision\PYGZsq{} (which makes passes at compilers) written by James Hacker.

  \PYGZlt{}signature of Ty Coon\PYGZgt{}, 1 April 1989
  Ty Coon, President of Vice

This General Public License does not permit incorporating your program into
proprietary programs.  If your program is a subroutine library, you may
consider it more useful to permit linking proprietary applications with the
library.  If this is what you want to do, use the GNU Lesser General
Public License instead of this License.
\end{sphinxVerbatim}


\section{License for PyPop documentation}
\label{\detokenize{docs/licenses:license-for-pypop-documentation}}

\subsection{GNU Free Documentation License}
\label{\detokenize{docs/licenses:gnu-free-documentation-license}}\label{\detokenize{docs/licenses:gfdl}}\begin{quote}

\sphinxAtStartPar
Copyright (C) 2000,2001,2002 Free Software Foundation, Inc. 59 Temple
Place, Suite 330, Boston, MA 02111\sphinxhyphen{}1307 USA Everyone is permitted to
copy and distribute verbatim copies of this license document, but
changing it is not allowed.
\end{quote}

\sphinxAtStartPar
\sphinxstylestrong{PREAMBLE}

\sphinxAtStartPar
The purpose of this License is to make a manual, textbook, or other
functional and useful document “free” in the sense of freedom: to assure
everyone the effective freedom to copy and redistribute it, with or
without modifying it, either commercially or noncommercially.
Secondarily, this License preserves for the author and publisher a way
to get credit for their work, while not being considered responsible for
modifications made by others.

\sphinxAtStartPar
This License is a kind of “copyleft”, which means that derivative works
of the document must themselves be free in the same sense. It
complements the GNU General Public License, which is a copyleft license
designed for free software.

\sphinxAtStartPar
We have designed this License in order to use it for manuals for free
software, because free software needs free documentation: a free program
should come with manuals providing the same freedoms that the software
does. But this License is not limited to software manuals; it can be
used for any textual work, regardless of subject matter or whether it is
published as a printed book. We recommend this License principally for
works whose purpose is instruction or reference.

\sphinxAtStartPar
\sphinxstylestrong{APPLICABILITY AND DEFINITIONS}

\sphinxAtStartPar
This License applies to any manual or other work, in any medium, that
contains a notice placed by the copyright holder saying it can be
distributed under the terms of this License. Such a notice grants a
world\sphinxhyphen{}wide, royalty\sphinxhyphen{}free license, unlimited in duration, to use that
work under the conditions stated herein. The “Document”, below, refers
to any such manual or work. Any member of the public is a licensee, and
is addressed as “you”. You accept the license if you copy, modify or
distribute the work in a way requiring permission under copyright law.

\sphinxAtStartPar
A “Modified Version” of the Document means any work containing the
Document or a portion of it, either copied verbatim, or with
modifications and/or translated into another language.

\sphinxAtStartPar
A “Secondary Section” is a named appendix or a front\sphinxhyphen{}matter section of
the Document that deals exclusively with the relationship of the
publishers or authors of the Document to the Document’s overall subject
(or to related matters) and contains nothing that could fall directly
within that overall subject. (Thus, if the Document is in part a
textbook of mathematics, a Secondary Section may not explain any
mathematics.) The relationship could be a matter of historical
connection with the subject or with related matters, or of legal,
commercial, philosophical, ethical or political position regarding them.

\sphinxAtStartPar
The “Invariant Sections” are certain Secondary Sections whose titles are
designated, as being those of Invariant Sections, in the notice that
says that the Document is released under this License. If a section does
not fit the above definition of Secondary then it is not allowed to be
designated as Invariant. The Document may contain zero Invariant
Sections. If the Document does not identify any Invariant Sections then
there are none.

\sphinxAtStartPar
The “Cover Texts” are certain short passages of text that are listed, as
Front\sphinxhyphen{}Cover Texts or Back\sphinxhyphen{}Cover Texts, in the notice that says that the
Document is released under this License. A Front\sphinxhyphen{}Cover Text may be at
most 5 words, and a Back\sphinxhyphen{}Cover Text may be at most 25 words.

\sphinxAtStartPar
A “Transparent” copy of the Document means a machine\sphinxhyphen{}readable copy,
represented in a format whose specification is available to the general
public, that is suitable for revising the document straightforwardly
with generic text editors or (for images composed of pixels) generic
paint programs or (for drawings) some widely available drawing editor,
and that is suitable for input to text formatters or for automatic
translation to a variety of formats suitable for input to text
formatters. A copy made in an otherwise Transparent file format whose
markup, or absence of markup, has been arranged to thwart or discourage
subsequent modification by readers is not Transparent. An image format
is not Transparent if used for any substantial amount of text. A copy
that is not “Transparent” is called “Opaque”.

\sphinxAtStartPar
Examples of suitable formats for Transparent copies include plain ASCII
without markup, Texinfo input format, LaTeX input format, SGML or XML
using a publicly available DTD, and standard\sphinxhyphen{}conforming simple HTML,
PostScript or PDF designed for human modification. Examples of
transparent image formats include PNG, XCF and JPG. Opaque formats
include proprietary formats that can be read and edited only by
proprietary word processors, SGML or XML for which the DTD and/or
processing tools are not generally available, and the machine\sphinxhyphen{}generated
HTML, PostScript or PDF produced by some word processors for output
purposes only.

\sphinxAtStartPar
The “Title Page” means, for a printed book, the title page itself, plus
such following pages as are needed to hold, legibly, the material this
License requires to appear in the title page. For works in formats which
do not have any title page as such, “Title Page” means the text near the
most prominent appearance of the work’s title, preceding the beginning
of the body of the text.

\sphinxAtStartPar
A section “Entitled XYZ” means a named subunit of the Document whose
title either is precisely XYZ or contains XYZ in parentheses following
text that translates XYZ in another language. (Here XYZ stands for a
specific section name mentioned below, such as “Acknowledgements”,
“Dedications”, “Endorsements”, or “History”.) To “Preserve the Title” of
such a section when you modify the Document means that it remains a
section “Entitled XYZ” according to this definition.

\sphinxAtStartPar
The Document may include Warranty Disclaimers next to the notice which
states that this License applies to the Document. These Warranty
Disclaimers are considered to be included by reference in this License,
but only as regards disclaiming warranties: any other implication that
these Warranty Disclaimers may have is void and has no effect on the
meaning of this License.

\sphinxAtStartPar
\sphinxstylestrong{VERBATIM COPYING}

\sphinxAtStartPar
You may copy and distribute the Document in any medium, either
commercially or noncommercially, provided that this License, the
copyright notices, and the license notice saying this License applies to
the Document are reproduced in all copies, and that you add no other
conditions whatsoever to those of this License. You may not use
technical measures to obstruct or control the reading or further copying
of the copies you make or distribute. However, you may accept
compensation in exchange for copies. If you distribute a large enough
number of copies you must also follow the conditions in section 3.

\sphinxAtStartPar
You may also lend copies, under the same conditions stated above, and
you may publicly display copies.

\sphinxAtStartPar
\sphinxstylestrong{COPYING IN QUANTITY}

\sphinxAtStartPar
If you publish printed copies (or copies in media that commonly have
printed covers) of the Document, numbering more than 100, and the
Document’s license notice requires Cover Texts, you must enclose the
copies in covers that carry, clearly and legibly, all these Cover Texts:
Front\sphinxhyphen{}Cover Texts on the front cover, and Back\sphinxhyphen{}Cover Texts on the back
cover. Both covers must also clearly and legibly identify you as the
publisher of these copies. The front cover must present the full title
with all words of the title equally prominent and visible. You may add
other material on the covers in addition. Copying with changes limited
to the covers, as long as they preserve the title of the Document and
satisfy these conditions, can be treated as verbatim copying in other
respects.

\sphinxAtStartPar
If the required texts for either cover are too voluminous to fit
legibly, you should put the first ones listed (as many as fit
reasonably) on the actual cover, and continue the rest onto adjacent
pages.

\sphinxAtStartPar
If you publish or distribute Opaque copies of the Document numbering
more than 100, you must either include a machine\sphinxhyphen{}readable Transparent
copy along with each Opaque copy, or state in or with each Opaque copy a
computer\sphinxhyphen{}network location from which the general network\sphinxhyphen{}using public
has access to download using public\sphinxhyphen{}standard network protocols a
complete Transparent copy of the Document, free of added material. If
you use the latter option, you must take reasonably prudent steps, when
you begin distribution of Opaque copies in quantity, to ensure that this
Transparent copy will remain thus accessible at the stated location
until at least one year after the last time you distribute an Opaque
copy (directly or through your agents or retailers) of that edition to
the public.

\sphinxAtStartPar
It is requested, but not required, that you contact the authors of the
Document well before redistributing any large number of copies, to give
them a chance to provide you with an updated version of the Document.

\sphinxAtStartPar
\sphinxstylestrong{MODIFICATIONS}

\sphinxAtStartPar
You may copy and distribute a Modified Version of the Document under the
conditions of sections 2 and 3 above, provided that you release the
Modified Version under precisely this License, with the Modified Version
filling the role of the Document, thus licensing distribution and
modification of the Modified Version to whoever possesses a copy of it.
In addition, you must do these things in the Modified Version:
\begin{enumerate}
\sphinxsetlistlabels{\Alph}{enumi}{enumii}{}{.}%
\item {} 
\sphinxAtStartPar
Use in the Title Page (and on the covers, if any) a title distinct
from that of the Document, and from those of previous versions (which
should, if there were any, be listed in the History section of the
Document). You may use the same title as a previous version if the
original publisher of that version gives permission.

\item {} 
\sphinxAtStartPar
List on the Title Page, as authors, one or more persons or entities
responsible for authorship of the modifications in the Modified
Version, together with at least five of the principal authors of the
Document (all of its principal authors, if it has fewer than five),
unless they release you from this requirement.

\item {} 
\sphinxAtStartPar
State on the Title page the name of the publisher of the Modified
Version, as the publisher.

\item {} 
\sphinxAtStartPar
Preserve all the copyright notices of the Document.

\item {} 
\sphinxAtStartPar
Add an appropriate copyright notice for your modifications adjacent
to the other copyright notices.

\item {} 
\sphinxAtStartPar
Include, immediately after the copyright notices, a license notice
giving the public permission to use the Modified Version under the
terms of this License, in the form shown in the
Addendum below.

\item {} 
\sphinxAtStartPar
Preserve in that license notice the full lists of Invariant Sections
and required Cover Texts given in the Document’s license notice.

\item {} 
\sphinxAtStartPar
Include an unaltered copy of this License.

\item {} 
\sphinxAtStartPar
Preserve the section Entitled “History”, Preserve its Title, and add
to it an item stating at least the title, year, new authors, and
publisher of the Modified Version as given on the Title Page. If
there is no section Entitled “History” in the Document, create one
stating the title, year, authors, and publisher of the Document as
given on its Title Page, then add an item describing the Modified
Version as stated in the previous sentence.

\item {} 
\sphinxAtStartPar
Preserve the network location, if any, given in the Document for
public access to a Transparent copy of the Document, and likewise the
network locations given in the Document for previous versions it was
based on. These may be placed in the “History” section. You may omit
a network location for a work that was published at least four years
before the Document itself, or if the original publisher of the
version it refers to gives permission.

\item {} 
\sphinxAtStartPar
For any section Entitled “Acknowledgements” or “Dedications”,
Preserve the Title of the section, and preserve in the section all
the substance and tone of each of the contributor acknowledgements
and/or dedications given therein.

\item {} 
\sphinxAtStartPar
Preserve all the Invariant Sections of the Document, unaltered in
their text and in their titles. Section numbers or the equivalent are
not considered part of the section titles.

\item {} 
\sphinxAtStartPar
Delete any section Entitled “Endorsements”. Such a section may not be
included in the Modified Version.

\item {} 
\sphinxAtStartPar
Do not retitle any existing section to be Entitled “Endorsements” or
to conflict in title with any Invariant Section.

\item {} 
\sphinxAtStartPar
Preserve any Warranty Disclaimers.

\end{enumerate}

\sphinxAtStartPar
If the Modified Version includes new front\sphinxhyphen{}matter sections or appendices
that qualify as Secondary Sections and contain no material copied from
the Document, you may at your option designate some or all of these
sections as invariant. To do this, add their titles to the list of
Invariant Sections in the Modified Version’s license notice. These
titles must be distinct from any other section titles.

\sphinxAtStartPar
You may add a section Entitled “Endorsements”, provided it contains
nothing but endorsements of your Modified Version by various
parties\textendash{}for example, statements of peer review or that the text has
been approved by an organization as the authoritative definition of a
standard.

\sphinxAtStartPar
You may add a passage of up to five words as a Front\sphinxhyphen{}Cover Text, and a
passage of up to 25 words as a Back\sphinxhyphen{}Cover Text, to the end of the list
of Cover Texts in the Modified Version. Only one passage of Front\sphinxhyphen{}Cover
Text and one of Back\sphinxhyphen{}Cover Text may be added by (or through arrangements
made by) any one entity. If the Document already includes a cover text
for the same cover, previously added by you or by arrangement made by
the same entity you are acting on behalf of, you may not add another;
but you may replace the old one, on explicit permission from the
previous publisher that added the old one.

\sphinxAtStartPar
The author(s) and publisher(s) of the Document do not by this License
give permission to use their names for publicity for or to assert or
imply endorsement of any Modified Version.

\sphinxAtStartPar
\sphinxstylestrong{COMBINING DOCUMENTS}

\sphinxAtStartPar
You may combine the Document with other documents released under this
License, under the terms defined in section 4. above for
modified versions, provided that you include in the combination all of
the Invariant Sections of all of the original documents, unmodified, and
list them all as Invariant Sections of your combined work in its license
notice, and that you preserve all their Warranty Disclaimers.

\sphinxAtStartPar
The combined work need only contain one copy of this License, and
multiple identical Invariant Sections may be replaced with a single
copy. If there are multiple Invariant Sections with the same name but
different contents, make the title of each such section unique by adding
at the end of it, in parentheses, the name of the original author or
publisher of that section if known, or else a unique number. Make the
same adjustment to the section titles in the list of Invariant Sections
in the license notice of the combined work.

\sphinxAtStartPar
In the combination, you must combine any sections Entitled “History” in
the various original documents, forming one section Entitled “History”;
likewise combine any sections Entitled “Acknowledgements”, and any
sections Entitled “Dedications”. You must delete all sections Entitled
“Endorsements”.

\sphinxAtStartPar
\sphinxstylestrong{COLLECTIONS OF DOCUMENTS}

\sphinxAtStartPar
You may make a collection consisting of the Document and other documents
released under this License, and replace the individual copies of this
License in the various documents with a single copy that is included in
the collection, provided that you follow the rules of this License for
verbatim copying of each of the documents in all other respects.

\sphinxAtStartPar
You may extract a single document from such a collection, and distribute
it individually under this License, provided you insert a copy of this
License into the extracted document, and follow this License in all
other respects regarding verbatim copying of that document.

\sphinxAtStartPar
\sphinxstylestrong{AGGREGATION WITH INDEPENDENT WORKS}

\sphinxAtStartPar
A compilation of the Document or its derivatives with other separate and
independent documents or works, in or on a volume of a storage or
distribution medium, is called an “aggregate” if the copyright resulting
from the compilation is not used to limit the legal rights of the
compilation’s users beyond what the individual works permit. When the
Document is included in an aggregate, this License does not apply to the
other works in the aggregate which are not themselves derivative works
of the Document.

\sphinxAtStartPar
If the Cover Text requirement of section 3 is applicable to these copies
of the Document, then if the Document is less than one half of the
entire aggregate, the Document’s Cover Texts may be placed on covers
that bracket the Document within the aggregate, or the electronic
equivalent of covers if the Document is in electronic form. Otherwise
they must appear on printed covers that bracket the whole aggregate.

\sphinxAtStartPar
\sphinxstylestrong{TRANSLATION}

\sphinxAtStartPar
Translation is considered a kind of modification, so you may distribute
translations of the Document under the terms of section 4. Replacing
Invariant Sections with translations requires special permission from
their copyright holders, but you may include translations of some or all
Invariant Sections in addition to the original versions of these
Invariant Sections. You may include a translation of this License, and
all the license notices in the Document, and any Warranty Disclaimers,
provided that you also include the original English version of this
License and the original versions of those notices and disclaimers. In
case of a disagreement between the translation and the original version
of this License or a notice or disclaimer, the original version will
prevail.

\sphinxAtStartPar
If a section in the Document is Entitled “Acknowledgements”,
“Dedications”, or “History”, the requirement (section 4) to Preserve its
Title (section 1) will typically require changing the actual title.

\sphinxAtStartPar
\sphinxstylestrong{TERMINATION}

\sphinxAtStartPar
You may not copy, modify, sublicense, or distribute the Document except
as expressly provided for under this License. Any other attempt to copy,
modify, sublicense or distribute the Document is void, and will
automatically terminate your rights under this License. However, parties
who have received copies, or rights, from you under this License will
not have their licenses terminated so long as such parties remain in
full compliance.

\sphinxAtStartPar
\sphinxstylestrong{FUTURE REVISIONS OF THIS LICENSE}

\sphinxAtStartPar
The Free Software Foundation may publish new, revised versions of the
GNU Free Documentation License from time to time. Such new versions will
be similar in spirit to the present version, but may differ in detail to
address new problems or concerns. See \sphinxurl{http://www.gnu.org/copyleft/}.

\sphinxAtStartPar
Each version of the License is given a distinguishing version number. If
the Document specifies that a particular numbered version of this
License “or any later version” applies to it, you have the option of
following the terms and conditions either of that specified version or
of any later version that has been published (not as a draft) by the
Free Software Foundation. If the Document does not specify a version
number of this License, you may choose any version ever published (not
as a draft) by the Free Software Foundation.

\sphinxAtStartPar
\sphinxstylestrong{ADDENDUM: How to use this License for your documents}

\sphinxAtStartPar
To use this License in a document you have written, include a copy of
the License in the document and put the following copyright and license
notices just after the title page:
\begin{quote}

\sphinxAtStartPar
Copyright (c) YEAR YOUR NAME. Permission is granted to copy,
distribute and/or modify this document under the terms of the GNU
Free Documentation License, Version 1.2 or any later version
published by the Free Software Foundation; with no Invariant
Sections, no Front\sphinxhyphen{}Cover Texts, and no Back\sphinxhyphen{}Cover Texts. A copy of
the license is included in the section entitled “GNU Free
Documentation License”.
\end{quote}

\sphinxAtStartPar
If you have Invariant Sections, Front\sphinxhyphen{}Cover Texts and Back\sphinxhyphen{}Cover Texts,
replace the “with…Texts.” line with this:
\begin{quote}

\sphinxAtStartPar
with the Invariant Sections being LIST THEIR TITLES, with the
Front\sphinxhyphen{}Cover Texts being LIST, and with the Back\sphinxhyphen{}Cover Texts being
LIST.
\end{quote}

\sphinxAtStartPar
If you have Invariant Sections without Cover Texts, or some other
combination of the three, merge those two alternatives to suit the
situation.

\sphinxAtStartPar
If your document contains nontrivial examples of program code, we
recommend releasing these examples in parallel under your choice of free
software license, such as the GNU General Public License, to permit
their use in free software.

\endgroup

\sphinxstepscope

\cleardoublepage
\begingroup
\renewcommand\chapter[1]{\endgroup}
\phantomsection


\chapter{References}
\label{\detokenize{docs/biblio:references}}\label{\detokenize{docs/biblio:refs}}\label{\detokenize{docs/biblio::doc}}
\begin{sphinxthebibliography}{Excoffie}
\bibitem[Cano:etal:2007]{docs/biblio:cano-etal-2007}
\sphinxAtStartPar
Pedro Cano. 2007. “Common and well\sphinxhyphen{}documented HLA
alleles: report of the ad\sphinxhyphen{}hoc committee of the American Society
for Histocompatiblity and Immunogenetics”. 68. 5. 392\sphinxhyphen{}417.
\sphinxstyleemphasis{Human Immunology}.
\bibitem[Chen:etal:1999]{docs/biblio:chen-etal-1999}
\sphinxAtStartPar
JJ Chen, JA Hollenbach, EA Trachtenberg, JJ Just,
M Carrington, KS Ronningen, A Begovich, MC King, SK McWeeney, SJ
Mack, HA Erlich, and G Thomson. 1999. “Hardy\sphinxhyphen{}Weinberg testing for
HLA class II (DRB1, DQA1, DQB1 and DPB1) loci in 26 human ethnic
groups”. 54. 533\sphinxhyphen{}542. \sphinxstyleemphasis{Tissue Antigens}.
\bibitem[Cramer:1946]{docs/biblio:cramer-1946}
\sphinxAtStartPar
H Cramer. \sphinxstyleemphasis{Mathematical Models of Statistics}.
1946. Princeton University Press. Princeton NJ.
\bibitem[Dempster:1977]{docs/biblio:dempster-1977}
\sphinxAtStartPar
A Dempster, N Laird, and D Rubin. 1977. “Maximum
likelihood estimation from incomplete data using the EM
algorithm”. 39. 1\sphinxhyphen{}38. \sphinxstyleemphasis{J Royal Stat Soc}.
\bibitem[Ewens:1972]{docs/biblio:ewens-1972}
\sphinxAtStartPar
W Ewens. 1972. “The sampling theory of selectively
neutral alleles.”. \sphinxstyleemphasis{Theor. Pop. Biol}. 3. 87\sphinxhyphen{}112.
\bibitem[Excoffier:Slatkin:1995]{docs/biblio:excoffier-slatkin-1995}
\sphinxAtStartPar
Laurent Excoffier and Montgomery Slatkin.
1995. “Maximum\sphinxhyphen{}likelihood estimation of molecular haplotype
frequencies in a diploid population”. \sphinxstyleemphasis{Molecular Biology and
Evolution}. 12. 5. 921\sphinxhyphen{}927.
\bibitem[Guo:Thompson:1992]{docs/biblio:guo-thompson-1992}
\sphinxAtStartPar
S W Guo and E A Thompson. 1992. “Performing
the exact test of Hardy\sphinxhyphen{}Weinberg proportion for multiple alleles”.
\sphinxstyleemphasis{Biometrics}. 48. 361\sphinxhyphen{}72.
\bibitem[Hedrick:1987]{docs/biblio:hedrick-1987}
\sphinxAtStartPar
P W Hedrick. 1987. “Gametic disequilibrium
measures: proceed with caution”. 117. 2. 331\sphinxhyphen{}41. \sphinxstyleemphasis{Genetics}.
\bibitem[Lancaster:etal:2003]{docs/biblio:lancaster-etal-2003}
\sphinxAtStartPar
Alex Lancaster, Mark P Nelson, Richard M
Single, Diogo Meyer, and Glenys Thomson. 2003. “PyPop: a software
framework for population genomics: analyzing large\sphinxhyphen{}scale
multi\sphinxhyphen{}locus genotype data”. \sphinxstyleemphasis{Pacific Symposium on Biocomputing}.
R B Altman. et al.. vol 8. 514\sphinxhyphen{}525. .. {[}\sphinxcode{\sphinxupquote{PDF}} (344 kB){]}.
\bibitem[Lancaster:etal:2007a]{docs/biblio:lancaster-etal-2007a}
\sphinxAtStartPar
Alex Lancaster, Mark P Nelson, Richard M
Single, Diogo Meyer, and Glenys Thomson. 2007a. \sphinxstyleemphasis{Software
framework for Biostatistics Core}. JA Hansen. \sphinxstyleemphasis{Immunobiology of
the Human MHC: Proceedings of the 13th International
Histocompatibility Workshop and Conference}. IHWG Press. Seattle,
Washington. I. 510\sphinxhyphen{}517.
\bibitem[Lancaster:etal:2007b]{docs/biblio:lancaster-etal-2007b}
\sphinxAtStartPar
Alex Lancaster, Richard M Single, Owen D
Solberg, Mark P Nelson, and Glenys Thomson. 2007b. “PyPop update \sphinxhyphen{}
a software pipeline for large\sphinxhyphen{}scale multilocus population
genomics.”. \sphinxstyleemphasis{Tissue Antigens}. 69 Suppl 1. 192\sphinxhyphen{}197.
.. {[}\sphinxcode{\sphinxupquote{PDF}}
(150 kB){]}.
\bibitem[Mack:etal:2007]{docs/biblio:mack-etal-2007}
\sphinxAtStartPar
Steven J. Mack, Alicia Sanchez\sphinxhyphen{}Mazas, Diogo
Meyer, Richard M Single, Yingssu Tsai, and Henry Erlich. 2007.
\sphinxstyleemphasis{Methods used in the generation and preparation of data for
analysis in the 13th International Histocompatibility Workshop}.
JA Hansen. \sphinxstyleemphasis{Immunobiology of the Human MHC: Proceedings of the
13th International Histocompatibility Workshop and Conference}.
IHWG Press. Seattle, Washington. I. 564\sphinxhyphen{}579.
\bibitem[Meyer:etal:2007]{docs/biblio:meyer-etal-2007}
\sphinxAtStartPar
Diogo Meyer, Richard M Single, Steven J. Mack,
Alex Lancaster, Mark P Nelson, M. Fernánndez\sphinxhyphen{}Viñaa, Henry Erlich,
and Glenys Thomson. 2007. \sphinxstyleemphasis{Haplotype frequencies and linkage
disequilibrium among classical HLA genes}. JA Hansen.
\sphinxstyleemphasis{Immunobiology of the Human MHC: Proceedings of the 13th
International Histocompatibility Workshop and Conference}. IHWG
Press. Seattle, Washington. I. 705\sphinxhyphen{}746.
\bibitem[Salamon:etal:1999]{docs/biblio:salamon-etal-1999}
\sphinxAtStartPar
H Salamon, W Klitz, S Easteal, X Gao, HA
Erlich, M Fernandez\sphinxhyphen{}Vina, and EA Trachtenberg. 1999. “Evolution of
HLA class II molecules: Allelic and amino acid site variability
across populations”. 152. 393\sphinxhyphen{}400. \sphinxstyleemphasis{Genetics}.
\bibitem[Schneider:etal:2000]{docs/biblio:schneider-etal-2000}
\sphinxAtStartPar
S Schneider, D Roessli, and L Excoffier.
2000. “Arlequin: A software for population genetics data analysis.
\sphinxurl{http://lgb.unige.ch/arlequin/}”. Ver 2.000. Genetics and Biometry
Lab, Dept. of Anthropology, University of Geneva.
\bibitem[Single:etal:2007a]{docs/biblio:single-etal-2007a}
\sphinxAtStartPar
RM Single, D Meyer, and G Thomson. 2007a.
\sphinxstyleemphasis{Statistical methods for analysis of population genetic data}.
JA Hansen. \sphinxstyleemphasis{Immunobiology of the Human MHC: Proceedings of the
13th International Histocompatibility Workshop and Conference}.
IHWG Press. Seattle, Washington. I. 518\sphinxhyphen{}522.
\bibitem[Slatkin:1994]{docs/biblio:slatkin-1994}
\sphinxAtStartPar
M Slatkin. 1994. “An exact test for neutrality
based on the Ewens sampling distribution”. \sphinxstyleemphasis{Genetical Research}.
64. 71\sphinxhyphen{}74.
\bibitem[Slatkin:1996]{docs/biblio:slatkin-1996}
\sphinxAtStartPar
M Slatkin. 1996. “A correction to the exact test
based on the Ewens sampling distribution”. \sphinxstyleemphasis{Genetical Research}.
68. 259\sphinxhyphen{}260.
\bibitem[Solberg:etal:2008]{docs/biblio:solberg-etal-2008}
\sphinxAtStartPar
Owen D. Solberg, Steven J. Mack, Alex K. Lancaster,
Richard M. Single, Yingssu Tsai, Alicia Sanchez\sphinxhyphen{}Mazas, and Glenys
Thomson Hum Immunol (2008), doi: 10.1016/j.humimm.2008.05.001
\bibitem[Thomson:Single:2014]{docs/biblio:thomson-single-2014}
\sphinxAtStartPar
Glenys Thomson and Richard M. Single. 2014.
“Conditional Asymmetric Linkage Disequilibrium (ALD): Extending the
Biallelic r2 Measure” \sphinxstyleemphasis{Genetics}, 198, 321\textendash{}331.
\bibitem[Watterson:1978]{docs/biblio:watterson-1978}
\sphinxAtStartPar
G Watterson. 1978. “The homozygosity test of
neutrality”. \sphinxstyleemphasis{Genetics}. 88. 405\sphinxhyphen{}417.
\end{sphinxthebibliography}



\renewcommand{\indexname}{Index}
\printindex
\end{document}