<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="PyPop : Python for Population Genomics">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>PyPop</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/alexlancaster/pypop">View on GitHub</a>

          <h1 id="project_title">PyPop</h1>
          <h2 id="project_tagline">Python for Population Genomics</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/alexlancaster/pypop/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/alexlancaster/pypop/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>PyPop: Python for Population Genomics</p>

       
        
<blockquote>
<ul>
<li><strong>PyPop development is now on GitHub.  The current binary releases described below are very outdated and only work on Python 2 which is also no longer maintained.  
  We are working on a new Python 3-compatible v1.0.0 release to be made available as PyPi.org package, but for the time being, there are instructions 
  on the GitHub <a href="https://github.com/alexlancaster/pypop#readme">README.md</a> for how to build PyPop from source on a modern Python 3 system.<</strong>/li>
</ul>
</blockquote>

<h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span class="octicon octicon-link"></span></a>Overview</h3>

<p>PyPop (Python for Population Genomics) is an environment for doing large-scale population genetic analyses including: (1) conformity to Hardy-Weinberg expectations, (2) tests for balancing or directional selection; (3) estimates of haplotype frequencies (and their distributions) and measures and tests of significance for linkage disequilibrium (LD).</p>

<p>It is an object-oriented framework implemented in <a href="http://www.pypop.org/">Python</a>, a language with powerful features for interfacing with other languages, such as C (in which we have already implemented many routines and which is particularly suited to computationally intensive tasks).</p>

<p>The output of the analyses are stored in the XML. These output files can then be transformed using standard tools into many other data formats suitable for machine input (such as PHYLIP or input for spreadsheet programs such as Excel or statistical packages, such as R, plain text, or HTML for human-readable format. Storing the output in XML allows the final viewable output format to be redesigned at will, without requiring the (often time-consuming) re-running of the analyses themselves.</p>

<p>An outline of PyPop can be found in our excellent 2007 <em>Tissue Antigens</em> and 2003 <em>PSB</em> <a href="#how-to-cite-pypop">papers</a>
.
PyPop is available for download</p>

<ul>
<li>2023: ported to Python 3, developer can compile from source, working towards a v1.0.0 release - no official release date yet.</li>
<li>2017: all new development is now in GitHub, no official release yet</li>
<li>2008-09-09: 0.7.0 release (many new features and bug fixes) ‚Üê Most recent</li>
<li>2005-04-13: 0.6.0 released (new features and bug fixes)</li>
<li>2004-03-09: 0.5.2 released (bug fix release, fixes Windows 98 .bat file problems)</li>
<li>2004-02-26: 0.5.1 released (mainly a bug fix and maintainance release)</li>
<li>2003-12-31: 0.5 released (first public beta)</li>
</ul>

<p>(Please note that links and documentation below always refer to the most recent release).</p>

<h3>
<a id="binaries" class="anchor" href="#binaries" aria-hidden="true"><span class="octicon octicon-link"></span></a>
  Binaries [please note these binaries are obsolete and may or may not work]</h3>

<p>Beta binary versions of PyPop are now publicly available (see "Getting and installing PyPop" link below for links to the current binaries) for both:</p>

<ul>
<li>GNU/Linux (the Linux binary needs a recent distribution that a recent glibc (2.8 has been tested): Fedora Core 9 is known to work, earlier versions of Red Hat such as the 7.x series and 8.0 are known to not work)</li>
<li>As of early 2023, pypop is no longer an official Fedora package.  Future releases will be made available via PyPi.org</ul>

<ul>
<li>Windows (the binary has been tested on Windows 98, 2000 and XP)</li>
</ul>

<p>Both binary packages are approximately 5.5 Mb downloads.
Source</p>

<p>PyPop is <a href="http://www.gnu.org/philosophy/free-sw.html">free software</a> (sometimes referred to as <a href="http://www.opensource.org/">open source</a> software) and the source code is released under the terms of the "copyleft" GNU General Public License, or GPL (<a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a>). This means even if we haven't compiled a binary for your platform, it is possible for you to download the source code and compile it yourself.</p>

<h3>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h3>

<p>Documentation for PyPop is contained in the ''PyPop User Guide''.</p>

<ul>
<li>
<a href="docs/">''PyPop User Guide''</a> includes:

<ul>
<li>
<a href="docs/guide-chapter-install.html">Getting and installing PyPop</a> (including links to the binaries and source code)</li>
<li>
<a href="docs/guide-chapter-usage.html">Getting started with PyPop</a> (instructions on how to configure and run PyPop)</li>
<li>
<a href="docs/guide-chapter-instructions.html">Interpreting PyPop output</a> (a description of the output statistics)</li>
</ul>
</li>
<li>PyPop User Guide is also available as a <a href="pypop-guide.pdf">PDF</a> [220 kB]</li>
</ul>

<p>Please be aware that this is a beta release so it is highly likely that there may be bugs and wrinkles to iron out. Please direct all questions to Alex Lancaster at the address below.</p>

<h3>
<a id="how-to-cite-pypop" class="anchor" href="#how-to-cite-pypop" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to cite PyPop</h3>

<p>When citing PyPop, please cite the (2007) paper from <em>Tissue Antigens</em>:</p>

<ul>
<li>A. K. Lancaster, R. M. Single, O. D. Solberg, M. P. Nelson and G. Thomson (2007) "PyPop update - a software pipeline for large-scale multilocus population genomics" <em>Tissue Antigens</em> 69 (s1), 192-197. [<a href="http://dx.doi.org/10.1111/j.1399-0039.2006.00769.x">journal page</a>, <a href="tissue-antigens-lancaster-2007.pdf">preprint PDF (112 kB)</a>].</li>
</ul>

<p>In addition, you can also cite our 2003 Pacific Symposium on Biocomputing paper:</p>

<ul>
<li>Alex Lancaster, Mark P. Nelson, Richard M. Single, Diogo Meyer, and Glenys Thomson (2003) "PyPop: a software framework for population genomics: analyzing large-scale multi-locus genotype data", in <em>Pacific Symposium on Biocomputing</em> vol. 8:514-525 (edited by R B Altman. et al., World Scientific, Singapore, 2003) [<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3891851/">PubMed Central</a>, <a href="psb-pypop.pdf">PDF (344 kB)</a>].</li>
</ul>

<h3>
<a id="population-data-files" class="anchor" href="#population-data-files" aria-hidden="true"><span class="octicon octicon-link"></span></a>Population data files</h3>
<p>Population data files and online supporting materials for published studies listed in the Solberg et al. meta-analysis paper may be found <a href="http://www.pypop.org/popdata/">here</p>

<h3>
<a id="immportorg" class="anchor" href="#immportorg" aria-hidden="true"><span class="octicon octicon-link"></span></a>ImmPort.org</h3>

<p>PyPop is affiliated with ImmPort.org, the Immunology Database and Analysis Portal. The ImmPort system provides advanced information technology support in the production, analysis, archiving, and exchange of scientific data for the diverse community of life science researchers supported by NIAID/DAIT. The development of the ImmPort system was supported by the NIH/NIAID Bioinformatics Integration Support Contract (BISC), Phase II.</p>

<h3>
<a id="acknowlegements" class="anchor" href="#acknowlegements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Acknowlegements</h3>

<p>This work has benefited from the support of NIH grant AI49213 (13th IHW) and NIH/NIAID Contract number HHSN266200400076C N01-AI-40076. Thanks to Steven J. Mack, Kristie A. Mather, Steve Marsh, Mark Grote and Leslie Louie for helpful comments and testing.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">PyPop maintained by <a href="https://github.com/alexlancaster">alexlancaster</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
