
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>PyPop: Python for Population Genomics &#8212; PyPop 1.3.1 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css?v=4ae1632d" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=deb9c837" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=bb516dca"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'index';</script>
    <link rel="icon" href="_static/pypop-favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PyPop User Guide" href="docs/index.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="1.3.1" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
  <aside class="bd-header-announcement d-print-none d-none" aria-label="Announcement" data-pst-announcement-url="https://raw.githubusercontent.com/alexlancaster/pypop/refs/heads/main/website/_templates/announcement_banner.html"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="#">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/pypop-logo.png" class="logo__image only-light" alt="PyPop - Home"/>
    <img src="_static/pypop-logo.png" class="logo__image only-dark pst-js-only" alt="PyPop - Home"/>
  
  
    <p class="title logo__title">PyPop</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="docs/index.html">
    PyPop User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="api/index.html">
    API Reference (1.3.1)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="http://pypop.org/popdata/">
    Solberg et al. (2008)
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/alexlancaster/pypop/" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="docs/index.html">
    PyPop User Guide
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="api/index.html">
    API Reference (1.3.1)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="http://pypop.org/popdata/">
    Solberg et al. (2008)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/alexlancaster/pypop/" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none"></div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="pypop-python-for-population-genomics">
<span id="guide-preface-intro"></span><h1>PyPop: Python for Population Genomics<a class="headerlink" href="#pypop-python-for-population-genomics" title="Link to this heading">#</a></h1>
<div class="logo-container docutils container" id="guide-preface-1-start">
<div class="logo-cell docutils container">
<a class="no-background reference internal image-reference" href="_images/pypop-logo.png"><img alt="_images/pypop-logo.png" class="no-background align-center" src="_images/pypop-logo.png" style="width: 120px;" />
</a>
</div>
<div class="text-cell docutils container">
<p><strong>PyPop (Python for Population Genomics)</strong> is an environment for
doing large-scale population genetic analyses including:</p>
<ul class="simple">
<li><p>conformity to Hardy-Weinberg expectations</p></li>
<li><p>tests for balancing or directional selection</p></li>
<li><p>estimates of haplotype frequencies and measures and tests of
significance for linkage disequilibrium (LD).</p></li>
</ul>
</div>
</div>
<div class="dropdown toggle-shown admonition" id="news">
<span id="guide-preface-1-end"></span><p class="admonition-title">PyPop News</p>
<ul>
<li><p>2025-10-10: Add initial version of <a class="reference internal" href="api/index.html"><span class="doc">API Reference (1.3.1)</span></a> automatically
generated from code to document code for developers and users (see
<a class="reference internal" href="docs/guide-chapter-usage.html#guide-usage-examples-api"><span class="std std-ref">PyPop API examples</span></a>).</p></li>
<li><p>2025-10-07: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.3.1">PyPop 1.3.1</a> is
released. Bump <code class="docutils literal notranslate"><span class="pre">numpy</span></code> and <code class="docutils literal notranslate"><span class="pre">lxml</span></code> dependencies.</p>
<p>Experimental support for wheels built for Windows ARM64 is also
available in the <a class="reference external" href="https://test.pypi.org/project/pypop-genomics/">Test PyPI</a> repo, to test,
run:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>--extra-index-url<span class="w"> </span>https://test.pypi.org/simple/<span class="w"> </span>pypop-genomics
</pre></div>
</div>
<p>Please test and report issues <a class="reference external" href="https://github.com/alexlancaster/pypop/issues/new?assignees=&amp;labels=bug&amp;projects=&amp;template=bug_report.yml&amp;title=%5BWindows+ARM64+bug%5D%3A+Please+replace+this+with+a+clear+and+concise+description+of+your+bug">via the bug tracker</a>.</p>
</li>
<li><p>2024-04-01: PyPop paper <a class="reference external" href="https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2024.1378512/full">published</a>
in <em>Frontiers in Immunology</em>, <a class="reference internal" href="#citing-pypop"><span class="std std-ref">see citing PyPop</span></a>
for details.</p></li>
<li><p>More details, including recent previous releases:</p></li>
</ul>
<blockquote>
<div><div class="toggle docutils container">
<ul class="simple">
<li><p>2025-09-08: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.3.0">PyPop 1.3.0</a> is
released, adds Python wheels 3.14 and new behavior of the
<code class="docutils literal notranslate"><span class="pre">--filelist</span> <span class="pre">FILELIST</span></code> (files are <a class="reference internal" href="docs/guide-chapter-usage.html#guide-usage-filelist"><span class="std std-ref">resolved</span></a> relative to directory where <code class="docutils literal notranslate"><span class="pre">FILELIST</span></code>
is located, rather than working directory).</p></li>
<li><p>2025-07-28: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.2.2">PyPop 1.2.2</a> is
released, includes dependency updates and <code class="docutils literal notranslate"><span class="pre">nox</span></code> task framework
useful for contributors and developers.</p></li>
<li><p>2025-04-29: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.2.1">PyPop 1.2.1</a>
released, includes a bug fix in text output, and internal
changes and updates to dependencies.</p></li>
<li><p>2025-02-04: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.2.0">PyPop 1.2.0</a>
is released: includes updates and restoration of full
functionality for <code class="docutils literal notranslate"><span class="pre">[Sequence]</span></code> options as part of the
<code class="docutils literal notranslate"><span class="pre">[Filters]</span></code>, including dynamic downloads of HLA sequence data
and major updates to documentation, using new HLA nomenclature
throughout.</p></li>
<li><p>2025-01-16: Beta release <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.2.0-b2">PyPop 1.2.0b2</a>
is released.</p></li>
<li><p>2025-01-05: Pre-release <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.1.3-rc1">PyPop 1.1.3rc1</a>
is now available. Experimental wheels for Windows ARM64 are
added in this release.</p></li>
<li><p>2024-11-18: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.1.2">PyPop 1.1.2</a>
released: adds <code class="docutils literal notranslate"><span class="pre">--citation</span></code> command-line option to print citation
information and updates <code class="docutils literal notranslate"><span class="pre">numpy</span></code> to 2.1.3</p></li>
<li><p>2024-09-10: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.1.1">PyPop 1.1.1</a>
released, enables support for Python 3.13 and build Python 3.13 wheels.</p></li>
<li><p>2024-05-30: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.1.0">PyPop 1.1.0</a>
released. Increases the minimum macOS requirements to Catalina
(Intel) and Big Sur (Silicon) to ensure binary compatibility with
the GNU Scientific Library (<cite>gsl</cite>). Thanks to Steve Mack for
testing.</p></li>
<li><p>2024-03-08: PyPop paper, provisionally accepted.</p></li>
<li><p>2024-02-24: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.0.2">PyPop 1.0.2</a>
released. Code scanning updates and updated <code class="docutils literal notranslate"><span class="pre">numpy</span></code> to 1.26.4</p></li>
<li><p>2024-02-11: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.0.1">PyPop 1.0.1</a>
released. Added support for <code class="docutils literal notranslate"><span class="pre">ARM64</span></code> for Linux, and also
<code class="docutils literal notranslate"><span class="pre">muslinux</span></code> wheels. Improved support for scientific notation.</p></li>
<li><p>2024-02-01: <a class="reference external" href="https://zenodo.org/records/10602940">Preprint</a>
describing 1.0.0 released on Zenodo.</p></li>
<li><p>2023-11-07: <a class="reference external" href="https://github.com/alexlancaster/pypop/releases/tag/v1.0.0">PyPop 1.0.0</a>
released</p></li>
<li><p>Highlights of PyPop 1.0.0 include:</p>
<ul>
<li><p>PyPop now fully ported to Python 3.</p></li>
<li><p>New asymmetric linkage disequilibrium (ALD) computations
<span id="id1">(<a class="reference internal" href="docs/biblio.html#id25" title="Thomson, G. and Single, R. M. Conditional asymmetric linkage disequilibrium (ALD): extending the bi-allelic r² measure. Genetics, 198(1):321–331, 2014. doi:10.1534/genetics.114.165266.">Thomson and Single, 2014</a>)</span>:.</p></li>
<li><p>Improved tab-separated values (TSV) output file handling.</p></li>
<li><p>Preliminary support for <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3715123/">Genotype List (GL) String</a>.</p></li>
<li><p>Unit tests, new documentation system, continuous integration
framework and PyPI package</p></li>
<li><p>and even more minor features and bug fixes… (see
<a class="reference external" href="https://github.com/alexlancaster/pypop/blob/main/NEWS.md#100---2023-11-07">NEWS.md</a>).</p></li>
</ul>
</li>
<li><p>2023-11-04: release candidate 2 (1.0.0rc2) released. Fixes some
missing TSV output.</p></li>
<li><p>2023-11-01: release candidate 1 (1.0.0rc1) released.</p></li>
<li><p>2023-10-27: seventh beta pre-release 1.0.0b7, Previous <code class="docutils literal notranslate"><span class="pre">arm64</span></code>
issues have been resolved. Thanks to Owen Solberg for extensive
testing and debugging.</p></li>
<li><p>2023-10-13: fourth beta pre-release 1.0.0b4, . Although this
release contains packages that will install on <code class="docutils literal notranslate"><span class="pre">arm64</span></code>/M1
machines, these <code class="docutils literal notranslate"><span class="pre">arm64</span></code> packages should be considered as
<strong>alpha</strong>-only and are strictly for testing only. Please do not
use PyPop on M1 machines for any production analyses, until we
fix some underlying <code class="docutils literal notranslate"><span class="pre">arm64</span></code> numerical issues.</p></li>
<li><p>2023-10-10: second beta pre-release 1.0.0b2</p></li>
<li><p>2023-09-26: first beta pre-release 1.0.0b1</p></li>
<li><p>2023: ported to Python 3, pre-release alpha versions of 1.0.0
under development - no formal release yet.</p></li>
<li><p>2022: 0.7.0 binaries deprecated.</p></li>
<li><p>2020: pypop is no longer a Fedora package (to be replaced by PyPI package)</p></li>
<li><p>2017: all new development is now in GitHub</p></li>
</ul>
</div>
</div></blockquote>
<ul class="simple">
<li><p>See the <a class="reference internal" href="docs/guide-chapter-changes.html#pypop-release-history"><span class="std std-ref">PyPop Release History</span></a> in the <em>Python User Guide</em>
for even earlier history and full release notes.</p></li>
</ul>
</div>
<p id="guide-preface-2-start">PyPop is an object-oriented framework implemented in <a class="reference external" href="https://www.python.org/">Python</a>, but also contains C extensions for some
computationally intensive tasks. Output of analyses are stored in XML
format for maximum downstream flexibility. PyPop also has an internal
facility for additionally aggregating the output XML and generating
output tab-separated (TSV) files, as well as well as generating a
default plain text file summary for each population.</p>
<p>Although it can be run on any kind of genotype data, it has additional
support for analyzing population genotype with allelic nomenclature
from the human leukocyte antigen (HLA) region.  An outline of PyPop
can be found in our 2024 paper <span id="id2">(<a class="reference internal" href="docs/biblio.html#id13" title="Lancaster, A. K., Single, R. M., Mack, S. J., Sochat, V., Mariani, M. P., and Webster, G. D. PyPop: A mature open-source software pipeline for population genomics. Frontiers in Immunology, 2024. doi:10.3389/fimmu.2024.1378512.">Lancaster <em>et al.</em>, 2024</a>)</span>, and two
previous <a class="reference internal" href="#citing-pypop"><span class="std std-ref">papers</span></a>.</p>
<section id="installation-and-documentation">
<span id="guide-preface-2-end"></span><h2>Installation and documentation<a class="headerlink" href="#installation-and-documentation" title="Link to this heading">#</a></h2>
<p>Documentation, including instructions on <a class="reference internal" href="docs/guide-chapter-install.html"><span class="doc">installing</span></a>, <a class="reference internal" href="docs/guide-chapter-usage.html"><span class="doc">using</span></a>
and <a class="reference internal" href="docs/guide-chapter-instructions.html"><span class="doc">interpreting output of</span></a>
PyPop, is contained in the <a class="reference internal" href="docs/index.html#user-guide"><span class="std std-ref">PyPop User Guide</span></a>.  We
also provide an <a class="reference internal" href="api/index.html"><span class="doc">API Reference (1.3.1)</span></a> for that describes the
programmatic access to PyPop and for contributors to PyPop itself.</p>
<div class="admonition-contact-and-questions admonition">
<p class="admonition-title">Contact and questions</p>
<p>Please file all questions, support requests, and bug reports via
our <a class="reference external" href="https://github.com/alexlancaster/pypop/issues">GitHub issue tracker</a>. More details on
how to file bug reports can be found in our <a class="reference internal" href="docs/guide-chapter-contributing.html#guide-contributing-bug-report"><span class="std std-ref">contributors
chapter</span></a> of the <em>User Guide</em>.
Please don’t email developers individually.</p>
</div>
</section>
<section id="source-code">
<span id="id3"></span><h2>Source code<a class="headerlink" href="#source-code" title="Link to this heading">#</a></h2>
<p>PyPop is <a class="reference external" href="http://www.gnu.org/philosophy/free-sw.html">free software</a> (sometimes referred to
as <a class="reference external" href="https://opensource.org/">open source</a> software) and the
source code is released under the terms of the “copyleft” GNU General
Public License, or GPL (<a class="reference external" href="https://www.gnu.org/licenses/gpl.html">https://www.gnu.org/licenses/gpl.html</a>)
(specifically GPLv2, but any later version applies).  All source code is
available and maintained on our <a class="reference external" href="https://github.com/alexlancaster/pypop">GitHub website</a>.</p>
</section>
<section id="citing-pypop">
<h2>Citing PyPop<a class="headerlink" href="#citing-pypop" title="Link to this heading">#</a></h2>
<p>If you write a paper that uses PyPop in your analysis, please cite
<strong>both</strong>:</p>
<ul>
<li><p>our <a class="reference external" href="https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2024.1378512/full">2024 article</a>
in <em>Frontiers in Immunology</em>:</p>
<blockquote>
<div><p>Lancaster AK, Single RM, Mack SJ, Sochat V, Mariani MP, Webster
GD. (2024) “PyPop: A mature open-source software pipeline for
population genomics.”  <em>Front. Immunol.</em> <strong>15</strong>:1378512 doi:
<a class="reference external" href="https://doi.org/10.3389/fimmu.2024.1378512">10.3389/fimmu.2024.1378512</a></p>
</div></blockquote>
</li>
<li><p><strong>and</strong> a citation to the <a class="reference external" href="https://zenodo.org/records/10080667">Zenodo record</a> which includes a DOI for
the version of the software you used in your analyses. Citing this
record and DOI supports reproducibility by allowing researchers to
to determine the exact version of PyPop used in any particular
analysis. In addition, it allows retrieval of long-term software
source-code archives, independent of the original developers.</p>
<p>Here’s how to cite the correct version:</p>
<ul>
<li><p>If you have PyPop version 1.1.2 or later, currently installed, you
can run:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>pypop<span class="w"> </span>--citation
</pre></div>
</div>
<p>which outputs the Zenodo record citation in the simple “APA”
format (you can also choose from BibTeX, EndNote, RIS and other
formats, see the section on <a class="reference external" href="http://pypop.org/docs/guide-chapter-usage.html#command-line-interfaces">command-line interfaces</a>
in the <em>User Guide</em> for more details).</p>
</li>
<li><p>If you do not have PyPop installed, have a release of PyPop
earlier than 1.1.2, or otherwise want to obtain the DOI and
citation for specific versions, follow these steps:</p>
<ol class="arabic">
<li><p>First visit the DOI for the overall Zenodo record:
<a class="reference external" href="https://zenodo.org/doi/10.5281/zenodo.10080667">10.5281/zenodo.10080667</a>. This DOI
represents <strong>all versions</strong>, and will always resolve to the
latest one.</p></li>
<li><p>When you are viewing the record, look for the <strong>Versions</strong> box
in the right-sidebar. Here are listed all versions (including
older versions).</p></li>
<li><p>Select and click the version-specific DOI that matches the
specific version of PyPop that you used for your analysis.</p></li>
<li><p>Once you are visiting the Zenodo record for the specific version,
under the <strong>Citation</strong> box in the right-sidebar, select the
citation format you wish to use and click to copy the citation.
It will contain link to the version-specific DOI, and be of the
form:</p>
<blockquote>
<div><p>Lancaster, AK et al. (YYYY) “PyPop: Python for Population
Genomics” (Version X.Y.Z) [Computer
software]. Zenodo. <a class="reference external" href="https://doi.org/10.5281/zenodo.XXXXX">https://doi.org/10.5281/zenodo.XXXXX</a></p>
</div></blockquote>
</li>
</ol>
</li>
</ul>
<p>Note that citation metadata for the current Zenodo record is also
stored in <a class="reference external" href="https://github.com/alexlancaster/pypop/blob/main/CITATION.cff">CITATION.cff</a></p>
</li>
</ul>
<p id="guide-preface-3-start">Two previous papers are also available (but not necessary to cite):</p>
<ul class="simple">
<li><p>Lancaster AK, Single RM, Solberg OD, Nelson MP, Thomson G (2007)
“PyPop update - a software pipeline for large-scale multilocus
population genomics” <em>Tissue Antigens</em> 69 (s1), 192-197.  [<a class="reference external" href="http://dx.doi.org/10.1111/j.1399-0039.2006.00769.x">journal
page</a>,
<a class="reference external" href="http://pypop.org/tissue-antigens-lancaster-2007.pdf">preprint PDF (112 kB)</a>].</p></li>
<li><p>Lancaster A, Nelson MP, Single RM, Meyer D, Thomson G (2003)
“PyPop: a software framework for population genomics: analyzing
large-scale multi-locus genotype data”, in <em>Pacific Symposium on
Biocomputing</em> vol. 8:514-525 (edited by R B Altman. et al., World
Scientific, Singapore, 2003) [<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3891851/">PubMed Central</a>, <a class="reference external" href="http://pypop.org/psb-pypop.pdf">PDF
(344 kB)</a>].</p></li>
</ul>
<p>PyPop was originally developed for the analysis of data for the 13th
<a class="reference external" href="http://www.ihwg.org/">International Histocompatiblity Workshop and Conference</a> held in Seattle, Washington in 2002
(<span id="id4"><a class="reference internal" href="docs/biblio.html#id16" title="Meyer, D., Single, R. M., Mack, S. J., Lancaster, A. K., Nelson, M. P., Erlich, H. A., Fernández-Viña, M., and Thomson, G. Single locus polymorphism of classical HLA genes. In Hansen, J. A., editor, Immunobiology of the Human MHC: Proceedings of the 13th International Histocompatibility Workshop and Conference, volume I, pages 653–704. IHWG Press, Seattle, WA, 2007.">Meyer <em>et al.</em>, 2007</a></span>, Single <em>et al.</em>
<span id="id5"><a class="reference internal" href="docs/biblio.html#id20" title="Single, R. M., Meyer, D., Mack, S. J., Lancaster, A. K., Nelson, M. P., Erlich, H. A., Fernández-Viña, M., and Thomson, G. Haplotype frequencies and linkage disequilibrium among classical HLA genes. In Hansen, J. A., editor, Immunobiology of the Human MHC: Proceedings of the 13th International Histocompatibility Workshop and Conference, volume I, pages 705–746. IHWG Press, Seattle, WA, 2007.">2007</a>a, <a class="reference internal" href="docs/biblio.html#id21" title="Single, R. M., Meyer, D., and Thomson, G. Statistical methods for analysis of population genetic data. In Hansen, J. A., editor, Immunobiology of the Human MHC: Proceedings of the 13th International Histocompatibility Workshop and Conference, volume I, pages 518–522. IHWG Press, Seattle, WA, 2007.">2007</a>b</span>). For
more details on the design and technical details of PyPop, please
consult Lancaster <em>et al.</em>
<span id="id6">(<a class="reference internal" href="docs/biblio.html#id11" title="Lancaster, A., Nelson, M. P., Meyer, D., Single, R. M., and Thomson, G. PyPop: a software framework for population genomics: analyzing large-scale multi-locus genotype data. Pacific Symposium on Biocomputing, pages 514–525, 2003. URL: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3891851/ (accessed on 2024-01-04)">2003</a>, <a class="reference internal" href="docs/biblio.html#id14" title="Lancaster, A. K., Nelson, M. P., Single, R. M., Meyer, D., and Thomson, G. Software framework for the Biostatistics Core of the International Histocompatibility Working Group. In Hansen, J. A., editor, Immunobiology of the Human MHC: Proceedings of the 13th International Histocompatibility Workshop and Conference, volume I, pages 510–517. IHWG Press, Seattle, WA, 2007.">2007</a>a, <a class="reference internal" href="docs/biblio.html#id12" title="Lancaster, A. K., Single, R. M., Solberg, O. D., Nelson, M. P., and Thomson, G. PyPop update – a software pipeline for large-scale multilocus population genomics. Tissue Antigens, 69(s1):192–197, 2007. doi:10.1111/j.1399-0039.2006.00769.x.">2007</a>b, <a class="reference internal" href="docs/biblio.html#id13" title="Lancaster, A. K., Single, R. M., Mack, S. J., Sochat, V., Mariani, M. P., and Webster, G. D. PyPop: A mature open-source software pipeline for population genomics. Frontiers in Immunology, 2024. doi:10.3389/fimmu.2024.1378512.">2024</a>)</span>.</p>
</section>
<section id="acknowledgements">
<span id="guide-preface-3-end"></span><h2>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Link to this heading">#</a></h2>
<p id="guide-preface-acknowlegements-start">This work has benefited from the support of NIH grant AI49213 (13th
IHW) and NIH/NIAID Contract number HHSN266200400076C
N01-AI-40076. Thanks to Steven J. Mack, Kristie A. Mather, Steve G.E.
Marsh, Mark Grote and Leslie Louie for helpful comments and testing.</p>
</section>
<section id="supplementary-data-files">
<span id="popdata-files"></span><span id="guide-preface-acknowlegements-end"></span><h2>Supplementary data files<a class="headerlink" href="#supplementary-data-files" title="Link to this heading">#</a></h2>
<p><a class="reference external" href="popdata/">Population data files and online supporting materials</a> for
published studies listed in the <span id="id7">Solberg <em>et al.</em> (<a class="reference internal" href="docs/biblio.html#id24" title="Solberg, O. D., Mack, S. J., Lancaster, A. K., Single, R. M., Tsai, Y., Sanchez-Mazas, A., and Thomson, G. Balancing selection and heterogeneity across the classical human leukocyte antigen loci: A meta-analytic review of 497 population studies. Human Immunology, 69(7):443–464, 2008. doi:10.1016/j.humimm.2008.05.001.">2008</a>)</span> meta-analysis paper.</p>
</section>
<section id="immport-org">
<h2>ImmPort.org<a class="headerlink" href="#immport-org" title="Link to this heading">#</a></h2>
<p>PyPop is affiliated with <a class="reference external" href="https://ImmPort.org">https://ImmPort.org</a>, the Immunology Database and
Analysis Portal. The ImmPort system provides advanced information
technology support in the production, analysis, archiving, and
exchange of scientific data for the diverse community of life science
researchers supported by NIAID/DAIT. The development of the ImmPort
system was supported by the NIH/NIAID Bioinformatics Integration
Support Contract (BISC), Phase II.</p>
<div class="toctree-wrapper compound">
</div>
<div class="toctree-wrapper compound">
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="right-next"
       href="docs/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">PyPop User Guide</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025 PyPop contributors.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>