
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>PyPop.haplo &#8212; PyPop 1.4.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=9c3e77be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css?v=4ae1632d" />
    <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=deb9c837" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../_static/documentation_options.js?v=9172181d"></script>
    <script src="../../../_static/doctools.js?v=fd6eb6e6"></script>
    <script src="../../../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=1ae7504c"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'api/PyPop/haplo/index';</script>
    <link rel="icon" href="../../../_static/pypop-favicon.ico"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="PyPop.hardyweinberg" href="../hardyweinberg/index.html" />
    <link rel="prev" title="PyPop.filters" href="../filters/index.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="1.4.0" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
  <aside class="bd-header-announcement d-print-none d-none" aria-label="Announcement" data-pst-announcement-url="https://raw.githubusercontent.com/alexlancaster/pypop/refs/heads/main/website/_templates/announcement_banner.html"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/pypop-logo.png" class="logo__image only-light" alt="PyPop - Home"/>
    <img src="../../../_static/pypop-logo.png" class="logo__image only-dark pst-js-only" alt="PyPop - Home"/>
  
  
    <p class="title logo__title">PyPop</p>
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../docs/index.html">
    PyPop User Guide
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../../index.html">
    API Reference (1.4.0)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="http://pypop.org/popdata/">
    Solberg et al. (2008)
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/alexlancaster/pypop/" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../docs/index.html">
    PyPop User Guide
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../../index.html">
    API Reference (1.4.0)
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-external" href="http://pypop.org/popdata/">
    Solberg et al. (2008)
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/alexlancaster/pypop/" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../citation/index.html">PyPop.citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../command_line_interface/index.html">PyPop.command_line_interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="../datatypes/index.html">PyPop.datatypes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../filters/index.html">PyPop.filters</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">PyPop.haplo</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hardyweinberg/index.html">PyPop.hardyweinberg</a></li>
<li class="toctree-l1"><a class="reference internal" href="../homozygosity/index.html">PyPop.homozygosity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../parsers/index.html">PyPop.parsers</a></li>
<li class="toctree-l1"><a class="reference internal" href="../popaggregate/index.html">PyPop.popaggregate</a></li>
<li class="toctree-l1"><a class="reference internal" href="../popanalysis/index.html">PyPop.popanalysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../popmeta/index.html">PyPop.popmeta</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pypop/index.html">PyPop.pypop</a></li>
<li class="toctree-l1"><a class="reference internal" href="../randombinning/index.html">PyPop.randombinning</a></li>
<li class="toctree-l1"><a class="reference internal" href="../utils/index.html">PyPop.utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="../xslt/index.html">PyPop.xslt</a></li>
</ul>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../arlequin/index.html">PyPop.arlequin - DEPRECATED</a></li>
</ul>
</div>
</nav></div>
        <div class="sidebar-primary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#classes">Classes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-contents">Module Contents</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Haplo"><code class="docutils literal notranslate"><span class="pre">Haplo</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Emhaplofreq"><code class="docutils literal notranslate"><span class="pre">Emhaplofreq</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Emhaplofreq.serializeStart"><code class="docutils literal notranslate"><span class="pre">Emhaplofreq.serializeStart()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Emhaplofreq.serializeEnd"><code class="docutils literal notranslate"><span class="pre">Emhaplofreq.serializeEnd()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Emhaplofreq.estHaplotypes"><code class="docutils literal notranslate"><span class="pre">Emhaplofreq.estHaplotypes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Emhaplofreq.estLinkageDisequilibrium"><code class="docutils literal notranslate"><span class="pre">Emhaplofreq.estLinkageDisequilibrium()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Emhaplofreq.allPairwise"><code class="docutils literal notranslate"><span class="pre">Emhaplofreq.allPairwise()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Haplostats"><code class="docutils literal notranslate"><span class="pre">Haplostats</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Haplostats.serializeStart"><code class="docutils literal notranslate"><span class="pre">Haplostats.serializeStart()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Haplostats.serializeEnd"><code class="docutils literal notranslate"><span class="pre">Haplostats.serializeEnd()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Haplostats.estHaplotypes"><code class="docutils literal notranslate"><span class="pre">Haplostats.estHaplotypes()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.Haplostats.allPairwise"><code class="docutils literal notranslate"><span class="pre">Haplostats.allPairwise()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.HaploArlequin"><code class="docutils literal notranslate"><span class="pre">HaploArlequin</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.HaploArlequin.outputArlequin"><code class="docutils literal notranslate"><span class="pre">HaploArlequin.outputArlequin()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.HaploArlequin.runArlequin"><code class="docutils literal notranslate"><span class="pre">HaploArlequin.runArlequin()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#PyPop.haplo.HaploArlequin.genHaplotypes"><code class="docutils literal notranslate"><span class="pre">HaploArlequin.genHaplotypes()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item"><nav class="sidebar-indices-items" aria-labelledby="pst-indices-navigation-heading-2">
  <p id="pst-indices-navigation-heading-2" class="sidebar-indices-items__title" role="heading" aria-level="1">Indices</p>
  <ul class="indices-link">
        <li class="toctree-l1">
          <a class="reference internal"
             href="../../../genindex.html"
             accesskey="I">General Index</a>
        </li>
        <li class="toctree-l1">
          <a class="reference internal" href="../../../py-modindex.html">Python Module Index</a>
        </li>
  </ul>
</nav></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">API Reference (1.4.0)</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">PyPop.haplo</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="module-PyPop.haplo">
<span id="pypop-haplo"></span><h1>PyPop.haplo<a class="headerlink" href="#module-PyPop.haplo" title="Link to this heading">#</a></h1>
<p>Module for estimating haplotypes and linkage disequilibrium measures.</p>
<p>Currently there are two implementations: <a class="reference internal" href="#PyPop.haplo.Emhaplofreq" title="PyPop.haplo.Emhaplofreq"><code class="xref py py-class docutils literal notranslate"><span class="pre">Emhaplofreq</span></code></a> and
<a class="reference internal" href="#PyPop.haplo.Haplostats" title="PyPop.haplo.Haplostats"><code class="xref py py-class docutils literal notranslate"><span class="pre">Haplostats</span></code></a>.</p>
<section id="classes">
<h2>Classes<a class="headerlink" href="#classes" title="Link to this heading">#</a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#PyPop.haplo.Haplo" title="PyPop.haplo.Haplo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Haplo</span></code></a></p></td>
<td><p>Estimating haplotypes given genotype data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#PyPop.haplo.Emhaplofreq" title="PyPop.haplo.Emhaplofreq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Emhaplofreq</span></code></a></p></td>
<td><p>Haplotype and linkage disequilibrium (LD) estimation via emhaplofreq.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#PyPop.haplo.Haplostats" title="PyPop.haplo.Haplostats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Haplostats</span></code></a></p></td>
<td><p>Haplotype and LD estimation implemented via <code class="docutils literal notranslate"><span class="pre">haplo.stats</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#PyPop.haplo.HaploArlequin" title="PyPop.haplo.HaploArlequin"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HaploArlequin</span></code></a></p></td>
<td><p>Performs haplotype estimation via Arlequin.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="PyPop.haplo.Haplo">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">Haplo</span></span><a class="headerlink" href="#PyPop.haplo.Haplo" title="Link to this definition">#</a></dt>
<dd><p>Estimating haplotypes given genotype data.</p>
<p>This is abstract stub class (currently has no methods).</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PyPop.haplo.Emhaplofreq">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">Emhaplofreq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locusData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">untypedAllele</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'****'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testMode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Emhaplofreq" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PyPop.haplo.Haplo" title="PyPop.haplo.Haplo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Haplo</span></code></a></p>
<div class="graphviz"><img src="../../../_images/inheritance-bd1b326673edfbd722fbd584342ed1e972eb61a9.png" alt="Inheritance diagram of PyPop.haplo.Emhaplofreq" usemap="#inheritance3aa8b44eff" class="inheritance graphviz" /></div>
<map id="inheritance3aa8b44eff" name="inheritance3aa8b44eff">
<area shape="rect" id="node1" href="#PyPop.haplo.Emhaplofreq" target="_top" title="Haplotype and linkage disequilibrium (LD) estimation via emhaplofreq." alt="" coords="96,5,201,31"/>
<area shape="rect" id="node2" href="#PyPop.haplo.Haplo" target="_top" title="Estimating haplotypes given genotype data." alt="" coords="5,5,77,31"/>
</map><p>Haplotype and linkage disequilibrium (LD) estimation via emhaplofreq.</p>
<p>This is essentially a wrapper to a Python extension built on top
of the <code class="docutils literal notranslate"><span class="pre">emhaplofreq</span></code> command-line program.  Will refuse to
estimate haplotypes longer than that defined by <code class="docutils literal notranslate"><span class="pre">emhaplofreq</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locusData</strong> (<a class="reference internal" href="../utils/index.html#PyPop.utils.StringMatrix" title="PyPop.utils.StringMatrix"><em>StringMatrix</em></a>) – a StringMatrix</p></li>
<li><p><strong>untypedAllele</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – defaults to <code class="docutils literal notranslate"><span class="pre">****</span></code></p></li>
<li><p><strong>stream</strong> (<a class="reference internal" href="../utils/index.html#PyPop.utils.TextOutputStream" title="PyPop.utils.TextOutputStream"><em>TextOutputStream</em></a>) – output file</p></li>
<li><p><strong>testMode</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – default is <code class="docutils literal notranslate"><span class="pre">False</span></code></p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Emhaplofreq.serializeStart">
<span class="sig-name descname"><span class="pre">serializeStart</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Emhaplofreq.serializeStart" title="Link to this definition">#</a></dt>
<dd><p>Serialize start of XML output to the currently defined XML stream.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>must be paired with a subsequent <a class="reference internal" href="#PyPop.haplo.Emhaplofreq.serializeEnd" title="PyPop.haplo.Emhaplofreq.serializeEnd"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Emhaplofreq.serializeEnd()</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Emhaplofreq.serializeEnd">
<span class="sig-name descname"><span class="pre">serializeEnd</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Emhaplofreq.serializeEnd" title="Link to this definition">#</a></dt>
<dd><p>Serialize end of XML output to the currently defined XML stream.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>must be paired with a previous <a class="reference internal" href="#PyPop.haplo.Emhaplofreq.serializeStart" title="PyPop.haplo.Emhaplofreq.serializeStart"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Emhaplofreq.serializeStart()</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Emhaplofreq.estHaplotypes">
<span class="sig-name descname"><span class="pre">estHaplotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locusKeys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Emhaplofreq.estHaplotypes" title="Link to this definition">#</a></dt>
<dd><p>Estimate haplotypes for listed loci in <code class="docutils literal notranslate"><span class="pre">locusKeys</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locusKeys</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – <p>format is a string consisting of</p>
<ul>
<li><p>comma (<code class="docutils literal notranslate"><span class="pre">,</span></code>) separated haplotypes blocks for which to
estimate haplotypes</p></li>
<li><p>within each “block”, each locus is separated by colons
( <code class="docutils literal notranslate"><span class="pre">:</span></code> )</p></li>
</ul>
</p></li>
<li><p><strong>numInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of initial conditions to use</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<p><code class="docutils literal notranslate"><span class="pre">*DQA1:*DPB1,*DRB1:*DQB1</span></code>, means to estimate haplotypes for
<code class="docutils literal notranslate"><span class="pre">DQA1</span></code> and <code class="docutils literal notranslate"><span class="pre">DPB1</span></code> loci followed by estimation of haplotypes for
<code class="docutils literal notranslate"><span class="pre">DRB1</span></code> and <code class="docutils literal notranslate"><span class="pre">DQB1</span></code> loci.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Emhaplofreq.estLinkageDisequilibrium">
<span class="sig-name descname"><span class="pre">estLinkageDisequilibrium</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locusKeys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutationPrintFlag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numPermutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numPermuInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Emhaplofreq.estLinkageDisequilibrium" title="Link to this definition">#</a></dt>
<dd><p>Estimate linkage disequilibrium (LD) for listed loci.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locusKeys</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – see <a class="reference internal" href="#PyPop.haplo.Emhaplofreq.estHaplotypes" title="PyPop.haplo.Emhaplofreq.estHaplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">estHaplotypes()</span></code></a></p></li>
<li><p><strong>permutationPrintFlag</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – print all permutations (default <code class="docutils literal notranslate"><span class="pre">0</span></code>)</p></li>
<li><p><strong>numInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of initial conditions (default <code class="docutils literal notranslate"><span class="pre">None</span></code>)</p></li>
<li><p><strong>numPermutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of permutations (default <code class="docutils literal notranslate"><span class="pre">None</span></code>)</p></li>
<li><p><strong>numPermuInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of initial conditions for
each permutation (default <code class="docutils literal notranslate"><span class="pre">None</span></code>)</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<p>See <a class="reference internal" href="#PyPop.haplo.Emhaplofreq.estHaplotypes" title="PyPop.haplo.Emhaplofreq.estHaplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">estHaplotypes()</span></code></a> for an example that estimates LD</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Emhaplofreq.allPairwise">
<span class="sig-name descname"><span class="pre">allPairwise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">permutationPrintFlag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numPermutations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numPermuInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">haploSuppressFlag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">haplosToShow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Emhaplofreq.allPairwise" title="Link to this definition">#</a></dt>
<dd><p>Estimate pairwise statistics for a given set of loci.</p>
<p>Depending on the flags passed, this can be used to estimate
both LD (linkage disequilibrium) and HF (haplotype
frequencies), an optional permutation test on LD can be run.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>permutationPrintFlag</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – sets whether the result from
permutation output run will be included in the output XML.
Default: <code class="docutils literal notranslate"><span class="pre">0</span></code> (disabled).</p></li>
<li><p><strong>numInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – sets number of initial conditions before
performing the permutation test. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>numPermutations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – sets number of permutations that will
be performed. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>numPermuInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – sets number of initial conditions
tried per-permutation. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>haploSuppressFlag</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – sets whether haplotype information
is generated in the output. Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p></li>
<li><p><strong>haplosToShow</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>) – list of haplotypes to show in output</p></li>
<li><p><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – mode for haplotype output</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PyPop.haplo.Haplostats">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">Haplostats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locusData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">untypedAllele</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'****'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stream</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testMode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Haplostats" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PyPop.haplo.Haplo" title="PyPop.haplo.Haplo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Haplo</span></code></a></p>
<div class="graphviz"><img src="../../../_images/inheritance-32fdf120ff1f2a44fa9bde99ecfd5b3209a6eeeb.png" alt="Inheritance diagram of PyPop.haplo.Haplostats" usemap="#inheritance35c6a9e0c3" class="inheritance graphviz" /></div>
<map id="inheritance35c6a9e0c3" name="inheritance35c6a9e0c3">
<area shape="rect" id="node1" href="#PyPop.haplo.Haplo" target="_top" title="Estimating haplotypes given genotype data." alt="" coords="5,5,77,31"/>
<area shape="rect" id="node2" href="#PyPop.haplo.Haplostats" target="_top" title="Haplotype and LD estimation implemented via ``haplo.stats``." alt="" coords="96,5,188,31"/>
</map><p>Haplotype and LD estimation implemented via <code class="docutils literal notranslate"><span class="pre">haplo.stats</span></code>.</p>
<p>This is a wrapper to a portion of the <code class="docutils literal notranslate"><span class="pre">haplo.stats</span></code> R package.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locusData</strong> (<a class="reference internal" href="../utils/index.html#PyPop.utils.StringMatrix" title="PyPop.utils.StringMatrix"><em>StringMatrix</em></a>) – a StringMatrix</p></li>
<li><p><strong>untypedAllele</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – defaults to <code class="docutils literal notranslate"><span class="pre">****</span></code></p></li>
<li><p><strong>stream</strong> (<a class="reference internal" href="../utils/index.html#PyPop.utils.TextOutputStream" title="PyPop.utils.TextOutputStream"><em>TextOutputStream</em></a>) – output file</p></li>
<li><p><strong>testMode</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – default is <code class="docutils literal notranslate"><span class="pre">False</span></code></p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Haplostats.serializeStart">
<span class="sig-name descname"><span class="pre">serializeStart</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Haplostats.serializeStart" title="Link to this definition">#</a></dt>
<dd><p>Serialize start of XML output to currently defined XML stream.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>must be paired with a subsequent <a class="reference internal" href="#PyPop.haplo.Haplostats.serializeEnd" title="PyPop.haplo.Haplostats.serializeEnd"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplostats.serializeEnd()</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Haplostats.serializeEnd">
<span class="sig-name descname"><span class="pre">serializeEnd</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Haplostats.serializeEnd" title="Link to this definition">#</a></dt>
<dd><p>Serialize end of XML output to currently defined XML stream.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>must be paired with a previous <a class="reference internal" href="#PyPop.haplo.Haplostats.serializeStart" title="PyPop.haplo.Haplostats.serializeStart"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplostats.serializeStart()</span></code></a></p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Haplostats.estHaplotypes">
<span class="sig-name descname"><span class="pre">estHaplotypes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">locusKeys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">testMode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Haplostats.estHaplotypes" title="Link to this definition">#</a></dt>
<dd><p>Estimate haplotypes for listed loci in <code class="docutils literal notranslate"><span class="pre">locusKeys</span></code>.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">locusKeys</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>, assume entire matrix.  LD is
also estimated if there are <code class="docutils literal notranslate"><span class="pre">locusKeys</span></code> consisting of only
two loci.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>FIXME: this does <em>not</em> yet remove missing data before haplotype estimations</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>locusKeys</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – see <a class="reference internal" href="#PyPop.haplo.Emhaplofreq.estHaplotypes" title="PyPop.haplo.Emhaplofreq.estHaplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Emhaplofreq.estHaplotypes()</span></code></a> for format</p></li>
<li><p><strong>weight</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>) – set weights (default <code class="docutils literal notranslate"><span class="pre">None</span></code>, which sets all weights equal)</p></li>
<li><p><strong>control</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>) – a dictionary of control parameters</p></li>
<li><p><strong>numInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of initial conditions (default <code class="docutils literal notranslate"><span class="pre">None</span></code>)</p></li>
<li><p><strong>testMode</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.14)"><em>bool</em></a>) – run in test mode default is <code class="docutils literal notranslate"><span class="pre">False</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>multiple statistics</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.14)">tuple</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.Haplostats.allPairwise">
<span class="sig-name descname"><span class="pre">allPairwise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numInitCond</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.Haplostats.allPairwise" title="Link to this definition">#</a></dt>
<dd><p>Estimate pairwise statistics for all pairs of loci.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>weight</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>) – see <a class="reference internal" href="#PyPop.haplo.Haplostats.estHaplotypes" title="PyPop.haplo.Haplostats.estHaplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplostats.estHaplotypes()</span></code></a></p></li>
<li><p><strong>control</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.14)"><em>dict</em></a>) – see <a class="reference internal" href="#PyPop.haplo.Haplostats.estHaplotypes" title="PyPop.haplo.Haplostats.estHaplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplostats.estHaplotypes()</span></code></a></p></li>
<li><p><strong>numInitCond</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – see <a class="reference internal" href="#PyPop.haplo.Haplostats.estHaplotypes" title="PyPop.haplo.Haplostats.estHaplotypes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Haplostats.estHaplotypes()</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PyPop.haplo.HaploArlequin">
<span class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></span><span class="sig-name descname"><span class="pre">HaploArlequin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arpFilename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idCol</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefixCols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffixCols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">windowSize</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapOrder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">untypedAllele</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arlequinPrefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'arl_run'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.HaploArlequin" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference internal" href="#PyPop.haplo.Haplo" title="PyPop.haplo.Haplo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Haplo</span></code></a></p>
<div class="graphviz"><img src="../../../_images/inheritance-6ca56e97de9959ad57445a5a89ae82047e4b0cc9.png" alt="Inheritance diagram of PyPop.haplo.HaploArlequin" usemap="#inheritance62f3da738b" class="inheritance graphviz" /></div>
<map id="inheritance62f3da738b" name="inheritance62f3da738b">
<area shape="rect" id="node1" href="#PyPop.haplo.Haplo" target="_top" title="Estimating haplotypes given genotype data." alt="" coords="5,5,77,31"/>
<area shape="rect" id="node2" href="#PyPop.haplo.HaploArlequin" target="_top" title="Performs haplotype estimation via Arlequin." alt="" coords="96,5,211,31"/>
</map><p>Performs haplotype estimation via Arlequin.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 1.0.0.</span></p>
</div>
<p>Outputs Arlequin format data files and runtime info, also runs and
parses the resulting Arlequin data so it can be made available
programmatically to rest of Python framework.</p>
<p>Delegates all calls Arlequin to an internally instantiated
ArlequinBatch Python object called ‘batch’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>arpFilename</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – Arlequin filename (must have <code class="docutils literal notranslate"><span class="pre">.arp</span></code> file
extension)</p></li>
<li><p><strong>idCol</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – column in input file that contains the individual <code class="docutils literal notranslate"><span class="pre">id</span></code>.</p></li>
<li><p><strong>prefixCols</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of columns to ignore before allele data
starts</p></li>
<li><p><strong>suffixCols</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – number of columns to ignore after allele data
stops</p></li>
<li><p><strong>windowSize</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.14)"><em>int</em></a>) – size of sliding window</p></li>
<li><p><strong>mapOrder</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>) – list order of columns if different to column order in file
(defaults to order in file)</p></li>
<li><p><strong>untypedAllele</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – (defaults to <code class="docutils literal notranslate"><span class="pre">0</span></code>)</p></li>
<li><p><strong>arlequinPrefix</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.14)"><em>str</em></a>) – prefix for all Arlequin run-time files
(defaults to <code class="docutils literal notranslate"><span class="pre">arl_run</span></code>).</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.HaploArlequin.outputArlequin">
<span class="sig-name descname"><span class="pre">outputArlequin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.HaploArlequin.outputArlequin" title="Link to this definition">#</a></dt>
<dd><p>Outputs the specified <code class="docutils literal notranslate"><span class="pre">.arp</span></code> sample file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)"><em>list</em></a>) – list of strings containing the <code class="docutils literal notranslate"><span class="pre">.arp</span></code> sample file</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.HaploArlequin.runArlequin">
<span class="sig-name descname"><span class="pre">runArlequin</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.HaploArlequin.runArlequin" title="Link to this definition">#</a></dt>
<dd><p>Run the Arlequin haplotyping program.</p>
<p>Generates the expected <code class="docutils literal notranslate"><span class="pre">.txt</span></code> set-up files for Arlequin, then
forks a copy of <code class="docutils literal notranslate"><span class="pre">arlecore.exe</span></code>, which must be on <code class="docutils literal notranslate"><span class="pre">PATH</span></code> to
actually generate the haplotype estimates from the generated
<code class="docutils literal notranslate"><span class="pre">.arp</span></code> file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PyPop.haplo.HaploArlequin.genHaplotypes">
<span class="sig-name descname"><span class="pre">genHaplotypes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#PyPop.haplo.HaploArlequin.genHaplotypes" title="Link to this definition">#</a></dt>
<dd><p>Parses Arlequin output to retrieve estimated haplotypes.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>a list of the sliding <code class="docutils literal notranslate"><span class="pre">windows</span></code> which consists of tuples. Each tuple consists of:</p>
<ul class="simple">
<li><p>freqs (dict): dictionary entry (the haplotype-frequency) key-value pairs.</p></li>
<li><p>popName (str): population name (original <code class="docutils literal notranslate"><span class="pre">.arp</span></code> file prefix)</p></li>
<li><p>sampleCount (int): sample count (number of samples for that window)</p></li>
<li><p>lociList (list): ordered list of loci considered</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.14)">list</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../filters/index.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">PyPop.filters</p>
      </div>
    </a>
    <a class="right-next"
       href="../hardyweinberg/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">PyPop.hardyweinberg</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025 PyPop contributors.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 9.1.0.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>