
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>2. Getting started with PyPop &#8212; PyPop: Python for Population Genomics 0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/basic_mod.css?v=0.7.0-1" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/js/theme.js"></script>
    <script src="../_static/js/petite-vue.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="3. Interpreting PyPop output" href="guide-chapter-instructions.html" />
    <link rel="prev" title="1. Installing PyPop" href="guide-chapter-install.html" /> 
  </head><body data-dark_mode_code_blocks="true">

<div id="top_nav">
    
        <div id="notification_banner" data-banner-hiding="temporary" v-show="!permanentlyHidden">
    <p v-if="visible" id="content">PyPop 0.7.0 binaries are deprecated, please check the <a href="https://github.com/alexlancaster/pypop#readme">README</a> for installing pre-releases or for compiling from source</p>
    <a class="close" href="#" @click.prevent="toggleVisible()">[[ visible ? '&#x25B2; HIDE' : '&#x25BC; SHOW BANNER' ]]</a>
</div>

<script>
(function() {
    const topNav = document.querySelector('#top_nav')
    const notificationContent = document.querySelector('#notification_banner p#content').innerText
    const localStorageKey = 'readNotification'
    const bannerHiding = document.querySelector('#notification_banner').dataset['bannerHiding']
    const cssVariableName = '--navbarHeight'
    const rootElement = document.documentElement

    /*************************************************************************/
    // Local storage for remembering if the user has read the notification.

    function checkAlreadyRead() {
        return localStorage.getItem(localStorageKey) == notificationContent
    }

    function setRead() {
        localStorage.setItem(localStorageKey, notificationContent)
    }

    function clearRead() {
        localStorage.removeItem(localStorageKey)
    }

    const alreadyRead = checkAlreadyRead()
    const permanentlyHidden = alreadyRead && bannerHiding == 'permanent'

    /*************************************************************************/
    // Updating a CSS variable so other elements adjust to the nav bar height.

    function updateNavbarHeight() {
        // Only update it if the delta is significant. Otherwise it causes
        // unnecessary browser repaints.
        const documentStyles = getComputedStyle(rootElement)
        // We store the value in REM, so need to convert to pixels
        const currentValue = parseFloat(
            documentStyles.getPropertyValue(cssVariableName)
        ) * parseFloat(documentStyles.fontSize)

        const newValue = topNav.clientHeight

        if (newValue - 5 > currentValue) {
            console.log(`Updating ${cssVariableName} - overlapping`)
            rootElement.style.setProperty(
                cssVariableName, newValue + "px"
            );
        } else if (currentValue - newValue >= 30) {
            console.log(`Updating ${cssVariableName} - gap too large`)
            rootElement.style.setProperty(
                cssVariableName, newValue + "px"
            );
        }
    }

    /*************************************************************************/
    // After loading the page, and resizing the window, recalculate the nav bar
    // height.

    if (!permanentlyHidden) {
        // This height is approximately correct when there's a banner, so
        // shouldn't require any page reflow:
        rootElement.style.setProperty(
            '--navbarHeight', "5.5rem"
        );

        document.addEventListener("DOMContentLoaded", function() {
            updateNavbarHeight()
        });

        var interval = undefined

        window.addEventListener('resize', () => {
            if (interval) {
                clearTimeout(interval)
            }
            interval = setTimeout(() => {
                console.log("Finished resizing")
                updateNavbarHeight()
            }, 100)
        })
    }

    /*************************************************************************/

    PetiteVue.createApp({
        visible: !alreadyRead,
        permanentlyHidden: permanentlyHidden,
        bannerHiding: bannerHiding,
        $delimiters: ['[[', ']]'],
        toggleVisible() {
            this.visible = !this.visible

            if (this.visible) {
                clearRead()
            } else {
                setRead()
            }

            if (!this.visible && bannerHiding == 'permanent') {
                this.permanentlyHidden = true
            }

            // Run this after Vue has had time to update the DOM:
            setTimeout(
                updateNavbarHeight,
                0
            )
        }
    }).mount('#notification_banner')
})()
</script>
    

    <nav>
        
            
        

        <p id="toggle_sidebar">
            <a href="#" title="Toggle sidebar">|||</a>
        </p>
        <h1><a href="../index.html" title="Go to homepage">PyPop</a></h1>
            <a id="source_link" href="https://github.com/alexlancaster/pypop/">
    
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512">
            <path fill="white" d="M 244.8,8 C 106.1,8 0,113.3 0,252 c 0,110.9 69.8,205.8 169.5,239.2 12.8,2.3 17.3,-5.6 17.3,-12.1 0,-6.2 -0.3,-40.4 -0.3,-61.4 0,0 -70,15 -84.7,-29.8 0,0 -11.4,-29.1 -27.8,-36.6 0,0 -22.9,-15.7 1.6,-15.4 0,0 24.9,2 38.6,25.8 21.9,38.6 58.6,27.5 72.9,20.9 2.3,-16 8.8,-27.1 16,-33.7 -55.9,-6.2 -112.3,-14.3 -112.3,-110.5 0,-27.5 7.6,-41.3 23.6,-58.9 -2.6,-6.5 -11.1,-33.3 2.6,-67.9 20.9,-6.5 69,27 69,27 20,-5.6 41.5,-8.5 62.8,-8.5 21.3,0 42.8,2.9 62.8,8.5 0,0 48.1,-33.6 69,-27 13.7,34.7 5.2,61.4 2.6,67.9 16,17.7 25.8,31.5 25.8,58.9 0,96.5 -58.9,104.2 -114.8,110.5 9.2,7.9 17,22.9 17,46.4 0,33.7 -0.3,75.4 -0.3,83.6 0,6.5 4.6,14.4 17.3,12.1 C 428.2,457.8 496,362.9 496,252 496,113.3 383.5,8 244.8,8 Z"/>
        </svg>
    
</a>
        

        <a id="mode_toggle" href="#" @click.prevent="handleClick" :title="mode">
    <template v-if="mode == 'light'">
        <svg width="100%" height="100%" viewBox="0 0 79 80" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;"><g id="mode_light"><rect id="Bounds" x="0" y="-0" width="78.623" height="79.049" style="fill:none;"/><circle cx="39.311" cy="39.524" r="15.734" style="fill:#fff;"/><g id="beams"><g id="beam"><path id="beam1" serif:id="beam" d="M44.212,4.901c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.9,2.196 -4.9,4.901l-0,9.614c-0,2.705 2.196,4.901 4.9,4.901c2.705,0 4.901,-2.196 4.901,-4.901l0,-9.614Z" style="fill:#fff;"/></g><g id="beam2" serif:id="beam"><path id="beam3" serif:id="beam" d="M67.48,18.073c1.913,-1.912 1.913,-5.018 0,-6.931c-1.912,-1.912 -5.018,-1.912 -6.931,0l-6.798,6.799c-1.912,1.912 -1.912,5.018 0,6.931c1.913,1.912 5.018,1.912 6.931,-0l6.798,-6.799Z" style="fill:#fff;"/></g><g id="beam4" serif:id="beam"><path id="beam5" serif:id="beam" d="M25.728,61.108c1.912,-1.913 1.912,-5.018 -0,-6.931c-1.913,-1.913 -5.019,-1.913 -6.931,-0l-6.799,6.798c-1.912,1.913 -1.912,5.019 0,6.931c1.913,1.913 5.019,1.913 6.931,0l6.799,-6.798Z" style="fill:#fff;"/></g><g id="beam6" serif:id="beam"><path id="beam7" serif:id="beam" d="M60.682,54.177c-1.913,-1.913 -5.018,-1.913 -6.931,-0c-1.912,1.913 -1.912,5.018 0,6.931l6.798,6.798c1.913,1.913 5.019,1.913 6.931,0c1.913,-1.912 1.913,-5.018 0,-6.931l-6.798,-6.798Z" style="fill:#fff;"/></g><g id="beam8" serif:id="beam"><path id="beam9" serif:id="beam" d="M4.901,34.623c-2.705,0 -4.901,2.196 -4.901,4.901c0,2.705 2.196,4.901 4.901,4.901l9.614,0c2.705,0 4.901,-2.196 4.901,-4.901c0,-2.705 -2.196,-4.901 -4.901,-4.901l-9.614,0Z" style="fill:#fff;"/></g><g id="beam10" serif:id="beam"><path id="beam11" serif:id="beam" d="M44.212,64.534c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.9,2.196 -4.9,4.901l-0,9.614c-0,2.705 2.196,4.901 4.9,4.901c2.705,-0 4.901,-2.196 4.901,-4.901l0,-9.614Z" style="fill:#fff;"/></g><g id="beam12" serif:id="beam"><path id="beam13" serif:id="beam" d="M18.929,11.142c-1.912,-1.912 -5.018,-1.912 -6.931,0c-1.912,1.913 -1.912,5.019 0,6.931l6.799,6.799c1.912,1.912 5.018,1.912 6.931,-0c1.912,-1.913 1.912,-5.019 -0,-6.931l-6.799,-6.799Z" style="fill:#fff;"/></g><g id="beam14" serif:id="beam"><path id="beam15" serif:id="beam" d="M64.108,34.623c-2.705,0 -4.901,2.196 -4.901,4.901c-0,2.705 2.196,4.901 4.901,4.901l9.614,0c2.705,0 4.901,-2.196 4.901,-4.901c-0,-2.705 -2.196,-4.901 -4.901,-4.901l-9.614,0Z" style="fill:#fff;"/></g></g></g></svg>
    </template>

    <template v-if="mode == 'dark'">
        <svg width="100%" height="100%" viewBox="0 0 79 80" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;"><g id="mode_dark"><rect id="Bounds" x="0" y="-0" width="78.623" height="79.049" style="fill:none;"/><circle cx="39.311" cy="39.524" r="15.734" style="fill:#fff;"/><g id="beams"><g id="beam"><path id="beam1" serif:id="beam" d="M44.212,14.515c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,0 -4.901,2.196 -4.901,4.901c0,2.705 2.197,4.901 4.901,4.901c2.705,0 4.901,-2.196 4.901,-4.901Z" style="fill:#fff;"/></g><g id="beam2" serif:id="beam"><path id="beam3" serif:id="beam" d="M60.662,24.892c1.902,-1.902 1.902,-4.99 0,-6.892l-0.04,-0.039c-1.901,-1.902 -4.989,-1.902 -6.891,-0c-1.901,1.901 -1.901,4.989 0,6.891l0.04,0.04c1.902,1.901 4.989,1.901 6.891,-0Z" style="fill:#fff;"/></g><g id="beam4" serif:id="beam"><path id="beam5" serif:id="beam" d="M25.732,61.103c1.91,-1.91 1.91,-5.011 0,-6.921l-0.009,-0.01c-1.91,-1.91 -5.012,-1.91 -6.921,-0c-1.91,1.91 -1.91,5.011 -0,6.921l0.01,0.01c1.909,1.91 5.011,1.91 6.92,-0Z" style="fill:#fff;"/></g><g id="beam6" serif:id="beam"><path id="beam7" serif:id="beam" d="M60.672,54.167c-1.907,-1.907 -5.004,-1.907 -6.911,0l-0.02,0.02c-1.907,1.907 -1.907,5.004 0,6.911c1.907,1.907 5.004,1.907 6.911,-0l0.02,-0.02c1.907,-1.907 1.907,-5.004 0,-6.911Z" style="fill:#fff;"/></g><g id="beam8" serif:id="beam"><path id="beam9" serif:id="beam" d="M14.52,34.623c-2.702,0 -4.896,2.194 -4.896,4.896l0,0.01c0,2.702 2.194,4.896 4.896,4.896c2.702,0 4.896,-2.194 4.896,-4.896l-0,-0.01c-0,-2.702 -2.194,-4.896 -4.896,-4.896Z" style="fill:#fff;"/></g><g id="beam10" serif:id="beam"><path id="beam11" serif:id="beam" d="M44.212,64.534c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.901,2.196 -4.901,4.901c0,2.704 2.197,4.9 4.901,4.9c2.705,0 4.901,-2.196 4.901,-4.9Z" style="fill:#fff;"/></g><g id="beam12" serif:id="beam"><path id="beam13" serif:id="beam" d="M25.73,17.943c-1.911,-1.911 -5.015,-1.911 -6.926,0l-0.005,0.005c-1.911,1.911 -1.911,5.015 0,6.926c1.911,1.911 5.015,1.911 6.926,0l0.005,-0.005c1.911,-1.911 1.911,-5.014 -0,-6.926Z" style="fill:#fff;"/></g><g id="beam14" serif:id="beam"><path id="beam15" serif:id="beam" d="M64.098,34.623c-2.699,0 -4.891,2.192 -4.891,4.892l-0,0.019c-0,2.699 2.192,4.891 4.891,4.891c2.7,0 4.892,-2.192 4.892,-4.891l0,-0.019c0,-2.7 -2.192,-4.892 -4.892,-4.892Z" style="fill:#fff;"/></g></g></g></svg>
    </template>

    <template v-if="mode == 'darkest'">
        <svg width="100%" height="100%" viewBox="0 0 79 80" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xml:space="preserve" style="fill-rule:evenodd;clip-rule:evenodd;stroke-linejoin:round;stroke-miterlimit:2;"><g id="mode_darkest"><rect id="Bounds" x="0" y="-0" width="78.623" height="79.049" style="fill:none;"/><path d="M39.315,23.791c8.684,-0 15.734,7.05 15.734,15.733c0,8.684 -7.05,15.734 -15.734,15.734c-8.683,0 -15.733,-7.05 -15.733,-15.734c-0,-8.683 7.05,-15.733 15.733,-15.733Zm0,4.737c6.069,0 10.997,4.927 10.997,10.996c-0,6.069 -4.928,10.996 -10.997,10.996c-6.068,0 -10.996,-4.927 -10.996,-10.996c0,-6.069 4.928,-10.996 10.996,-10.996Z" style="fill:#fff;"/><g id="beams"><g id="beam"><path id="beam1" serif:id="beam" d="M44.216,14.515c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,0 -4.9,2.196 -4.9,4.901c-0,2.705 2.196,4.901 4.9,4.901c2.705,0 4.901,-2.196 4.901,-4.901Z" style="fill:#fff;"/></g><g id="beam2" serif:id="beam"><path id="beam3" serif:id="beam" d="M60.666,24.892c1.902,-1.902 1.902,-4.99 0,-6.892l-0.04,-0.039c-1.901,-1.902 -4.989,-1.902 -6.891,-0c-1.901,1.901 -1.901,4.989 0,6.891l0.04,0.04c1.902,1.901 4.99,1.901 6.891,-0Z" style="fill:#fff;"/></g><g id="beam4" serif:id="beam"><path id="beam5" serif:id="beam" d="M25.737,61.103c1.909,-1.91 1.909,-5.011 -0,-6.921l-0.01,-0.01c-1.91,-1.91 -5.011,-1.91 -6.921,-0c-1.91,1.91 -1.91,5.011 -0,6.921l0.01,0.01c1.91,1.91 5.011,1.91 6.921,-0Z" style="fill:#fff;"/></g><g id="beam6" serif:id="beam"><path id="beam7" serif:id="beam" d="M60.676,54.167c-1.907,-1.907 -5.004,-1.907 -6.911,0l-0.02,0.02c-1.907,1.907 -1.907,5.004 0,6.911c1.907,1.907 5.004,1.907 6.911,-0l0.02,-0.02c1.907,-1.907 1.907,-5.004 0,-6.911Z" style="fill:#fff;"/></g><g id="beam8" serif:id="beam"><path id="beam9" serif:id="beam" d="M14.524,34.623c-2.702,0 -4.896,2.194 -4.896,4.896l0,0.01c0,2.702 2.194,4.896 4.896,4.896c2.702,0 4.896,-2.194 4.896,-4.896l0,-0.01c0,-2.702 -2.194,-4.896 -4.896,-4.896Z" style="fill:#fff;"/></g><g id="beam10" serif:id="beam"><path id="beam11" serif:id="beam" d="M44.216,64.534c0,-2.705 -2.196,-4.901 -4.901,-4.901c-2.704,-0 -4.9,2.196 -4.9,4.901c-0,2.704 2.196,4.9 4.9,4.9c2.705,0 4.901,-2.196 4.901,-4.9Z" style="fill:#fff;"/></g><g id="beam12" serif:id="beam"><path id="beam13" serif:id="beam" d="M25.734,17.943c-1.911,-1.911 -5.015,-1.911 -6.926,0l-0.005,0.005c-1.911,1.911 -1.911,5.015 0,6.926c1.911,1.911 5.015,1.911 6.926,0l0.005,-0.005c1.911,-1.911 1.911,-5.014 0,-6.926Z" style="fill:#fff;"/></g><g id="beam14" serif:id="beam"><path id="beam15" serif:id="beam" d="M64.103,34.623c-2.7,0 -4.892,2.192 -4.892,4.892l-0,0.019c-0,2.699 2.192,4.891 4.892,4.891c2.699,0 4.891,-2.192 4.891,-4.891l0,-0.019c0,-2.7 -2.192,-4.892 -4.891,-4.892Z" style="fill:#fff;"/></g></g></g></svg>
    </template>
</a>

<script>
(function() {
    const LOCAL_STORAGE_KEY = 'piccoloThemeMode'

    var initialMode = localStorage.getItem(LOCAL_STORAGE_KEY)

    if (initialMode) {
        // Make sure the value in local storage is valid
        if (['light', 'dark', 'darkest'].indexOf(initialMode) == -1) {
            initialMode = 'light'
            localStorage.setItem(LOCAL_STORAGE_KEY, initialMode)
        }
    } else {
        // Check if the client prefers dark mode
        if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
            initialMode = 'dark'
        } else {
            initialMode = 'light'
        }
        localStorage.setItem(LOCAL_STORAGE_KEY, initialMode)
    }

    document.documentElement.dataset.mode = initialMode

    PetiteVue.createApp({
        'mode': initialMode,
        handleClick() {
            let currentMode = this.mode

            if (currentMode == 'light') {
                this.mode = 'dark'
            } else if (currentMode == 'dark') {
                this.mode = 'darkest'
            } else if (currentMode == 'darkest') {
                this.mode = 'light'
            }

            document.documentElement.dataset.mode = this.mode
            localStorage.setItem(LOCAL_STORAGE_KEY, this.mode)

            console.log(this.mode)
        }
    }).mount('#mode_toggle')
})()
</script>
            <p class="mobile_search_link">
                <a href="../search.html" title="Search">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 65 64" fill-rule="evenodd" stroke-linejoin="round" stroke-miterlimit="2">
                        <path d="M14.873 40.009c-2.315-3.943-3.642-8.532-3.642-13.429C11.231 11.91 23.141 0 37.811 0s26.58 11.91 26.58 26.58-11.91 26.58-26.58 26.58a26.44 26.44 0 0 1-14.277-4.161L9.739 62.794a3.12 3.12 0 0 1-4.413 0L.913 58.382c-1.217-1.218-1.217-3.196 0-4.413l13.96-13.96zM37.811 8.054c10.225 0 18.526 8.301 18.526 18.526s-8.301 18.526-18.526 18.526-18.526-8.301-18.526-18.526S27.586 8.054 37.811 8.054z" fill="#fff" />
                    </svg>
                </a>
            </p>
        

        <div class="searchbox_wrapper">
            
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
    </nav>
</div>

    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">PyPop User Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="guide-chapter-install.html">1. Installing PyPop</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">2. Getting started with PyPop</a></li>
<li class="toctree-l2"><a class="reference internal" href="guide-chapter-instructions.html">3. Interpreting PyPop output</a></li>
<li class="toctree-l2"><a class="reference internal" href="guide-chapter-contributing.html">4. Contributing to PyPop</a></li>
<li class="toctree-l2"><a class="reference internal" href="guide-chapter-changes.html">5. Authors and history</a></li>
<li class="toctree-l2"><a class="reference internal" href="licenses.html">6. Licenses</a></li>
<li class="toctree-l2"><a class="reference internal" href="biblio.html">7. References</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Supplementary data sets</span></p>
<ul>
<li class="toctree-l1"><a class="reference external" href="http://pypop.org/popdata/">Solberg et al. (2008)</a></li>
</ul>

        </div>
      </div>


    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="getting-started-with-pypop">
<h1><span class="section-number">2. </span>Getting started with PyPop<a class="headerlink" href="#getting-started-with-pypop" title="Permalink to this heading">¶</a></h1>
<section id="introduction">
<h2><span class="section-number">2.1. </span>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">¶</a></h2>
<p>You may use <strong class="program">PyPop</strong> to analyze many different kinds of data, including
allele-level genotype data (as in <a class="reference internal" href="#data-minimal-noheader-noids"><span class="std std-numref">Listing 2.1</span></a>), allele-level
frequency data (as in <a class="reference internal" href="#data-allelecount"><span class="std std-numref">Listing 2.6</span></a>),
microsatellite data, SNP data, and nucleotide and amino acid sequence
data.</p>
<p>As mentioned in the installation chapter, a minimal working example
of a <a class="reference external" href="https://github.com/alexlancaster/pypop/blob/main/tests/data/USAFEL-UchiTelle-small.pop">configuration file (.ini)</a>,
and a <a class="reference external" href="https://github.com/alexlancaster/pypop/blob/main/tests/data/minimal.ini">population file (.pop)</a>,
can be found by clicking the respective links.</p>
<p>There are two ways to run PyPop:</p>
<ul class="simple">
<li><p>interactive mode (where the program will prompt you to directly type
the input it needs); and</p></li>
<li><p>batch mode (where you supply all the command line options the program
needs).</p></li>
</ul>
<p>For the most simplest application of PyPop, where you wish to analyze
a single population, the interactive mode is the simplest to use. We
will describe this mode first then describe batch mode.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The following assumes you have already <a class="reference internal" href="guide-chapter-install.html#installing-pypop"><span class="std std-ref">installed PyPop</span></a>, done any <a class="reference internal" href="guide-chapter-install.html#post-install-path-adjustments"><span class="std std-ref">post-install adjustments</span></a> needed for your platform, and
verified that you can run the main commands (see the
<a class="reference internal" href="guide-chapter-install.html#examples"><span class="std std-ref">Examples</span></a> section).</p>
</div>
<section id="interactive-mode">
<h3><span class="section-number">2.1.1. </span>Interactive mode<a class="headerlink" href="#interactive-mode" title="Permalink to this heading">¶</a></h3>
<p>To run PyPop in interactive mode, with a minimal “GUI”, on Windows or
MacOS, you can directly click on the <code class="docutils literal notranslate"><span class="pre">pypop-interactive</span></code> file in the
directory where the scripts were installed (see <a class="reference internal" href="guide-chapter-install.html#post-install-path-adjustments"><span class="std std-ref">post-install
adjustments</span></a>).</p>
<p>You can also type <code class="docutils literal notranslate"><span class="pre">pypop-interactive</span></code> after starting a console
application on all platforms (on MacOS and GNU/Linux, this is normally
the <strong class="program">Terminal</strong> program, on Windows, it’s <strong class="program">Command
prompt</strong>).</p>
<p>In most cases, this will launch a console with the following:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>PyPop: Python for Population Genomics (1.0.0a15)
Copyright (C) 2003-2006 Regents of the University of California
Copyright (C) 2007-2023 PyPop team.
This is free software.  There is NO warranty; not even for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You may redistribute copies of PyPop under the terms of the GNU
General Public License.  For more information about these
matters, see the file named COPYING.

Select both an &#39;.ini&#39; configuration file and a &#39;.pop&#39; file via the
system file dialog.
</pre></div>
</div>
<p>Following this:</p>
<ol class="arabic">
<li><p>the system file dialog will appear prompting you to select an
<code class="docutils literal notranslate"><span class="pre">.ini</span></code> <a class="reference internal" href="#guide-usage-configfile"><span class="std std-ref">configuration file</span></a>.</p></li>
<li><p>a second system file dialog will prompt you for a <code class="docutils literal notranslate"><span class="pre">.pop</span></code>
<a class="reference internal" href="#guide-usage-datafile"><span class="std std-ref">data file</span></a>.</p></li>
<li><p>after both files are selected the console will display the
processing of the file:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>PyPop is processing sample.pop ...
PyPop run complete!
XML output(s) can be found in: [&#39;sample-out.xml&#39;]
Plain text output(s) can be found in: [&#39;sample-out.txt&#39;]
<span class="hll">Press Enter to continue...
</span></pre></div>
</div>
</li>
<li><p>when the run is completed, the last line will prompt you to press
<code class="docutils literal notranslate"><span class="pre">Enter</span></code> to leave the console window (highlighted above).</p></li>
</ol>
<p>If the system file GUI dialog does not appear (e.g. if you are running
on a terminal without a display), it will fall-back to text-mode entry
for the files, where you need to type the full (either relative or
absolute) paths to the files. The output should resemble:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>PyPop: Python for Population Genomics (1.0.0a15)
Copyright (C) 2003-2006 Regents of the University of California
Copyright (C) 2007-2023 PyPop team.
This is free software.  There is NO warranty; not even for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You may redistribute copies of PyPop under the terms of the GNU
General Public License.  For more information about these
matters, see the file named COPYING.

To accept the default in brackets for each filename, simply press
return for each prompt.

<span class="hll">Please enter config filename [config.ini]: sample.ini
</span><span class="hll">Please enter population filename [no default]: sample.pop
</span>PyPop is processing sample.pop ...
PyPop run complete!
XML output(s) can be found in: [&#39;sample-out.xml&#39;]
Plain text output(s) can be found in: [&#39;sample-out.txt&#39;]
Press Enter to continue...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Some messages with the prefix “LOG:” may appear during the console
operation.  They are informational only and do not indicate
improper operation of the program.</p>
</div>
<p>In both cases you should substitute the names of your own
configuration (e.g., <code class="file docutils literal notranslate"><span class="pre">config.ini</span></code>) and population file (e.g.,
<code class="file docutils literal notranslate"><span class="pre">Guatemalan.pop</span></code>) for <code class="file docutils literal notranslate"><span class="pre">sample.ini</span></code> and <code class="file docutils literal notranslate"><span class="pre">sample.pop</span></code>
(highlighted above). The formats for these files are described in the
sections on the <a class="reference internal" href="#guide-usage-datafile"><span class="std std-ref">data file</span></a> and
<a class="reference internal" href="#guide-usage-configfile"><span class="std std-ref">configuration file</span></a>, below.</p>
</section>
<section id="batch-mode">
<h3><span class="section-number">2.1.2. </span>Batch mode<a class="headerlink" href="#batch-mode" title="Permalink to this heading">¶</a></h3>
<p>To run PyPop in the more common “batch mode”, you can run PyPop from
the console (as noted above, on Windows: open <strong class="program">Command
prompt</strong>, aka a “DOS shell”; on MacOS or GNU/Linux: open the
<strong class="program">Terminal</strong> application). Change to a directory where your
<code class="docutils literal notranslate"><span class="pre">.pop</span></code> file is located, and type the command:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>pypop Guatemalan.pop
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If your system administrator has installed PyPop the name of the
script may be renamed to something different.</p>
</div>
<p>Batch mode assumes two things: that you have a file called
<code class="file docutils literal notranslate"><span class="pre">config.ini</span></code> in your current folder and that you also have your
population file is in the current folder, otherwise you will need to
supply the full path to the file. You can specify a particular
configuration file for PyPop to use, by supplying the <code class="docutils literal notranslate"><span class="pre">-c</span></code> option as
follows:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>pypop -c newconfig.ini Guatemalan.pop
</pre></div>
</div>
<p>You may also redirect the output to a different directory (which must
already exist) by using the <code class="docutils literal notranslate"><span class="pre">-o</span></code> option:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>pypop -c newconfig.ini -o altdir Guatemalan.pop
</pre></div>
</div>
<p>Please see <a class="reference internal" href="#guide-pypop-cli"><span class="std std-ref">pypop usage</span></a> for the full list of command-line
options.</p>
</section>
<section id="what-happens-when-you-run-pypop">
<span id="guide-usage-intro-run-details"></span><h3><span class="section-number">2.1.3. </span>What happens when you run PyPop?<a class="headerlink" href="#what-happens-when-you-run-pypop" title="Permalink to this heading">¶</a></h3>
<p>The most common types of analysis will involve the editing of your
<code class="file docutils literal notranslate"><span class="pre">config.ini</span></code> file to suit your data (see <a class="reference external" href="guide-usage-configfile">The configuration
file</a>) followed by the selection of either
the interactive or batch mode described above. If your input
configuration file is <code class="file docutils literal notranslate"><em><span class="pre">configfilename</span></em></code> and your population file name
is <code class="file docutils literal notranslate"><em><span class="pre">popfilename</span></em><span class="pre">.txt</span></code> the initial output will be generated quickly, but
your the PyPop execution will not be finished until the text output file
named <code class="file docutils literal notranslate"><em><span class="pre">popfilename</span></em><span class="pre">-out.txt</span></code> has been created. A successful run will
produce two output files: <code class="file docutils literal notranslate"><em><span class="pre">popfilename</span></em><span class="pre">-out.xml</span></code>,
<code class="file docutils literal notranslate"><em><span class="pre">popfilename</span></em><span class="pre">-out.txt</span></code>. A third output file will be created if you are
using the Anthony Nolan HLA filter option for HLA data to check your
input for valid/known HLA alleles: <code class="file docutils literal notranslate"><span class="pre">popfilename-filter.xml</span></code>).</p>
<p>The <code class="file docutils literal notranslate"><span class="pre">popfilename-out.xml</span></code> file is the primary output created by
PyPop and the human-readable <code class="file docutils literal notranslate"><span class="pre">popfilename-out.txt</span></code> file is a
summary of the complete XML output. The XML output can be further
transformed into plain text TSV files, either directly via <code class="docutils literal notranslate"><span class="pre">pypop</span></code>
if invoked on multiple input files (using the <code class="docutils literal notranslate"><span class="pre">--enable-tsv</span></code> option,
see <a class="reference internal" href="#guide-pypop-cli"><span class="std std-ref">pypop usage</span></a>), or via the <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> tool that
aggregates results from different <code class="docutils literal notranslate"><span class="pre">pypop</span></code> runs (see
<a class="reference internal" href="#guide-usage-popmeta"><span class="std std-ref">Using popmeta to aggregate results</span></a>).</p>
<p>A typical PyPop run might take anywhere from a few of minutes to a few
hours, depending on how large your data set is and who else is using the
system at the same time. Note that performing the
<code class="docutils literal notranslate"><span class="pre">allPairwiseLDWithPermu</span></code> test may take several <strong>days</strong> if you have
highly polymorphic loci in your data set.</p>
</section>
</section>
<section id="using-popmeta-to-aggregate-results">
<span id="guide-usage-popmeta"></span><h2><span class="section-number">2.2. </span>Using <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> to aggregate results<a class="headerlink" href="#using-popmeta-to-aggregate-results" title="Permalink to this heading">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> script (<code class="docutils literal notranslate"><span class="pre">popmeta.bat</span></code> on Windows, <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> on
GNU/Linux) can aggregate results from a number of output XML files
from individual populations into a set of tab-separated (TSV) files
containing summary statistics via customized XSLT (eXtensible
Stylesheet Language for Transformations) stylesheets.  These TSV files
can be directly imported into a spreadsheet or statistical software
(e.g., <strong class="program">R</strong>, <strong class="program">SAS</strong>).  In addition, there is some
preliminary support for export into other formats, such as the
population genetic software (e.g., <strong class="program">PHYLIP</strong>).</p>
<p>Here is an example of a <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> run, following on from the XML outputs
generated in similar fashion in the previous <code class="docutils literal notranslate"><span class="pre">pypop</span></code> runs:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>popmeta -o altdir Guatemalan-out.xml NorthAmerican-out.xml
</pre></div>
</div>
<p>This will generate a number of <code class="docutils literal notranslate"><span class="pre">.dat</span></code> files, including
<code class="file docutils literal notranslate"><span class="pre">1-locus-allele.dat</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It’s highly recommended to use the <code class="docutils literal notranslate"><span class="pre">-o</span></code> option to save the output
in a separate subdirectory, as the output <code class="docutils literal notranslate"><span class="pre">.dat</span></code> files have
fixed names, and will overwrite any files in the local directory with the
same name).  See <a class="reference internal" href="#guide-popmeta-cli"><span class="std std-ref">popmeta usage</span></a> for the full list of
options.</p>
</div>
<p>Note that a similar effect can be achieved directly from a <code class="docutils literal notranslate"><span class="pre">pypop</span></code>
run (assuming that the configuration file can be used for both
<code class="docutils literal notranslate"><span class="pre">.pop</span></code> population files), by invoking <code class="docutils literal notranslate"><span class="pre">pypop</span></code> with the
<code class="docutils literal notranslate"><span class="pre">--enable-tsv</span></code> option:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>pypop -c newconfig.ini -o altdir Guatemalan.pop NorthAmerican.pop --enable-tsv
</pre></div>
</div>
</section>
<section id="command-line-interfaces">
<h2><span class="section-number">2.3. </span>Command-line interfaces<a class="headerlink" href="#command-line-interfaces" title="Permalink to this heading">¶</a></h2>
<p>Described below is the usage for both programs, including a full list
of the current command-line options and arguments.  Note that you can
also view this full list of options from the program itself by
supplying the <code class="docutils literal notranslate"><span class="pre">--help</span></code> option, i.e. <code class="docutils literal notranslate"><span class="pre">pypop</span> <span class="pre">--help</span></code>, or <code class="docutils literal notranslate"><span class="pre">popmeta</span>
<span class="pre">--help</span></code>, respectively.</p>
<section id="pypop-usage">
<span id="guide-pypop-cli"></span><h3><span class="section-number">2.3.1. </span><code class="docutils literal notranslate"><span class="pre">pypop</span></code> usage<a class="headerlink" href="#pypop-usage" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pypop</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUTDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">V</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CONFIG</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">XSLFILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">enable</span><span class="o">-</span><span class="n">ihwg</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">enable</span><span class="o">-</span><span class="n">phylip</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span> <span class="n">FILELIST</span><span class="p">]</span>
             <span class="p">[</span><span class="n">POPFILE</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="options-for-pypop">
<h4>Options for pypop<a class="headerlink" href="#options-for-pypop" title="Permalink to this heading">¶</a></h4>
<dl class="option-list">
<dt><kbd>-o, --outputdir</kbd></dt>
<dd><p>put output in directory OUTPUTDIR</p>
</dd>
<dt><kbd>-V, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd>-c, --config</kbd></dt>
<dd><p>select config file</p>
<p>Default: “config.ini”</p>
</dd>
<dt><kbd>-m, --testmode</kbd></dt>
<dd><p>run PyPop in test mode for unit testing</p>
</dd>
<dt><kbd>-d, --debug</kbd></dt>
<dd><p>enable debugging output (overrides config file setting)</p>
</dd>
<dt><kbd>-x, --xsl</kbd></dt>
<dd><p>override the default XSLT translation with XSLFILE</p>
</dd>
</dl>
</section>
<section id="tsv-output-options">
<h4>TSV output options<a class="headerlink" href="#tsv-output-options" title="Permalink to this heading">¶</a></h4>
<p>Note that <code class="docutils literal notranslate"><span class="pre">--enable-</span></code> flags only valid if <code class="docutils literal notranslate"><span class="pre">--enable-tsv</span></code>/<code class="docutils literal notranslate"><span class="pre">-t</span></code> selected</p>
<dl class="option-list">
<dt><kbd>-t, --enable-tsv</kbd></dt>
<dd><p>generate TSV output files (aka run ‘popmeta’)</p>
</dd>
<dt><kbd>--enable-ihwg</kbd></dt>
<dd><p>enable 13th IWHG workshop populationdata default headers</p>
</dd>
<dt><kbd>--enable-phylip</kbd></dt>
<dd><p>enable generation of PHYLIP <code class="docutils literal notranslate"><span class="pre">.phy</span></code> files</p>
</dd>
</dl>
</section>
<section id="mutually-exclusive-input-options">
<h4>Mutually exclusive input options<a class="headerlink" href="#mutually-exclusive-input-options" title="Permalink to this heading">¶</a></h4>
<dl class="option-list">
<dt><kbd>-i, --interactive</kbd></dt>
<dd><p>run in interactive mode, prompting user for file names</p>
</dd>
<dt><kbd>-f, --filelist</kbd></dt>
<dd><p>file containing list of files (one per line) to process
(mutually exclusive with supplying POPFILEs)</p>
</dd>
<dt><kbd>POPFILE</kbd></dt>
<dd><p>input population (<code class="docutils literal notranslate"><span class="pre">.pop</span></code>) file(s)</p>
<p>Default: []</p>
</dd>
</dl>
</section>
</section>
<section id="popmeta-usage">
<span id="guide-popmeta-cli"></span><h3><span class="section-number">2.3.2. </span><code class="docutils literal notranslate"><span class="pre">popmeta</span></code> usage<a class="headerlink" href="#popmeta-usage" title="Permalink to this heading">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">popmeta</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUTDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">V</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">disable</span><span class="o">-</span><span class="n">tsv</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">meta</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">x</span> <span class="n">XSLDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">enable</span><span class="o">-</span><span class="n">ihwg</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">enable</span><span class="o">-</span><span class="n">phylip</span> <span class="o">|</span> <span class="o">-</span><span class="n">b</span> <span class="n">FACTOR</span><span class="p">]</span>
               <span class="n">XMLFILE</span> <span class="p">[</span><span class="n">XMLFILE</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<section id="positional-arguments">
<h4>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this heading">¶</a></h4>
<dl class="option-list">
<dt><kbd>XMLFILE</kbd></dt>
<dd><p>XML (<code class="docutils literal notranslate"><span class="pre">.xml</span></code>) file(s) generated by pypop runs</p>
<p>Default: []</p>
</dd>
</dl>
</section>
<section id="options-for-popmeta">
<h4>Options for popmeta<a class="headerlink" href="#options-for-popmeta" title="Permalink to this heading">¶</a></h4>
<dl class="option-list">
<dt><kbd>-o, --outputdir</kbd></dt>
<dd><p>put output in directory OUTPUTDIR</p>
</dd>
<dt><kbd>-V, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
<dt><kbd>--disable-tsv</kbd></dt>
<dd><p>disable generation of <code class="docutils literal notranslate"><span class="pre">.dat</span></code> TSV files</p>
</dd>
<dt><kbd>--output-meta</kbd></dt>
<dd><p>dump the meta output file to stdout, ignore xslt file</p>
</dd>
<dt><kbd>-x, --xsldir</kbd></dt>
<dd><p>use specified directory to find meta XSLT</p>
</dd>
<dt><kbd>--enable-ihwg</kbd></dt>
<dd><p>enable 13th IWHG workshop populationdata default headers</p>
</dd>
</dl>
</section>
<section id="mutually-exclusive-popmeta-options">
<h4>Mutually exclusive popmeta options<a class="headerlink" href="#mutually-exclusive-popmeta-options" title="Permalink to this heading">¶</a></h4>
<dl class="option-list">
<dt><kbd>--enable-phylip</kbd></dt>
<dd><p>enable generation of PHYLIP <code class="docutils literal notranslate"><span class="pre">.phy</span></code> files</p>
</dd>
<dt><kbd>-b, --batchsize</kbd></dt>
<dd><p>process in batches of size total/FACTOR rather than all at once, by default do separately (batchsize=0)</p>
<p>Default: 0</p>
</dd>
</dl>
</section>
</section>
</section>
<section id="the-data-file">
<span id="guide-usage-datafile"></span><h2><span class="section-number">2.4. </span>The data file<a class="headerlink" href="#the-data-file" title="Permalink to this heading">¶</a></h2>
<section id="sample-files">
<h3><span class="section-number">2.4.1. </span>Sample files<a class="headerlink" href="#sample-files" title="Permalink to this heading">¶</a></h3>
<p>Data can be input either as genotypes, or in an allele count format,
depending on the format of your data.</p>
<p>As you will see in the following examples, population files begin with
header information. In the simplest case, the first line contains the
column headers for the genotype, allele count, or, sequence information
from the population. If the file contains a population data-block, then
the first line consists of headers identifying the data on the second
line, and the third line contains the column headers for the genotype or
allele count information.</p>
<p>Note that for genotype data, each locus corresponds to two columns in
the population file. The locus name must repeated, with a suffix such as
<code class="docutils literal notranslate"><span class="pre">_1</span></code>, <code class="docutils literal notranslate"><span class="pre">_2</span></code> (the default) or <code class="docutils literal notranslate"><span class="pre">_a</span></code>, <code class="docutils literal notranslate"><span class="pre">_b</span></code> and must match the format
defined in the <code class="file docutils literal notranslate"><span class="pre">config.ini</span></code> (see
<a class="reference internal" href="#validsamplefields"><span class="std std-ref">validSampleFields</span></a>). Although PyPop needs this
distinction to be made, phase is NOT assumed, and if known it is
ignored.</p>
<p><a class="reference internal" href="#config-minimal-example"><span class="std std-numref">Listing 2.7</span></a> shows the relevant lines for the
configuration to read in the data shown in
<a class="reference internal" href="#data-minimal-noheader-noids"><span class="std std-numref">Listing 2.1</span></a> through to <a class="reference internal" href="#data-allelecount"><span class="std std-numref">Listing 2.6</span></a>.</p>
<div class="literal-block-wrapper docutils container" id="data-minimal-noheader-noids">
<div class="code-block-caption"><span class="caption-number">Listing 2.1 </span><span class="caption-text">Multi-locus allele-level genotype data</span><a class="headerlink" href="#data-minimal-noheader-noids" title="Permalink to this code">¶</a></div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>a_1   a_2   c_1   c_2   b_1   b_2
****  ****  0102  02025 1301  18012
0101  0201  0307  0605  1401  39021
0210  03012 0712  0102  1520  1301
0101  0218  0804  1202  35091 4005
2501  0201  1507  0307  51013 1401
0210  3204  1801  0102  78021 1301
03012 3204  1507  0605  51013 39021
</pre></div>
</div>
</div>
<p>This is an example of the simplest kind of data file.</p>
<div class="literal-block-wrapper docutils container" id="data-minimal-noheader">
<div class="code-block-caption"><span class="caption-number">Listing 2.2 </span><span class="caption-text">Multi-locus allele-level HLA genotype data with sample information</span><a class="headerlink" href="#data-minimal-noheader" title="Permalink to this code">¶</a></div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>populat    id        a_1   a_2   c_1   c_2   b_1   b_2
UchiTelle  UT900-23  ****  ****  0102  02025 1301  18012
UchiTelle  UT900-24  0101  0201  0307  0605  1401  39021
UchiTelle  UT900-25  0210  03012 0712  0102  1520  1301
UchiTelle  UT900-26  0101  0218  0804  1202  35091 4005
UchiTelle  UT910-01  2501  0201  1507  0307  51013 1401
UchiTelle  UT910-02  0210  3204  1801  0102  78021 1301
UchiTelle  UT910-03  03012 3204  1507  0605  51013 39021
</pre></div>
</div>
</div>
<p>This example shows a data file which has non-allele data in some
columns, here we have population (<code class="docutils literal notranslate"><span class="pre">populat</span></code>) and sample identifiers
(<code class="docutils literal notranslate"><span class="pre">id</span></code>).</p>
<div class="literal-block-wrapper docutils container" id="data-hla">
<div class="code-block-caption"><span class="caption-number">Listing 2.3 </span><span class="caption-text">Multi-locus allele-level HLA genotype data with sample and header information</span><a class="headerlink" href="#data-hla" title="Permalink to this code">¶</a></div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>labcode method              ethnic  contin  collect        latit           longit
USAFEL  12th Workshop SSOP  Telle   NW Asia Targen Village 41 deg 12 min N 94 deg 7 min E
populat     id         a_1     a_2     c_1     c_2     b_1     b_2
UchiTelle   UT900-23   ****    ****    0102    02025   1301    18012
UchiTelle   UT900-24   0101    0201    0307    0605    1401    39021
UchiTelle   UT900-25   0210    03012   0712    0102    1520    1301
UchiTelle   UT900-26   0101    0218    0804    1202    35091   4005
UchiTelle   UT910-01   2501    0201    1507    0307    51013   1401
UchiTelle   UT910-02   0210    3204    1801    0102    78021   1301
UchiTelle   UT910-03   03012   3204    1507    0605    51013   39021
</pre></div>
</div>
</div>
<p>This is an example of a data file which is identical to
<a class="reference internal" href="#data-minimal-noheader"><span class="std std-numref">Listing 2.2</span></a>, but which includes population level
information.</p>
<div class="literal-block-wrapper docutils container" id="data-hla-microsat">
<div class="code-block-caption"><span class="caption-number">Listing 2.4 </span><span class="caption-text">Multi-locus allele-level HLA genotype and microsatellite genotype data with header information</span><a class="headerlink" href="#data-hla-microsat" title="Permalink to this code">¶</a></div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>labcode ethnic  complex
USAFEL  ****    0
populat    id      drb1_1  drb1_2  dqb1_1  dqb1_2  d6s2222_1  d6s2222_2
UchiTelle  HJK_2   01      0301    0201     0501    249        249
UchiTelle  HJK_1   0301    0301    0201     0201    249        249
UchiTelle  HJK_3   01      0301    0201     0501    249        249
UchiTelle  HJK_4   01      0301    0201     0501    249        249
UchiTelle  MYU_2   02      0401    0302     0602    247        249
UchiTelle  MYU_1   0301    0301    0201     0201    247        249
UchiTelle  MYU_3   0301    0401    0201     0302    249        249
UchiTelle  MYU_4   0301    0401    0201     0302    247        249
</pre></div>
</div>
</div>
<p>This example mixes different kinds of data: HLA allele data (from DRB1
and DQB1 loci) with microsatellite data (locus D6S2222).</p>
<div class="literal-block-wrapper docutils container" id="data-nucleotide">
<div class="code-block-caption"><span class="caption-number">Listing 2.5 </span><span class="caption-text">Sequence genotype data with header information</span><a class="headerlink" href="#data-nucleotide" title="Permalink to this code">¶</a></div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>labcode file
BLOGGS  C_New
popName ID       TGFB1cdn10(1) TGFB1cdn10(2) TGFBhapl(1) TGFBhapl(2)
Urboro  XQ-1     C             T             CG          TG
Urboro  XQ-2     C             C             CG          CG
Urboro  XQ-5     C             T             CG          TG
Urboro  XQ-21    C             T             CG          TG
Urboro  XQ-7     C             T             CG          TG
Urboro  XQ-20    C             T             CG          TG
Urboro  XQ-6     T             T             TG          TG
Urboro  XQ-8     C             T             CG          TG
Urboro  XQ-9     T             T             TG          TG
Urboro  XQ-10    C             T             CG          TG
</pre></div>
</div>
</div>
<p>This example includes nucleotide sequence data: the TGFB1CDN10 locus
consists of one nucleotide, the TGFBhapl locus is actually haplotype
data, but PyPop simply treats each combination as a separate “allele”
for subsequent analysis.</p>
<div class="literal-block-wrapper docutils container" id="data-allelecount">
<div class="code-block-caption"><span class="caption-number">Listing 2.6 </span><span class="caption-text">Allele count data</span><a class="headerlink" href="#data-allelecount" title="Permalink to this code">¶</a></div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>populat    method  ethnic     country    latit   longit
UchiTelle  PCR-SSO Klingon    QZ         052.81N 100.25E
dqa1  count
0101  31
0102  37
0103  17
0201  21
0301  32
0401  9
0501  35
</pre></div>
</div>
</div>
<p>PyPop can also process allele count data. However, you cannot mix allele
count data and genotype data together in the one file.</p>
<div class="admonition note" id="data-allelecount-note">
<p class="admonition-title">Note</p>
<p>Currently each <code class="docutils literal notranslate"><span class="pre">.pop</span></code> file can only contain allele count data for
<em>one locus</em>. In order to process multiple loci for one population you
must create a separate <code class="docutils literal notranslate"><span class="pre">.pop</span></code> for each locus.</p>
</div>
<p>These population files are plain text files, such as you might save
out of the <strong class="program">Notepad</strong> application on Windows (or
<strong class="program">Emacs</strong>). The columns are all tab-delimited, so you can
include spaces in your labels. If you have your data in a spreadsheet
application, such as <strong class="program">Excel</strong> or <strong class="program">LibreOffice</strong>, export the file as
tab-delimited text, in order to use it as PyPop data file.</p>
</section>
<section id="missing-data">
<h3><span class="section-number">2.4.2. </span>Missing data<a class="headerlink" href="#missing-data" title="Permalink to this heading">¶</a></h3>
<p>Untyped or missing data may be represented in a variety of ways. The
default value for untyped or missing data is a series of four asterisks
(<code class="docutils literal notranslate"><span class="pre">****</span></code>) as specified by the <code class="file docutils literal notranslate"><span class="pre">config.ini</span></code>. You may not “represent”
untyped data by leaving a column blank, nor may you represent a
homozygote by leaving the second column blank. All cells for which you
have data must include data, and all cells for which you do not have
data must also be filled in, using a missing data value.</p>
<p>For individuals who were not typed at all loci, the data in loci for
which they are typed will be used on all single-locus analyses for that
individual and locus, so that you see the value of the number of
individuals (<code class="docutils literal notranslate"><span class="pre">n</span></code>) vary from locus to locus in the output. These
individuals’ data will also be used for multi-locus analyses. Only the
loci that contain no missing data will be included in any multi-locus
analysis.</p>
<p>If an individual is only partially typed at a locus, it will be treated
as if it were completely untyped, and data for that individual for that
locus will be dropped from ALL analyses.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<ul class="simple">
<li><p>Do not leave trailing blank lines at the end of your data file, as
this currently causes PyPop to terminate with an error message
that takes experience to diagnose.</p></li>
<li><p>For haplotype estimation and linkage disequilibrium calculations
(i.e., the emhaplofreq part of the program) you are currently
restricted to a maximum of seven loci per haplotype request. For
haplotype estimation there is a limit of 5000 for the number of
individuals (<code class="docutils literal notranslate"><span class="pre">n</span></code>) <a class="footnote-reference brackets" href="#id7" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a></p></li>
</ul>
</div>
</section>
</section>
<section id="the-configuration-file">
<span id="guide-usage-configfile"></span><h2><span class="section-number">2.5. </span>The configuration file<a class="headerlink" href="#the-configuration-file" title="Permalink to this heading">¶</a></h2>
<p>The sets of population genetic analyses that are run on your population
data file and the manner in which the data file is interpreted by PyPop
is controlled by a configuration file, the default name for which is
<code class="file docutils literal notranslate"><span class="pre">config.ini</span></code>. This is another plain text file consisting of comments
(which are lines that start with a semi-colon), sections (which are
lines with labels in square brackets), and options (which are lines
specifying settings relevant to that section in the <code class="docutils literal notranslate"><span class="pre">option=value</span></code>
format).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If any option runs over one line (such as <code class="docutils literal notranslate"><span class="pre">validSampleFields</span></code>) then
the second and subsequent lines must be indented by exactly <strong>one
space</strong>.</p>
</div>
<section id="a-minimal-configuration-file">
<span id="config-minimal"></span><h3><span class="section-number">2.5.1. </span>A minimal configuration file<a class="headerlink" href="#a-minimal-configuration-file" title="Permalink to this heading">¶</a></h3>
<p>Here we present a minimal <code class="docutils literal notranslate"><span class="pre">.ini</span></code> file corresponding to
<a class="reference internal" href="#data-minimal-noheader-noids"><span class="std std-numref">Listing 2.1</span></a> A section by section
review of this file follows. (Note comment lines have been omitted in
the above example for clarity). A description of more advanced options
is contained in <a class="reference internal" href="#config-advanced"><span class="std std-ref">Advanced options</span></a>.</p>
<div class="literal-block-wrapper docutils container" id="config-minimal-example">
<div class="code-block-caption"><span class="caption-number">Listing 2.7 </span><span class="caption-text">Minimal config.ini file</span><a class="headerlink" href="#config-minimal-example" title="Permalink to this code">¶</a></div>
<div class="highlight-ini notranslate"><div class="highlight"><pre><span></span><span class="hll"><span class="k">[General]</span>
</span><span class="na">debug</span><span class="o">=</span><span class="s">0</span>

<span class="hll"><span class="k">[ParseGenotypeFile]</span>
</span><span class="na">untypedAllele</span><span class="o">=</span><span class="s">****</span>
<span class="na">alleleDesignator</span><span class="o">=</span><span class="s">*</span>
<span class="na">validSampleFields</span><span class="o">=</span><span class="s">*a_1</span>
<span class="w"> </span><span class="na">*a_2</span>
<span class="w"> </span><span class="na">*c_1</span>
<span class="w"> </span><span class="na">*c_2</span>
<span class="w"> </span><span class="na">*b_1</span>
<span class="w"> </span><span class="na">*b_2</span>

<span class="hll"><span class="k">[HardyWeinberg]</span>
</span><span class="na">lumpBelow</span><span class="o">=</span><span class="s">5</span>

<span class="hll"><span class="k">[HardyWeinbergGuoThompson]</span>
</span><span class="na">dememorizationSteps</span><span class="o">=</span><span class="s">2000</span>
<span class="na">samplingNum</span><span class="o">=</span><span class="s">1000</span>
<span class="na">samplingSize</span><span class="o">=</span><span class="s">1000</span>

<span class="hll"><span class="k">[HomozygosityEWSlatkinExact]</span>
</span><span class="na">numReplicates</span><span class="o">=</span><span class="s">10000</span>

<span class="hll"><span class="k">[Emhaplofreq]</span>
</span><span class="na">allPairwiseLD</span><span class="o">=</span><span class="s">1</span>
<span class="na">allPairwiseLDWithPermu</span><span class="o">=</span><span class="s">0</span>
<span class="c1">;;numPermuInitCond=5</span>
</pre></div>
</div>
</div>
<p><strong>Configuration file sections</strong> (highlighted above)</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">[General]</span></code></p>
<p>This section contains variables that control the overall behavior of
PyPop.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">debug=0</span></code>.</p>
<p>This setting is for debugging. Setting it to 1 will set off a
large amount of output of no interest to the general user. It
should not be used unless you are running into trouble and need to
communicate with the PyPop developers about the problems.</p>
</li>
</ul>
</li>
<li><p>Specifying data formats</p>
<p>There are two possible formats: <code class="docutils literal notranslate"><span class="pre">[ParseGenotypeFile]</span></code> and
<code class="docutils literal notranslate"><span class="pre">[ParseAlleleCountFile]</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">[ParseGenotypeFile]</span></code>.</p>
<p>If your data is genotype data, you will want a section labeled:
<code class="docutils literal notranslate"><span class="pre">[ParseGenotypeFile]</span></code>.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">alleleDesignator</span></code>.</p>
<p>This option is used to tell PyPop what is allele data and what
isn’t. You must use this symbol in :ref:<code class="docutils literal notranslate"><span class="pre">`validSampleFields</span></code>
option. The default is <code class="docutils literal notranslate"><span class="pre">*</span></code>. In general, you won’t need to
change it. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">*</span></code> <strong>]</strong></p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">untypedAllele</span></code>.</p>
<p>This option is used to tell PyPop what symbol you have used in
your data files to represent untyped or unknown data
fields. These fields MAY NOT BE LEFT BLANK. You must use
something consistent that cannot be confused with real data
here. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">****</span></code> <strong>]</strong></p>
</li>
</ul>
</li>
</ul>
<blockquote id="validsamplefields">
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">validSampleFields</span></code>.</p>
<p>This option should contain the names of the loci immediately
preceding your genotype data (if it has three header lines, this
information will be on the third line, otherwise it will be the
first line of the file).<strong>[There is no default, this option must
always be present]</strong></p>
<p>The format is as follows, for each sample field (which may either
be an identifying field for the sample such as <code class="docutils literal notranslate"><span class="pre">populat</span></code>, or
contain allele data) create a new line where:</p>
<ul class="simple">
<li><p>The first line (<code class="docutils literal notranslate"><span class="pre">validSampleFields=</span></code>) consists of the name of
your sample field (if it contains allele data, the name of the
field should be preceded by the character designated in the
<code class="docutils literal notranslate"><span class="pre">alleleDesignator</span></code> option above).</p></li>
<li><p>All subsequent lines after the first <em>must</em> be preceded by <em>one
space</em> (again if it contains allele data, the name of the field
should be preceded by the character designated in the
<code class="docutils literal notranslate"><span class="pre">alleleDesignator</span></code> option above).</p></li>
</ul>
<p>Here is an example:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>validSampleFields=*a_1
 *a_2
 *c_1
 *c_2
 *b_1
 *b_2    Note initial space at start of line.
</pre></div>
</div>
<p>Here is example that includes identifying (non-allele data)
information such as sample id (<code class="docutils literal notranslate"><span class="pre">id</span></code>) and population name
(<code class="docutils literal notranslate"><span class="pre">populat</span></code>):</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>validSampleFields=populat
 id
 *a_1
 *a_2
 *c_1
 *c_2
 *b_1
 *b_2
</pre></div>
</div>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[ParseAlleleCountFile]</span></code>.</p>
<p>If your data is not genotype data, but rather, data of the
allele-name count format, then you will want to use the
<code class="docutils literal notranslate"><span class="pre">[ParseAlleleCountFile]</span></code> section INSTEAD of the
<code class="docutils literal notranslate"><span class="pre">[ParseGenotypeFile]</span></code> section. The <code class="docutils literal notranslate"><span class="pre">alleleDesignator</span></code> and
<code class="docutils literal notranslate"><span class="pre">untypedAllele</span></code> options work identically to that described for
<code class="docutils literal notranslate"><span class="pre">[ParseGenotypeFile]</span></code>.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">validSampleFields</span></code>.</p>
<p>This option should contain either a single locus name or a
colon-separated list of all loci that will be in the data files
you intend to analyze using a specific <code class="docutils literal notranslate"><span class="pre">.ini</span></code> file. The
colon-separated list allows you to avoid changing the <code class="docutils literal notranslate"><span class="pre">.ini</span></code>
file when running over a collection of data files containing
different loci. e.g.,</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>validSampleFields=A:B:C:DQA1:DQB1:DRB1:DPB1:DPA1
 count
</pre></div>
</div>
<p>Note that each <code class="docutils literal notranslate"><span class="pre">.pop</span></code> file must contain only one locus (see
<a class="reference external" href="data-allelecount-note">note_title</a> in
<a class="reference internal" href="#data-allelecount"><span class="std std-numref">Listing 2.6</span></a>). Listing multiple loci
simply permits the same <code class="docutils literal notranslate"><span class="pre">.ini</span></code> file to be reused for each data
file.</p>
</li>
</ul>
</div></blockquote>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">[HardyWeinberg]</span></code></p>
<p>Hardy-Weinberg analysis is enabled by the presence of this section.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">lumpBelow</span></code>.</p>
<p>This option value represents a cut-off value. Alleles with an
expected value equal to or less than <code class="docutils literal notranslate"><span class="pre">lumpBelow</span></code> will be lumped
together into a single category for the purpose of calculating the
degrees of freedom and overall <code class="docutils literal notranslate"><span class="pre">p</span></code>-value for the chi-squared
Hardy-Weinberg test.</p>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">[HardyWeinbergGuoThompson]</span></code></p>
<p>When this section is present, an implementation of the
Hardy-Weinberg exact test is run using the original
<a class="reference internal" href="biblio.html#guo-thompson-1992" id="id2"><span>[Guo:Thompson:1992]</span></a> code, using a Monte-Carlo Markov chain (MCMC). In
addition, two measures (Chen and Diff) of the goodness of it of
individual genotypes are reported under this option <a class="reference internal" href="biblio.html#chen-etal-1999" id="id3"><span>[Chen:etal:1999]</span></a>
By default this section is not enabled. This is a different
implementation to the <strong class="program">Arlequin</strong> version listed in
<a class="reference internal" href="#config-advanced"><span class="std std-ref">Advanced options</span></a>, below.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">dememorizationSteps</span></code>.</p>
<p>Number of steps of to “burn-in” the Markov chain before statistics
are collected.<strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">2000</span></code> <strong>]</strong></p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">samplingNum</span></code>.</p>
<p>Number of Markov chain samples <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">1000</span></code> <strong>]</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">samplingSize</span></code>.</p>
<p>Markov chain sample size<strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">1000</span></code> <strong>]</strong>.</p>
</li>
</ul>
<p>Note that the <strong>total</strong> number of steps in the Monte-Carlo Markov
chain is the product of <code class="docutils literal notranslate"><span class="pre">samplingNum</span></code> and <code class="docutils literal notranslate"><span class="pre">samplingSize</span></code>, so the
default values described above would contain 1,000,000 (= 1000 x
1000) steps in the MCMC chain.</p>
<p>The default values for options described above have proved to be
optimal for us and if the options are not provided these defaults
will be used. If you change the values and have problems, please let
us <strong>know</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">[HomozygosityEWSlatkinExact]</span></code></p>
<p>The presence of this section enables Slatkin’s <a class="reference internal" href="biblio.html#slatkin-1994" id="id4"><span>[Slatkin:1994]</span></a>
implementation of the Ewens-Watterson exact test of neutrality.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">numReplicates</span></code>.</p>
<p>The default values have proved to be optimal for us. There is no
reason to change them unless you are particularly curious. If you
change the default values and have problems, please let us know.</p>
</li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">[Emhaplofreq]</span></code></p>
<p>The presence of this section enables haplotype estimation and
calculation of linkage disequilibrium (LD) measures.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">lociToEstHaplo</span></code>.</p>
<p>In this option you can list the multi-locus haplotypes for which
you wish the program to estimate and to calculate the LD. It
should be a comma-separated list of colon-joined loci. e.g.,</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>lociToEstHaplo=a:b:drb1,a:b:c,drb1:dqa1:dpb1,drb1:dqb1:dpb1
</pre></div>
</div>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">allPairwiseLD</span></code>.</p>
<p>Set this to <code class="docutils literal notranslate"><span class="pre">1</span></code> (one) if you want the program to calculate all
pairwise LD for your data, otherwise set this to <code class="docutils literal notranslate"><span class="pre">0</span></code> (zero).</p>
</li>
</ul>
</li>
</ul>
<blockquote id="config-allpairwiseldwithpermu">
<div><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">allPairwiseLDWithPermu</span></code>.</p>
<p>Set this to a positive integer greater than 1 if you need to
determine the significance of the pairwise LD measures in the
previous section. The number you use is the number of permutations
that will be run to ascertain the significance (this should be at
least 1000 or greater). (Note this is done via permutation testing
performed after the pairwise LD test for all pairs of loci. Note
also that this test can take <em>DAYS</em> if your data is highly
polymorphic.)</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">numPermuInitCond</span></code>.</p>
<p>Set this to change the number of initial conditions used per
permutation. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">5</span></code> <strong>]</strong>. (<em>Note: this parameter is only used
if</em> <code class="docutils literal notranslate"><span class="pre">allPairwiseLDWithPermu</span></code> <em>is set and nonzero</em>).</p>
</li>
</ul>
</div></blockquote>
</section>
<section id="advanced-options">
<span id="config-advanced"></span><h3><span class="section-number">2.5.2. </span>Advanced options<a class="headerlink" href="#advanced-options" title="Permalink to this heading">¶</a></h3>
<p>The following section describes additional options to previously
described sections. Most of the time these options can be omitted and
PyPop will choose defaults, however these advanced options do offer
greater control over the application. In particular, customization will
be required for data that has sample identifiers as in
<a class="reference internal" href="#data-minimal-noheader"><span class="std std-numref">Listing 2.2</span></a> or header data block as in
<a class="reference internal" href="#data-hla"><span class="std std-numref">Listing 2.3</span></a> and both <code class="docutils literal notranslate"><span class="pre">validSampleFields</span></code> (described
above) and <code class="docutils literal notranslate"><span class="pre">validPopFields</span></code> (described below) will need to be
modified.</p>
<p>It also describes two extra sections related to using PyPop in
conjunction with <strong class="program">Arlequin</strong>: <code class="docutils literal notranslate"><span class="pre">[Arlequin]</span></code> and
<code class="docutils literal notranslate"><span class="pre">[HardyWeinbergGuoThompsonArlequin]</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">[General]</span></code> <strong>advanced options</strong></p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">txtOutFilename</span></code> and <code class="docutils literal notranslate"><span class="pre">xmlOutFilename</span></code>.</p>
<p>If you wish to specify a particular name for the output file, which
you want to remain identical over several runs, you can set these
two items to particular values. The default is to have the program
select the output filename, which can be controlled by the next
variable. <strong>[Default: not used]</strong></p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">outFilePrefixType</span></code>.</p>
<p>This option can either be omitted entirely (in which case the
default will be <code class="docutils literal notranslate"><span class="pre">filename</span></code>) or be set in several ways. The
default is set as <code class="docutils literal notranslate"><span class="pre">filename</span></code>, which will result in three output
files named <code class="file docutils literal notranslate"><span class="pre">original-filename-minus-suffix-out.xml</span></code>,
<code class="file docutils literal notranslate"><span class="pre">original-filename-minus-suffix-out.txt</span></code>, and
<code class="file docutils literal notranslate"><span class="pre">original-filename-minus-suffix-filter.xml</span></code>. <strong>[Default:</strong>
<code class="docutils literal notranslate"><span class="pre">filename</span></code> <strong>]</strong></p>
<p>If you set the value to <code class="docutils literal notranslate"><span class="pre">date</span></code> instead of filename, you’ll get the
date incorporated in the filename as follows:
<code class="file docutils literal notranslate"><span class="pre">original-filename-minus-suffix-YYYY-nn-dd-HH-MM-SS-out.</span><em><span class="pre">xml,txt</span></em></code>.
e.g., <code class="file docutils literal notranslate"><span class="pre">USAFEL-UchiTelle-2003-09-21-01-29-35-out.xml</span></code> (where Y, n,
d, H, M, S refer to year, month, day, hour, minute and second,
respectively).</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">xslFilename</span></code>.</p>
<p>This option specifies where to find the XSLT file to use for
transforming PyPop’s xml output into human-readable form. Most users
will not normally need to set this option, and the default is the
system-installed <code class="file docutils literal notranslate"><span class="pre">text.xsl</span></code> file.</p>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[ParseGenotypeFile]</span></code> <strong>advanced options</strong></p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">fieldPairDesignator</span></code>.</p>
<p>This option allows you to override the coding for the headers for
each pair of alleles at each locus; it must match the entry in the
config file under <code class="docutils literal notranslate"><span class="pre">validSampleFields</span></code> and the entries in your
population data file. If you want to use something other than <code class="docutils literal notranslate"><span class="pre">_1</span></code>
and <code class="docutils literal notranslate"><span class="pre">_2</span></code>, change this option, for instance, to use letters and
parentheses, change it as follows: <code class="docutils literal notranslate"><span class="pre">fieldPairDesignator=(a):(b)</span></code>
<strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">_1:_2</span></code> <strong>]</strong></p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">popNameDesignator</span></code>.</p>
<p>There is a special designator to mark the population name field,
which is usually the first field in the data block. <strong>[Default:</strong>
<code class="docutils literal notranslate"><span class="pre">+</span></code> <strong>]</strong></p>
<p>If you are analyzing data that contains a population name for each
sample, then the first entry in your <code class="docutils literal notranslate"><span class="pre">validSampleFields</span></code> section
should have a prefixed +, as below:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>validSampleFields=+populat
 *a_1
 *a_2
 ...
</pre></div>
</div>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">validPopFields</span></code>.</p>
<p>If you are analyzing data with an initial two line population header
block information as in <a class="reference internal" href="#data-hla"><span class="std std-ref">Multi-locus allele-level HLA genotype data with sample and header information</span></a>, then you will
need to set this option. In this case, it should contain the field
names in the first line of the header information of your file.
<strong>[Default: required when a population data-block is present in data
file]</strong>, e.g.:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>validPopFields=labcode
 method
 ethnic
 country
 latit
 longit
</pre></div>
</div>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[Emhaplofreq]</span></code> <strong>advanced options</strong></p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">permutationPrintFlag</span></code>.</p>
<p>Determines whether the likelihood ratio for each permutation will be
logged to the XML output file, this is disabled by default.
<strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">0</span></code> <strong>(i.e. OFF)]</strong>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If this is enabled it can <em>drastically</em> increase the size of the
output XML file on the order of the product of the number of
possible pairwise comparisons and permutations. Machines with
lower RAM and disk space may have difficulty coping with this.</p>
</div>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[Arlequin]</span></code> <strong>extra section</strong></p>
<p>This section sets characteristics of the <strong class="program">Arlequin</strong>
application if it has been installed (it must be installed separately
from PyPop as we cannot distribute it). The options in this section
are only used when a test requiring <strong class="program">Arlequin</strong>, such as it’s
implementation of Guo and Thompson’s <a class="reference internal" href="biblio.html#guo-thompson-1992" id="id5"><span>[Guo:Thompson:1992]</span></a> Hardy-Weinberg
exact test is invoked (see below).</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">arlequinExec</span></code>.</p>
<p>This option specifies where to find the <strong class="program">Arlequin</strong>
executable on your system. The default assumes it is on your system
path. <strong>[Default:</strong> <code class="file docutils literal notranslate"><span class="pre">arlecore.exe</span></code> <strong>]</strong></p>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[HardyWeinbergGuoThompsonArlequin]</span></code> <strong>extra section</strong></p>
<p>When this section is present, <strong class="program">Arlequin</strong>’s implementation of the
Hardy-Weinberg exact test is run, using a Monte-Carlo Markov Chain
implementation. By default this section is not enabled.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">markovChainStepsHW</span></code>.</p>
<p>Length of steps in the Markov chain <strong>[Default: 2500000]</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">markovChainDememorisationStepsHW</span></code>.</p>
<p>Number of steps of to “burn-in” the Markov chain before statistics
are collected.<strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">5000</span></code> <strong>]</strong></p>
</li>
</ul>
<p>The default values for options described above have proved to be optimal
for us and if the options are not provided these defaults will be used.
If you change the values and have problems, please let us <strong>know</strong>.</p>
<p><code class="docutils literal notranslate"><span class="pre">[Filters]</span></code> <strong>extra section</strong></p>
<p>When this section is present, it allows you to specify succesive filters
to the data.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">filtersToApply</span></code>.</p>
<p>Here you specify which filters you want applied to the data and the
order in which you want them applied. Separate each filter name with
a colon (<code class="docutils literal notranslate"><span class="pre">:</span></code>). Currently there are four predefined filter:
<code class="docutils literal notranslate"><span class="pre">AnthonyNolan</span></code>, <code class="docutils literal notranslate"><span class="pre">Sequence</span></code>, <code class="docutils literal notranslate"><span class="pre">DigitBinning</span></code>, and
<code class="docutils literal notranslate"><span class="pre">CustomBinning</span></code>. If you specify one or more of these filters, you
will get the default behavior of the filter. If you wish to modify
the default behavior, you should add a section with the same name as
the specified filter(s). See next section for more on this. Please
note that, while you are allowed to specify any ordering for the
filters, some orderings may not make sense. For example, the ordering
Sequence:AnthonyNolan would not make sense (because as far as PyPop
is concerned, your alleles are now amino acid residues.) However, the
reverse ordering, AnthonyNolan:Sequence, would be logical and perhaps
even advisable.</p>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[AnthonyNolan]</span></code> <strong>filter section</strong></p>
<p>This section is <em>only</em> useful for HLA data. Like all filter sections, it
will only be used if present in the <code class="docutils literal notranslate"><span class="pre">filtersToApply</span></code> line specified
above. If so enabled, your data will be filtered through the Anthony
Nolan database of known HLA allele names before processing. The data
files this filter relies on are <em>not</em> currently distributed with PyPop
but can be obtained via the <a class="reference external" href="ftp://ftp.ebi.ac.uk/pub/databases/imgt/mhc/hla/">IMGT ftp
site</a>. Invocation of
this filter will produce a <code class="docutils literal notranslate"><span class="pre">popfile-filter.xml</span></code> file output showing
what was resolved and what could not be resolved.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">alleleFileFormat</span></code>.</p>
<p>This options specifies which of the formats the Anthony Nolan
allele data will be used. The option can be set to either <code class="docutils literal notranslate"><span class="pre">txt</span></code>
(for the plain free text format) or <code class="docutils literal notranslate"><span class="pre">msf</span></code> (for the <a class="reference external" href="http://www.ebi.ac.uk/imgt/hla/download.html">Multiple
Sequence Format</a>)
<strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">msf</span></code> <strong>]</strong></p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">directory</span></code>.</p>
<p>Specifies the path to the root of the sequence files. For <code class="docutils literal notranslate"><span class="pre">txt</span></code>:
<strong>[Default:</strong>
<code class="file docutils literal notranslate"><em><span class="pre">prefix</span></em><span class="pre">/share/PyPop/anthonynolan/HIG-seq-pep-text/</span></code>
<strong>]</strong>.  For <code class="docutils literal notranslate"><span class="pre">msf</span></code> files <strong>[Default:</strong>
<code class="file docutils literal notranslate"><em><span class="pre">prefix</span></em><span class="pre">/share/PyPop/anthonynolan/msf/</span></code> <strong>]</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">preserve-ambiguous</span></code>.</p>
<p>The default behavior of the <code class="docutils literal notranslate"><span class="pre">AnthonyNolan</span></code> filter is to ignore
allele ambiguity (“slash”) notation. This notation, common in the
literature, looks like: <code class="docutils literal notranslate"><span class="pre">010101/0102/010301</span></code>. The default behavior
will simply truncate this to <code class="docutils literal notranslate"><span class="pre">0101</span></code>. If you want to preserve the
notation, set the option to <code class="docutils literal notranslate"><span class="pre">1</span></code>. This will result in a filtered
allele “name” of <code class="docutils literal notranslate"><span class="pre">0101/0102/0103</span></code> in the above hypothetical
example. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">0</span></code> <strong>]</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">preserve-unknown</span></code>.</p>
<p>The default behavior of the <code class="docutils literal notranslate"><span class="pre">AnthonyNolan</span></code> filter is to replace
unknown alleles with the <code class="docutils literal notranslate"><span class="pre">untypedAllele</span></code> designator. If you want
the filter to keep allele names it does not recognize, set the option
to <code class="docutils literal notranslate"><span class="pre">1</span></code>. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">0</span></code> <strong>]</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">preserve-lowres</span></code>.</p>
<p>This option is similar to <code class="docutils literal notranslate"><span class="pre">preserve-unknown</span></code>, but only applies to
lowres alleles. If set to <code class="docutils literal notranslate"><span class="pre">1</span></code>, PyPop will keep allele names that are
shorter than the default allele name length, usually 4 digits long.
But if the preserve-unknown flag is set, this one has no effect,
because all unknown alleles are preserved. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">0</span></code> <strong>]</strong>.</p>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[Sequence]</span></code> <strong>filter section</strong></p>
<p>This section allows configuration of the sequence filter. Like all
filter sections, it will only will be used if present in the
<code class="docutils literal notranslate"><span class="pre">filtersToApply</span></code> line specified above. If so enabled, your allele
names will be translated into sequences, and all ensuing analyses will
consider each position in the sequence to be a distinct locus. This
filter makes use of the same msf format alignment files as used above in
the AnthonyNolan filter. It does not work with the txt format alignment
files.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">sequenceFileSuffix</span></code>.</p>
<p>Determines the files that will be examined in order to read in a
sequence for each allele. (ie, if the file for locus A is
<code class="file docutils literal notranslate"><span class="pre">A_prot.msf</span></code>, the value would be <code class="docutils literal notranslate"><span class="pre">_prot</span></code> whereas if you
wanted to use the nucleotide sequence files, you might use
<code class="docutils literal notranslate"><span class="pre">_nuc</span></code>.) <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">_prot</span></code> <strong>]</strong>.</p>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">directory</span></code>.</p>
<p>Specifies the path to the root of the sequence files, in the same
manner as in the AnthonyNolan section, above.</p>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[DigitBinning]</span></code> <strong>filter section</strong></p>
<p>This section allows configuration of the DigitBinning filter. Like all
filter sections, it will be used if present in the <code class="docutils literal notranslate"><span class="pre">filtersToApply</span></code>
line specified above. If so enabled, your allele names will be truncated
after the nth digit.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">binningDigits</span></code>.</p>
<p>An integer that specifies how many digits to keep after the
truncation. <strong>[Default:</strong> <code class="docutils literal notranslate"><span class="pre">4</span></code> <strong>]</strong>.</p>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">[CustomBinning]</span></code> <strong>filter section</strong></p>
<p>This section allows configuration of the CustomBinning filter. Like all
filter sections, it will only be used if present in the
<code class="docutils literal notranslate"><span class="pre">filtersToApply</span></code> line specified above.</p>
<p>You can provide a set of custom rules for replacing allele names. Allele
names should be separated by <code class="docutils literal notranslate"><span class="pre">/</span></code> marks. This filter matches any allele
names that are exactly the same as the ones you list here, and will also
find “close matches” (but only if there are no exact matches.). Here is
an example:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>A=01/02/03
 04/05/0306
 !06/1201/1301
 !07/0805
</pre></div>
</div>
<p>In the example above, <code class="docutils literal notranslate"><span class="pre">A*03</span></code> alleles will match to <code class="docutils literal notranslate"><span class="pre">01/02/03</span></code>,
except for <code class="docutils literal notranslate"><span class="pre">A*0306</span></code>, which will match to <code class="docutils literal notranslate"><span class="pre">04/05/0306</span></code>. If you place
a <code class="docutils literal notranslate"><span class="pre">!</span></code> mark in front of the first allele name, that first name will be
used as the “new name” for the binned group (for example, <code class="docutils literal notranslate"><span class="pre">A*0805</span></code>
will be called <code class="docutils literal notranslate"><span class="pre">07</span></code> in the custom-binned data.) Note that the space at
the beginning of the lines (following the first line of each locus) is
important. The above rules are just dummy examples, provided to
illustrate how the filter works. PyPop is distributed with a
biologically relevant set of <code class="docutils literal notranslate"><span class="pre">CustomBinning</span></code> rules that have been
compiled from several sources <a class="footnote-reference brackets" href="#id8" id="id6" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a></p>
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id7" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>These hardcoded numbers can be changed if you obtain the source code
yourself and change the appropriate #define <code class="docutils literal notranslate"><span class="pre">emhaplofreq.h</span></code> and
recompile the program.</p>
</aside>
<aside class="footnote brackets" id="id8" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id6">2</a><span class="fn-bracket">]</span></span>
<p><a class="reference internal" href="biblio.html#mack-etal-2007" id="id9"><span>[Mack:etal:2007]</span></a>; <a class="reference internal" href="biblio.html#cano-etal-2007" id="id10"><span>[Cano:etal:2007]</span></a>; The Anthony Nolan list of deleted
allele names
(` &lt;<a class="reference external" href="http://www.anthonynolan.com/HIG/lists/delnames.html">http://www.anthonynolan.com/HIG/lists/delnames.html</a>&gt;`__); and the
Ambiguous Allele Combinations, release 2.18.0
(` &lt;<a class="reference external" href="http://www.ebi.ac.uk/imgt/hla/ambig.html">http://www.ebi.ac.uk/imgt/hla/ambig.html</a>&gt;`__).</p>
</aside>
</aside>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
    
        <div id="show_right_sidebar">
            <p><a class="toggle_right_sidebar" href="#"><span class="icon">&lt;</span><span>Page contents<span></a></p>
        </div>

        <div id="right_sidebar">
            <p><a class="toggle_right_sidebar" href="#"><span class="icon">&gt;</span><span>Page contents:<span></a></p>
            <div class="page_toc">
                <ul>
<li><a class="reference internal" href="#">2. Getting started with PyPop</a><ul>
<li><a class="reference internal" href="#introduction">2.1. Introduction</a><ul>
<li><a class="reference internal" href="#interactive-mode">2.1.1. Interactive mode</a></li>
<li><a class="reference internal" href="#batch-mode">2.1.2. Batch mode</a></li>
<li><a class="reference internal" href="#what-happens-when-you-run-pypop">2.1.3. What happens when you run PyPop?</a></li>
</ul>
</li>
<li><a class="reference internal" href="#using-popmeta-to-aggregate-results">2.2. Using <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> to aggregate results</a></li>
<li><a class="reference internal" href="#command-line-interfaces">2.3. Command-line interfaces</a><ul>
<li><a class="reference internal" href="#pypop-usage">2.3.1. <code class="docutils literal notranslate"><span class="pre">pypop</span></code> usage</a><ul>
<li><a class="reference internal" href="#options-for-pypop">Options for pypop</a></li>
<li><a class="reference internal" href="#tsv-output-options">TSV output options</a></li>
<li><a class="reference internal" href="#mutually-exclusive-input-options">Mutually exclusive input options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#popmeta-usage">2.3.2. <code class="docutils literal notranslate"><span class="pre">popmeta</span></code> usage</a><ul>
<li><a class="reference internal" href="#positional-arguments">Positional Arguments</a></li>
<li><a class="reference internal" href="#options-for-popmeta">Options for popmeta</a></li>
<li><a class="reference internal" href="#mutually-exclusive-popmeta-options">Mutually exclusive popmeta options</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#the-data-file">2.4. The data file</a><ul>
<li><a class="reference internal" href="#sample-files">2.4.1. Sample files</a></li>
<li><a class="reference internal" href="#missing-data">2.4.2. Missing data</a></li>
</ul>
</li>
<li><a class="reference internal" href="#the-configuration-file">2.5. The configuration file</a><ul>
<li><a class="reference internal" href="#a-minimal-configuration-file">2.5.1. A minimal configuration file</a></li>
<li><a class="reference internal" href="#advanced-options">2.5.2. Advanced options</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            </div>
        </div>
    

      <div class="clearer"></div>
    </div>
    <div class="button_nav_wrapper">
        <div class="button_nav">
            <div class="left">
                
                <a href="guide-chapter-install.html">
                    <span class="icon">&lt;</span><span><span class="section-number">1. </span>Installing PyPop</span></a>
                
            </div>

            <div class="right">
                
                    <a href="guide-chapter-instructions.html"><span><span class="section-number">3. </span>Interpreting PyPop output</span><span class="icon">&gt;</span></a>
                
            </div>
        </div>
    </div>


    <div class="footer" role="contentinfo">
        &#169; Copyright 2023 PyPop contributors.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 6.2.1.
    </div>

<p id="theme_credit">Styled using the <a href="https://github.com/piccolo-orm/piccolo_theme">Piccolo Theme</a></p>
  </body>
</html>