<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
        "http://www.w3.org/TR/html4/loose.dtd">
<html>

<head>
  <title>PyPop: Python for Population Genetics</title>
  <meta name="generator" content="/usr/bin/emacs">
  <meta http-equiv="Content-type" content="text/html; charset=iso-8859-1">
</head>

 <body text="#000000" bgcolor="#ffffff">
  
  <table cellpadding="3" cellspacing="0" style="float:right; margin:20px; border:2px solid;">
   <tr><td><a href="#overview">Overview</a></td></tr>
   <tr><td><a href="#release">Download</a></td></tr>
   <tr><td>
      + <a href="#binaries">Binaries</a><br>
      + <a href="#source">Source code</a><br>
      + <a href="#docs">Documentation</a><br>
    </td></tr>
   <tr><td><a href="#citation">Citing PyPop</a></td></tr>
   <tr><td><a href="#ackno">Acknowlegments</a></td></tr>
  </table>

  <h1 align="center">PyPop: <em>Py</em>thon for <em>Pop</em>ulation Genetics</h1>

  <h4 align="center"><a href="#release">PyPop public beta available for download</a></h4>



  <h3><a name="overview">Overview</a></h3>

  <p><strong>PyPop</strong>
   (<em>Py</em>thon for <em>Pop</em>ulation
   Genetics) is an environment developed by the Thomson lab
    for doing large-scale population genetic
   analyses including: (1) conformity to Hardy-Weinberg expectations, (2) tests
    for balancing or directional selection; (3) estimates of haplotype
    frequencies (and their distributions) and measures and tests of
    significance for linkage disequilibrium (LD). </p>

  <p>It is an
   object-oriented framework implemented in the programming language
   Python.  <a href="http://www.python.org/">Python</a> is a flexible scripting
   language which allows rapid prototyping of code and has powerful
   features for interfacing with other languages, such as C (in which
   we have already implemented many routines and which is particularly
   suited to computationally intensive tasks).</p>

  <p>The output of the
   analyses are stored in the XML format (XML is the eXtensible Markup
   Language devised by the World Wide Web Consortium, and is a
   platform-independent, vendor-neutral, non-proprietary, open
   standard for storing data).  These
   output files can then be transformed using standard tools into many
   other data formats suitable for machine input (such as
   PHYLIP or input for  spreadsheet
   programs such as Excel or statistical
   packages, such as R,  plain text, or HTML
   for human-readable format.  Storing the output in XML allows the
   final viewable output format to be redesigned at will, without
   requiring the (often time-consuming) re-running of the analyses
   themselves.</p>

  <p>An outline of PyPop can be found in our<a href="#citation"> 2003 PSB paper</a>.</p>

   <h3><a name="release"><tt>PyPop</tt> (beta) is available for download</a></h3>

  <ul>
   <li>2004-03-09: 0.5.2 released (bug fix release, fixes Windows 98 .bat file problems) <strong>&larr; Most recent</strong> </li>
   <li>2004-02-26: 0.5.1 released
   (mainly a bug fix and maintainance release)</li>
   <li>2003-12-31: 0.5 released (first public beta) </li>
  </ul>
  
  (Please note that links and documentation below always refer to the
  most recent release).

   <h4><a name="binaries">Binaries</a></h4>
   
   <p>Beta binary versions of PyPop are now publicly available 
    (see "Getting and installing PyPop" link below for links to the
    current binaries) for both:</p>
   
    <ul>
     <li><strong>GNU/Linux</strong> (the Linux binary needs a recent
      distribution that uses
      <em>at least</em> glibc 2.3.2: Red Hat 9, Fedora Core 1 and Slackware 9.1
      are known to work, earlier versions of Red Hat such as the 7.x series
      and 8.0 are known to <em>not</em> work)</li>
     <li><strong>Windows</strong> (the binary has been tested on Windows 98, 
      2000 and XP)</li>
    </ul>

   <p>Both binary packages are approximately 3.5 Mb downloads.</p>

   <h4><a name="source">Source</a></h4>

   <p>PyPop is
    <a href="http://www.gnu.org/philosophy/free-sw.html">free
     software</a> (sometimes referred to as <a
     href="http://www.opensource.org/">open source</a> software) and the
    source code is released under the terms of the "copyleft" GNU General
    Public License, or GPL
   (<a
  href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a>).  This means even if we haven't compiled a binary for your platform, 
    it is possible for you to download the source code and compile it 
    yourself.</p>

   <h4><a name="docs">Documentation</a></h4>

  <p>Documentation for PyPop is contained in the <em>PyPop User Guide</em>.</p>

  <ul><li><a href="docs/"><em>PyPop User Guide</em></a> includes:
    <ul>
     <li><a href="docs/guide-chapter-install.html">Getting and installing PyPop</a> (including links to the binaries and source code)</li>
     <li><a href="docs/guide-chapter-usage.html">Getting started with PyPop</a> (instructions on how to configure and run PyPop)</li>
     <li><a href="docs/guide-chapter-instructions.html">Interpreting PyPop output</a> (a description of the output statistics)</li>
    </ul>
   <li><em>PyPop User Guide</em> is also available as a <a href="pypop-guide.pdf">PDF</a> [220 kB]</li>
  </ul>

  <p>Please be aware that this is a <em>beta</em> release so it is highly
   likely that there may  be bugs and wrinkles to iron out.  Please direct
   all questions to Alex Lancaster at the address below.</p>

  <h3><a name="citation">How to cite PyPop</a></h3>

  <p>When citing PyPop, please cite the following paper from <a href="http://www-smi.stanford.edu/projects/helix/psb03/">PSB 2003</a>:  </p>

   <ul>
   <li>Alex Lancaster, Mark P. Nelson, Richard M. Single, Diogo Meyer, and
   Glenys Thomson (2003) "PyPop: a software framework for population
   genomics: analyzing large-scale multi-locus genotype data", in
   <em>Pacific Symposium on Biocomputing 8</em> (edited
   by R B Altman. et al.) 514-525, (World Scientific, Singapore, 2003) [<a href="psb-pypop.pdf">PDF</a> (344 kB)]. </li>
   </ul>

  <h3><a name="ackno">Acknowlegements</a></h3>

  <p>This  work  has  benefited from the support of NIH grant
   AI49213 (13th IHW).  Thanks  to  Steven J. Mack, Kristie A. Mather,
   Steve Marsh, Mark Grote and Leslie Louie for helpful comments and
   testing.</p>


<hr size="1" width="100%">

Last modified: $Date$<br>
<address>
E-mail maintainer: alexl at socrates . berkeley . edu</address>
    <p>
      <a href="http://validator.w3.org/check/referer"><img border="0"
          src="http://www.w3.org/Icons/valid-html401"
          alt="Valid HTML 4.01!" height="31" width="88"></a>
    </p>

</body>

</html>
