<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><title>Chapter 2. Getting started with PyPop</title><link rel="stylesheet" href="style.css" type="text/css"><meta name="generator" content="DocBook XSL Stylesheets V1.74.0"><link rel="home" href="pypop-guide.html" title="PyPop User Guide"><link rel="up" href="pypop-guide.html" title="PyPop User Guide"><link rel="prev" href="install-from-source.html" title="1.2. Installing from source"><link rel="next" href="guide-usage-datafile.html" title="2.2. The data file"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="navheader"><table width="100%" summary="Navigation header"><tr><th colspan="3" align="center">Chapter 2. Getting started with PyPop</th></tr><tr><td width="20%" align="left"><a accesskey="p" href="install-from-source.html">Prev</a> </td><th width="60%" align="center"> </th><td width="20%" align="right"> <a accesskey="n" href="guide-usage-datafile.html">Next</a></td></tr></table><hr></div><div class="chapter" lang="en"><div class="titlepage"><div><div><h2 class="title"><a name="guide-chapter-usage"></a>Chapter 2. Getting started with <code class="application">PyPop</code></h2></div><div><p class="pubdate">Last updated: $Date: 2008/11/19 05:14:04 $ by $Author:
  alancast $</p></div></div></div><div class="toc"><p><b>Table of Contents</b></p><dl><dt><span class="section"><a href="guide-chapter-usage.html#guide-usage-intro">2.1. Introduction</a></span></dt><dd><dl><dt><span class="section"><a href="guide-chapter-usage.html#id3144373">2.1.1. Interactive mode</a></span></dt><dt><span class="section"><a href="guide-chapter-usage.html#id3129007">2.1.2. Batch mode</a></span></dt><dt><span class="section"><a href="guide-chapter-usage.html#guide-usage-intro-run-details">2.1.3. What happens when you run
   PyPop?</a></span></dt></dl></dd><dt><span class="section"><a href="guide-usage-datafile.html">2.2. The data file</a></span></dt><dd><dl><dt><span class="section"><a href="guide-usage-datafile.html#id3155140">2.2.1. Sample files</a></span></dt><dt><span class="section"><a href="guide-usage-datafile.html#id3135827">2.2.2. Missing data</a></span></dt></dl></dd><dt><span class="section"><a href="guide-usage-configfile.html">2.3. The configuration file</a></span></dt><dd><dl><dt><span class="section"><a href="guide-usage-configfile.html#config-minimal">2.3.1. A minimal configuration file</a></span></dt><dt><span class="section"><a href="guide-usage-configfile.html#config-advanced">2.3.2. Advanced options</a></span></dt></dl></dd></dl></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="guide-usage-intro"></a>2.1. Introduction</h2></div></div></div><p>You may use <code class="application">PyPop</code> to analyze many
   different kinds of data, including allele-level genotype data (as
   in <a class="xref" href="guide-usage-datafile.html#data-minimal-noheader-noids" title="Example 2.1. Multi-locus allele-level genotype data">Example 2.1, “Multi-locus allele-level genotype data”</a>), allele-level
   frequency data (as in <a class="xref" href="guide-usage-datafile.html#data-allelecount" title="Example 2.6. Allele count data">Example 2.6, “Allele count data”</a>),
   microsatellite data, SNP data, and nucleotide and amino acid
   sequence data.</p><p>There are two ways to run <code class="application">PyPop</code>:
   </p><div class="itemizedlist"><ul type="disc"><li><p><em class="firstterm">interactive</em> mode (where the program
      will prompt you to directly type the input it needs); and</p></li><li><p><em class="firstterm">batch</em> mode (where you supply all the
      command line options the program needs).</p></li></ul></div><p> For the most straightforward application of
   <code class="application">PyPop</code>, where you wish to analyze a
   single population, the interactive mode is the simplest to use. We
   will describe this mode first then describe batch mode.</p><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="id3144373"></a>2.1.1. Interactive mode</h3></div></div></div><p>To run <code class="application">PyPop</code>, click the
    <code class="filename">pypop.bat</code> file (Windows) or type
    <strong class="userinput"><code>./pypop</code></strong> at the command prompt (GNU/Linux).
    You should see something like the following output (this is also
    described in detail in the instructions in the installation
    guide):</p><pre class="screen">PyPop: Python for Population Genomics (0.4.3)
Copyright (C) 2003 Regents of the University of California
This is free software.  There is NO warranty; not even for
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 
You may redistribute copies of PyPop under the terms of the
GNU General Public License.  For more information about these
matters, see the file named COPYING.
 
To accept the default in brackets for each filename, simply press
return for each prompt.
Please enter config filename [config.ini]: <strong class="userinput"><code>sample.ini</code></strong>
Please enter population filename [no default]: <strong class="userinput"><code>sample.pop</code></strong>
PyPop is processing sample.pop 

<span class="emphasis"><em>(Note: some messages with the prefix "LOG:" may appear here.
They are informational only and do not indicate improper operation 
of the program)</em></span>

PyPop run complete!
XML output can be found in: sample-out.xml
Plain text output can be found in: sample-out.txt
</pre><p>You should substitute the names of your own configuration
    (e.g., <code class="filename">config.ini</code>) and population file (e.g.,
    <code class="filename">Guatemalan.pop</code>) for
    <code class="filename">sample.ini</code> and
    <code class="filename">sample.pop</code>.  The formats for these files are
    described in <a class="xref" href="guide-usage-datafile.html" title="2.2. The data file">Section 2.2, “The data file”</a> and <a class="xref" href="guide-usage-configfile.html" title="2.3. The configuration file">Section 2.3, “The configuration file”</a>, below.
   </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="id3129007"></a>2.1.2. Batch mode</h3></div></div></div><p>To run <code class="application">PyPop</code> in batch mode, you
    can start <code class="application">PyPop</code> from the command line
    (in Windows: open a DOS shell, GNU/Linux: open a terminal window),
    change to the directory where you unpacked
    <code class="application">PyPop</code> and type</p><pre class="screen"><span class="command"><strong>pypop-batch</strong></span> <code class="filename">Guatemalan.pop</code></pre><div class="note" style="margin-left: 0.5in; margin-right: 0.5in;"><table border="0" summary="Note"><tr><td rowspan="2" align="center" valign="top" width="25"><img alt="[Note]" src="images/note.png"></td><th align="left">Note</th></tr><tr><td align="left" valign="top"><p>If your system administrator has installed
     <code class="application">PyPop</code> the name of the script may be
     renamed to something different.</p></td></tr></table></div><p>Batch mode assumes two things: that you have a file called
    <code class="filename">config.ini</code> in your current folder and that
    you also have your population file also in the current folder. You
    can specify a particular configuration file for
    <code class="application">PyPop</code> to use, by supplying the
    <code class="option">-c</code> option as follows:</p><pre class="screen"><span class="command"><strong>pypop-batch</strong></span> <code class="option">-c</code> <code class="filename">newconfig.ini</code> <code class="filename">Guatemalan.pop</code></pre><p>You may also redirect the output to a different directory
    (which must already exist) by using the <code class="option">-o</code>
    option:</p><pre class="screen"><span class="command"><strong>pypop-batch</strong></span> <code class="option">-c</code> <code class="filename">newconfig.ini</code> <code class="option">-o altdir</code> <code class="filename">Guatemalan.pop</code></pre><p>For a full list of options supported by
    <code class="application">PyPop</code>,  type <strong class="userinput"><code>pypop-batch
     --help</code></strong>.   You should receive a screen resembling the
    following:</p><pre class="screen">Usage: pypop [OPTION] INPUTFILE
Process and run population genetics statistics on an INPUTFILE.
Expects to find a configuration file called 'config.ini' in the
current directory or in /usr/share/PyPop/config.ini.

  -l, --use-libxslt    filter XML via XSLT using libxslt (default)
  -s, --use-4suite     filter XML via XSLT using 4Suite
  -x, --xsl=FILE       use XSLT translation file FILE
  -h, --help           show this message
  -c, --config=FILE    select alternative config file
  -d, --debug          enable debugging output (overrides config file setting)
  -i, --interactive    run in interactive mode, prompting user for file names
  -g, --gui            run GUI (currently disabled)
  -o, --outputdir=DIR  put output in directory DIR
  -V, --version        print version of PyPop
   
    INPUTFILE   input text file</pre><div class="warning" style="margin-left: 0.5in; margin-right: 0.5in;"><table border="0" summary="Warning"><tr><td rowspan="2" align="center" valign="top" width="25"><img alt="[Warning]" src="images/warning.png"></td><th align="left">Warning</th></tr><tr><td align="left" valign="top"><p>Documentation for these options is underway, but not
    currently available.</p></td></tr></table></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h3 class="title"><a name="guide-usage-intro-run-details"></a>2.1.3. What happens when you run
   <code class="application">PyPop</code>?</h3></div></div></div><p>The most common types of analysis will involve the editing of
    your <code class="filename">config.ini</code> file to suit your data (see
    <a class="xref" href="guide-usage-configfile.html" title="2.3. The configuration file">Section 2.3, “The configuration file”</a>) followed by the
    selection of either the interactive or batch mode described
    above. If your input configuration file is
    <code class="filename"><em class="replaceable"><code>configfilename</code></em></code> and
    your population file name is
    <code class="filename"><em class="replaceable"><code>popfilename</code></em>.txt</code>
    the initial output will be generated quickly, but your the
    <code class="application">PyPop</code> execution will not be finished
    until the text output file named
    <em class="replaceable"><code>popfilename</code></em>-out.txt has been created. A
    successful run will produce two output files:
    <code class="filename"><em class="replaceable"><code>popfilename</code></em>-out.xml</code>,
    <code class="filename"><em class="replaceable"><code>popfilename</code></em>-out.txt</code>.
    A third output file will be created if you are using the Anthony
    Nolan HLA filter option for HLA data to check your input for
    valid/known HLA alleles:
    <code class="filename"><em class="replaceable"><code>popfilename</code></em>-filter.xml</code>).</p><p>The
    <code class="filename"><em class="replaceable"><code>popfilename</code></em>-out.xml</code>
    file is the primary output created by PyPop and the human-readable
    <code class="filename"><em class="replaceable"><code>popfilename</code></em>-out.txt</code>
    file is a summary of the complete XML output. It is generated from
    the XML output via XSLT (eXtensible Stylesheet Language for
    Transformations) using the default XSLT stylesheet
    <code class="filename">text.xsl</code>, which is located in the xslt
    directory. The XML output can be further transformed using
    customized XSLT stylesheets into other formats for input to
    statistical software (e.g., R/Splus, SAS) or other population
    genetic software (e.g., PHYLIP). The <span class="command"><strong>popmeta</strong></span>
    script (<code class="filename">popmeta.bat</code> on Windows,
    <code class="filename">popmeta</code> on GNU/Linux) calls on other XSLT
    stylesheets to aggregate results from a number of output XML files
    from individual populations into a set of tab-separated (TSV)
    files containing summary statistics. These TSV files can be
    directly imported into a spreadsheet or statistical software. This
    script will be further documented in the next release.</p><p>A typical <code class="application">PyPop</code> run might take
    anywhere from a few of minutes to a few hours, depending on how
    large your data set is and who else is using the system at the
    same time.  Note that performing the
    <code class="literal">allPairwiseLDWithPermu</code> test may take several
    <span class="bold"><strong>days</strong></span> if you have highly
    polymorphic loci in your data set.</p></div></div></div><div class="navfooter"><hr><table width="100%" summary="Navigation footer"><tr><td width="40%" align="left"><a accesskey="p" href="install-from-source.html">Prev</a> </td><td width="20%" align="center"> </td><td width="40%" align="right"> <a accesskey="n" href="guide-usage-datafile.html">Next</a></td></tr><tr><td width="40%" align="left" valign="top">1.2. Installing from source </td><td width="20%" align="center"><a accesskey="h" href="pypop-guide.html">Home</a></td><td width="40%" align="right" valign="top"> 2.2. The data file</td></tr></table></div></body></html>
