<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><title>PyPop User Guide</title><link rel="stylesheet" href="style.css" type="text/css"><meta name="generator" content="DocBook XSL Stylesheets V1.74.0"><link rel="home" href="pypop-guide.html" title="PyPop User Guide"><link rel="next" href="guide-preface.html" title="Preface"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div class="navheader"><table width="100%" summary="Navigation header"><tr><th colspan="3" align="center">PyPop User Guide</th></tr><tr><td width="20%" align="left"> </td><th width="60%" align="center"> </th><td width="20%" align="right"> <a accesskey="n" href="guide-preface.html">Next</a></td></tr></table><hr></div><div class="book" lang="en"><div class="titlepage"><div><div><h1 class="title"><a name="guide"></a><code class="application">PyPop</code> User Guide</h1></div><div><h2 class="subtitle">User Guide for Python for Population Genomics</h2></div><div><div class="authorgroup"><div class="author"><h3 class="author"><span class="firstname">Alex</span> <span class="othername">K.</span> <span class="surname">Lancaster</span></h3><div class="affiliation"><span class="orgname">University of California, Berkeley<br></span> <span class="orgdiv">Department of Integrative Biology<br></span></div></div><div class="author"><h3 class="author"><span class="firstname">Mark</span> <span class="othername">P.</span> <span class="surname">Nelson</span></h3><div class="affiliation"><span class="orgname">University of California, Berkeley<br></span> <span class="orgdiv">Department of Integrative Biology<br></span></div></div><div class="author"><h3 class="author"><span class="firstname">Diogo</span> <span class="surname">Meyer</span></h3><div class="affiliation"><span class="orgname">University of California, Berkeley<br></span> <span class="orgdiv">Department of Integrative Biology<br></span></div></div><div class="author"><h3 class="author"><span class="firstname">Richard</span> <span class="othername">M.</span> <span class="surname">Single</span></h3><div class="affiliation"><span class="orgname">University of Vermont<br></span> <span class="orgdiv">Department of Medical Biostatistics<br></span></div></div><div class="author"><h3 class="author"><span class="firstname">Owen</span> <span class="othername">D.</span> <span class="surname">Solberg</span></h3><div class="affiliation"><span class="orgname">University of California, Berkeley<br></span> <span class="orgdiv">Department of Integrative Biology<br></span></div></div></div></div><div><p class="releaseinfo">Documenting version 0.7.0 of PyPop</p></div><div><p class="copyright">Copyright © 2003, 2004, 2005, 2006, 2007, 2008, 2009 Regents of the University of California</p></div><div><div class="legalnotice"><a name="id2942089"></a><p class="legalnotice-title"><b>Licence terms for PyPop documentation</b></p><p>Permission is granted to copy, distribute and/or modify this
    document under the terms of the GNU Free Documentation License,
    Version 1.2 or any later version published by the Free Software
    Foundation; with no Invariant Sections no Front-Cover Texts and no
    Back-Cover Texts.  A copy of the license is included in:
    <a class="xref" href="gfdl.html" title="A.2. GNU Free Documentation License">Section A.2, “GNU Free Documentation License”</a>.</p></div></div><div><p class="pubdate">11 May 2009</p></div></div><hr></div><div class="toc"><p><b>Table of Contents</b></p><dl><dt><span class="preface"><a href="guide-preface.html">Preface</a></span></dt><dd><dl><dt><span class="section"><a href="guide-preface.html#guide-preface-intro">1. Introduction</a></span></dt><dt><span class="section"><a href="guide-preface-howto.html">2. How to use this guide</a></span></dt><dt><span class="section"><a href="guide-preface-news.html">3. Recent changes to PyPop</a></span></dt><dt><span class="section"><a href="guide-preface-authors.html">4. Authors of software components</a></span></dt></dl></dd><dt><span class="chapter"><a href="guide-chapter-install.html">1. Installing PyPop</a></span></dt><dd><dl><dt><span class="section"><a href="guide-chapter-install.html#install-standalone">1.1. Installing standalone binary</a></span></dt><dd><dl><dt><span class="section"><a href="guide-chapter-install.html#install-standalone-linux">1.1.1. Installing on GNU/Linux</a></span></dt><dt><span class="section"><a href="guide-chapter-install.html#install-standalone-windows">1.1.2. Installing on Windows</a></span></dt></dl></dd><dt><span class="section"><a href="install-from-source.html">1.2. Installing from source</a></span></dt><dd><dl><dt><span class="section"><a href="install-from-source.html#install-from-source-sysreq">1.2.1. System requirements</a></span></dt><dt><span class="section"><a href="install-from-source.html#install-from-source-install">1.2.2. Installation</a></span></dt><dt><span class="section"><a href="install-from-source.html#install-from-source-tests">1.2.3. Test suite</a></span></dt><dt><span class="section"><a href="install-from-source.html#install-from-source-bugs">1.2.4. Contributions, bug reports</a></span></dt><dt><span class="section"><a href="install-from-source.html#install-from-source-distribution">1.2.5. Distribution structure</a></span></dt></dl></dd></dl></dd><dt><span class="chapter"><a href="guide-chapter-usage.html">2. Getting started with PyPop</a></span></dt><dd><dl><dt><span class="section"><a href="guide-chapter-usage.html#guide-usage-intro">2.1. Introduction</a></span></dt><dd><dl><dt><span class="section"><a href="guide-chapter-usage.html#id3144373">2.1.1. Interactive mode</a></span></dt><dt><span class="section"><a href="guide-chapter-usage.html#id3129007">2.1.2. Batch mode</a></span></dt><dt><span class="section"><a href="guide-chapter-usage.html#guide-usage-intro-run-details">2.1.3. What happens when you run
   PyPop?</a></span></dt></dl></dd><dt><span class="section"><a href="guide-usage-datafile.html">2.2. The data file</a></span></dt><dd><dl><dt><span class="section"><a href="guide-usage-datafile.html#id3155140">2.2.1. Sample files</a></span></dt><dt><span class="section"><a href="guide-usage-datafile.html#id3135827">2.2.2. Missing data</a></span></dt></dl></dd><dt><span class="section"><a href="guide-usage-configfile.html">2.3. The configuration file</a></span></dt><dd><dl><dt><span class="section"><a href="guide-usage-configfile.html#config-minimal">2.3.1. A minimal configuration file</a></span></dt><dt><span class="section"><a href="guide-usage-configfile.html#config-advanced">2.3.2. Advanced options</a></span></dt></dl></dd></dl></dd><dt><span class="chapter"><a href="guide-chapter-instructions.html">3. Interpreting PyPop output</a></span></dt><dd><dl><dt><span class="sect1"><a href="guide-chapter-instructions.html#instructions-pop-summary">3.1. Population summary</a></span></dt><dt><span class="sect1"><a href="instructions-locus-info.html">3.2. Single locus analyses</a></span></dt><dd><dl><dt><span class="sect2"><a href="instructions-locus-info.html#instructions-allelecounts">3.2.1. Basic allele count information</a></span></dt><dt><span class="sect2"><a href="instructions-locus-info.html#instructions-hardyweinberg">3.2.2. Chi-square test for deviation from Hardy-Weinberg
    proportions (HWP).</a></span></dt><dt><span class="sect2"><a href="instructions-locus-info.html#instructions-hardyweinberg-exact">3.2.3. Exact test for deviation from HWP</a></span></dt><dt><span class="sect2"><a href="instructions-locus-info.html#instructions-homozygosity">3.2.4. The Ewens-Watterson homozygosity test of neutrality</a></span></dt></dl></dd><dt><span class="sect1"><a href="instructions-haplo.html">3.3. Multi-locus analyses</a></span></dt><dd><dl><dt><span class="sect2"><a href="instructions-haplo.html#instructions-pairwise-ld">3.3.1. All pairwise LD</a></span></dt><dt><span class="sect2"><a href="instructions-haplo.html#instructions-haplotype-freqs">3.3.2. Haplotype frequency estimation</a></span></dt></dl></dd></dl></dd><dt><span class="bibliography"><a href="references.html">References</a></span></dt><dt><span class="appendix"><a href="licenses.html">A. License terms</a></span></dt><dd><dl><dt><span class="section"><a href="licenses.html#gpl">A.1. GNU General Public License</a></span></dt><dt><span class="section"><a href="gfdl.html">A.2. GNU Free Documentation License</a></span></dt></dl></dd></dl></div><div class="list-of-examples"><p><b>List of Examples</b></p><dl><dt>2.1. <a href="guide-usage-datafile.html#data-minimal-noheader-noids">Multi-locus allele-level genotype data</a></dt><dt>2.2. <a href="guide-usage-datafile.html#data-minimal-noheader">Multi-locus allele-level HLA genotype data with sample
    information</a></dt><dt>2.3. <a href="guide-usage-datafile.html#data-hla">Multi-locus allele-level HLA genotype data with sample and
    header information</a></dt><dt>2.4. <a href="guide-usage-datafile.html#data-hla-microsat">Multi-locus allele-level HLA genotype and microsatellite
    genotype data with header information</a></dt><dt>2.5. <a href="guide-usage-datafile.html#data-nucleotide">Sequence genotype data with header information</a></dt><dt>2.6. <a href="guide-usage-datafile.html#data-allelecount">Allele count data</a></dt><dt>2.7. <a href="guide-usage-configfile.html#config-minimal-example">Minimal config.ini file</a></dt><dt>3.1. <a href="guide-chapter-instructions.html#id3174421">Population summary sample output</a></dt><dt>3.2. <a href="instructions-locus-info.html#id3174471">Basic locus information sample output</a></dt><dt>3.3. <a href="instructions-locus-info.html#id3174542">Sample output of Hardy-Weinberg genotype table</a></dt><dt>3.4. <a href="instructions-locus-info.html#id3174572">Sample output of HW genotype classes</a></dt><dt>3.5. <a href="instructions-locus-info.html#id3174982">Sample output for exact test using
     gthwe</a></dt><dt>3.6. <a href="instructions-locus-info.html#id3175013">Sample output for exact test using the
     Arlequin implementation</a></dt><dt>3.7. <a href="instructions-locus-info.html#id3175289">Sample output of homozygosity test from Monte-Carlo implementation</a></dt><dt>3.8. <a href="instructions-locus-info.html#id3175358">Sample output of homozygosity test from simulation look-up
      tables (disabled by default)</a></dt><dt>3.9. <a href="instructions-haplo.html#id3175450">Sample output of all pairwise LD</a></dt><dt>3.10. <a href="instructions-haplo.html#id3175772">Sample output of haplotype estimation parameters</a></dt><dt>3.11. <a href="instructions-haplo.html#id3175816">Sample output of estimated haplotype frequencies</a></dt></dl></div><div class="list-of-equations"><p><b>List of Equations</b></p><dl><dt>3.0. <a href="instructions-haplo.html#id3175574"></a></dt><dt>3.0. <a href="instructions-haplo.html#id3175631"></a></dt></dl></div></div><div class="navfooter"><hr><table width="100%" summary="Navigation footer"><tr><td width="40%" align="left"> </td><td width="20%" align="center"> </td><td width="40%" align="right"> <a accesskey="n" href="guide-preface.html">Next</a></td></tr><tr><td width="40%" align="left" valign="top"> </td><td width="20%" align="center"> </td><td width="40%" align="right" valign="top"> Preface</td></tr></table></div></body></html>
